---
title: "CBO_Seurat_EXPLORE1_50K"
author: "JJ"
date: "07/08/2022"
output: html_document
---

```{r graphical_output}
options(bitmapType='cairo-png') # this line of code needed BEFORE knitr code in chunk below
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = "png",
                      dpi = 300,
                      echo = TRUE,   # prints the code (if FALSE then only code output is shown)
                      cache = TRUE)
```

```{r libraries}
suppressPackageStartupMessages({
library(Seurat)
library(glmGamPoi)
library(Matrix)
library(clustree)
library(ggVennDiagram)
library(patchwork)
library(tidyverse)  
})

```

```{r load_filteredSeuratObject}
# load full filtered (genes and cells) seurat object

Fullfilt.cbo.seurat <- readRDS("outputs50K/CBO/filtered.cbo.seurat.rds")
```

```{r DimPlot1GeneCellFiltered}
# Perform SCTransform
# Perform dimensionality reduction by PCA and UMAP embedding
# FindClusters is run with only the default resolution parameter (= 0.8)
# 10 clusters identified

filtered.cbo.seurat <- SCTransform(Fullfilt.cbo.seurat, method = "glmGamPoi", verbose = FALSE)

filtered.cbo.seurat <- RunPCA(filtered.cbo.seurat, verbose = FALSE)

filtered.cbo.seurat <- RunUMAP(filtered.cbo.seurat, dims = 1:30, verbose = FALSE)

filtered.cbo.seurat <- FindNeighbors(filtered.cbo.seurat, dims = 1:30, verbose = FALSE)

filtered.cbo.clusters.seurat <- FindClusters(filtered.cbo.seurat, verbose = FALSE)

DimPlot(filtered.cbo.clusters.seurat, label = TRUE) # '+ NoLegend()' is an option to insert here
```

```
# saved the SCTransformed, clustered 'filtered.cbo.clusters.seurat' Seurat object 
# to the folder 'outputs50K'. Running DimPlot() command on this object should generate the same
# UMAP projection every time

saveRDS(filtered.cbo.clusters.seurat, file = "./outputs50K/CBO/filtered.cbo.clusters.seurat.rds")
```

```{r}
# load the SCTransformed, fully filtered seurat object

filt.cbo.sct.clusters <- readRDS("outputs50K/CBO/filtered.cbo.clusters.seurat.rds")
```

```{r ElbowPlotGeneCellFilt}
# elbow plot shows that there is an elbow at PC 15, but the graph does not level off after
# PC 15 showing there is some additional information to ~ PC 40.
# The input object is the CLUSTERED saved seurat object, and this gives the same result as
# the seurat object generated by the FindNeighbours() command.

ElbowPlot(filt.cbo.sct.clusters, ndims = 50, reduction = "pca")
```

```{r}
# Find positive markers for each of the 10 clusters
cbo.cluster.markers <- FindAllMarkers(filt.cbo.sct.clusters, min.pct = 0.25, only.pos = TRUE)
```

```{r}
# Cell counts per cluster.
# there are 3070 cells in total, which is correct, and the cell numbers per cluster are shown below.
# Two clusters are present with < 100 cells

cell.num <- table(filt.cbo.sct.clusters@active.ident)
cell.num
sum(cell.num)
```

```
# save cbo.cluster.markers as a csv file

write.csv(cbo.cluster.markers, "./outputs50K/CBO/cbo.sct.cluster.markers.csv", quote = F)
```

```{r}
# load cluster markers csv file

cbo.sct.cluster.markers <- read.csv("outputs50K/CBO/cbo.sct.cluster.markers.csv", 
                                    header = TRUE, 
                                    stringsAsFactors = FALSE)
```

```{r}
# cluster 0 top10 markers
# cluster 0 = probable roof plate/choroid plexus. n = 654 cells
cluster0.markers <- cbo.cluster.markers[cbo.cluster.markers$cluster == 0, ] %>%
  head(n=10)

cluster0.markers
```

```{r}
# cluster 1 top10 markers
# cluster 1 probably 4th ventricle derivatives/choroid plexus (+/- Purkinje neurons), n = 458 cells
cluster1.markers <- cbo.cluster.markers[cbo.cluster.markers$cluster == 1, ] %>%
  head(n=10)

cluster1.markers
```

```{r}
# cluster 2 top10 markers
# cluster 2 hard to classify, could be glia? Most markers are non-specific (expressed in pct1 and pct2)
# n = 422 cells
cluster2.markers <- cbo.cluster.markers[cbo.cluster.markers$cluster == 2, ] %>%
  head(n=10)

cluster2.markers
```

```{r}
# cluster 3 top10 markers
# cluster 3: these markers expressed in GABAergic/inhibitory interneurons and glutamatergic (NRXN1)
# of cerebellum
cluster3.markers <- cbo.cluster.markers[cbo.cluster.markers$cluster == 3, ] %>%
  head(n=10)

cluster3.markers
```

```{r}
# cluster 4 top10 markers
# cluster 4 - undetermined, n = 331 cells
cluster4.markers <- cbo.cluster.markers[cbo.cluster.markers$cluster == 4, ] %>%
  head(n=10)

cluster4.markers
```

```{r}
# cluster 5 top10 markers
# cluster 5 - many ribosomal genes, also pct1 and pct2 very similar
cluster5.markers <- cbo.cluster.markers[cbo.cluster.markers$cluster == 5, ] %>%
  head(n=10)

cluster5.markers
```

```{r}
# cluster 6 top10 markers
# cluster 6 most likely GCP/granule cells, n = 272 cells
cluster6.markers <- cbo.cluster.markers[cbo.cluster.markers$cluster == 6, ] %>%
  head(n=10)

cluster6.markers
```

```{r}
# cluster 7 top10 markers
# cluster 7 could be rhombic lip, GCP, proliferative granule cells
cluster7.markers <- cbo.cluster.markers[cbo.cluster.markers$cluster == 7, ] %>%
  head(n=10)

cluster7.markers
```

```{r}
# cluster 8 top10 markers
# cluster 8 is probably glia
cluster8.markers <- cbo.cluster.markers[cbo.cluster.markers$cluster == 8, ] %>%
  head(n=10)

cluster8.markers
```

```{r}
# cluster 9 top10 markers
# cluster 9 likely ciliated cells/ependyma/choroid plexus
cluster9.markers <- cbo.cluster.markers[cbo.cluster.markers$cluster == 9, ] %>%
  head(n=10)

cluster9.markers
```

```{r}
# cluster 5 contained many ribosomal genes, so TRY FILTERING ON RIBOSOMAL GENE PERCENTAGE
#----------------------------------------------------------------------------------------

# use the full filtered (for cells and genes and mitoch percent) seurat object
```

```{r}
# load the cell, gene and mitoch gene percentage seurat object

filt.cbo.seurat <- readRDS("outputs50K/CBO/filtered.cbo.seurat.rds")
```

```{r}
# calculate ribosomal percentage

filt.cbo.seurat[["percent.ribo"]] <- PercentageFeatureSet(filt.cbo.seurat, pattern = "^RP[SL]")
```

```{r}
summary(filt.cbo.seurat@meta.data$percent.ribo)
```


```{r ViolinPlot1CBOnoMitochNoRibo}
VlnPlot(filt.cbo.seurat, features = c("percent.mt", "percent.ribo"), ncol = 2)
```

```{r}
# In https://kb.10xgenomics.com/hc/en-us/articles/218169723-What-fraction-of-reads-map-to-ribosomal-proteins- on the 10X Genomics scRNA-seq website there is a table of percent ribosomal genes in different cell types. For neurons the upper limit is 20%. So try filtering 'filt.cbo.seurat' for ribosomal genes and repeat the clustering.

# View metadata data frame, stored in object@meta.data
metadata <- filt.cbo.seurat[[]]

# number of cells with percent.ribo > 20%: 1551 cells
# number of cells with percent.ribo < 20%: 1518 cells

nrow(metadata[metadata$percent.ribo > 20, ])

nrow(metadata[metadata$percent.ribo < 20, ])
```

```
# saved the last seurat object, 'full.filt.cbo.seurat'

saveRDS(full.filt.cbo.seurat, file = "./outputs50K/CBO/full.filt.cbo.seurat.rds")
```

```
# saved the clustered 'full.filt.cbo.clusters.seurat' Seurat object to the folder 'outputs50K'

saveRDS(full.filt.cbo.clusters.seurat, file = "./outputs50K/CBO/full.ribo.filt.cbo.clusters.seurat.rds")
```

```
# save cbo.cluster.markers.ribo20 as a csv file. This object has percent.ribo < 20

write.csv(cbo.cluster.markers.ribo20, "./outputs50K/CBO/cbo.cluster.markers.ribo20.csv", quote = F)
```

```{r remove_ribo_genes}
# FILTER OUT ALL RIBOSOMAL GENES, THEN ALL MITOCHONDRIAL GENES (ON THE CELLS WHICH HAVE ALREADY BEEN REMOVED
# THAT HAVE PERCENT.MITO < 10% BEFORE DOING THE CLUSTERING), DON'T REMOVE ANY MORE ACTUAL CELLS
#-------------------------------------------------------------------------------------

# remove ribosomal genes first
filt.noRibo.cbo <- Fullfilt.cbo.seurat[ ! grepl('^RP[SL]', rownames(Fullfilt.cbo.seurat)), ]
```

```{r}
# check ribosomal genes removed
# this has worked - 97 ribosomal genes have been removed from the new Seurat object
dim(Fullfilt.cbo.seurat)
dim(filt.noRibo.cbo)
```

```{r remove_mitoch_genes}
# remove mitochondrial genes 
filt.noRiboMito.cbo <- filt.noRibo.cbo[ ! grepl("^MT-", rownames(filt.noRibo.cbo)), ]
```

```{r}
# check if ribosome and mitochondrial gene filtering works - it does!
dim(filt.noRibo.cbo)
dim(filt.noRiboMito.cbo)
```

```{r DimPlot2NoRiboNoMito}
# Perform SCTransform
# Perform dimensionality reduction by PCA and UMAP embedding
# FindClusters is run with only the default resolution parameter (= 0.8)
# 11 clusters identified

filt.noRiboMito.cbo <- SCTransform(filt.noRiboMito.cbo, method = "glmGamPoi", verbose = FALSE)

filt.noRiboMito.cbo <- RunPCA(filt.noRiboMito.cbo, verbose = FALSE)

filt.noRiboMito.cbo <- RunUMAP(filt.noRiboMito.cbo, dims = 1:30, verbose = FALSE)

filt.noRiboMito.cbo <- FindNeighbors(filt.noRiboMito.cbo, dims = 1:30, verbose = FALSE)

filt.noRiboMito.cbo.clusters <- FindClusters(filt.noRiboMito.cbo, verbose = FALSE)

DimPlot(filt.noRiboMito.cbo.clusters, label = TRUE) # '+ NoLegend()' is an option to insert here
```

```
# SAVE THE 'filt.noRiboMito.cbo' seurat object - the cells expressing >10% mitochondrial genes are removed
# and all ribosome genes and mitochondrial genes have been removed from the count matrix. This object is related
# to the coding line 417 seurat object ('FindNeighbours()' output).

saveRDS(filt.noRiboMito.cbo, file = "./outputs50K/CBO/filt.NoRiboNoMitochGenes.cbo.rds")
```
```{r}
# load filt.NoRiboNoMitochGenes.cbo.rds

filt.noRiboMito.cbo <- readRDS("outputs50K/CBO/filt.NoRiboNoMitochGenes.cbo.rds")
```

```
# SAVE THE 'filt.noRiboMito.cbo.clusters' seurat object

saveRDS(filt.noRiboMito.cbo.clusters, file = "./outputs50K/CBO/filt.NoRiboNoMitoGenes.cbo.clusters.rds")
```

```{r}
# load 'filt.NoRiboNoMitoGenes.cbo.clusters.rds'

filt.noRiboMito.cbo.clusters <- readRDS("outputs50K/CBO/filt.NoRiboNoMitoGenes.cbo.clusters.rds")
```

```{r}
# Find positive markers for each of the 11 clusters of 'filt.noRiboMito.cbo.clusters'

cbo.cluster.markers.noRiboMito <- FindAllMarkers(filt.noRiboMito.cbo.clusters, min.pct = 0.25, only.pos = TRUE)
```

```{r}
# get cell counts per cluster
# there are 3070 cells in total, which is correct, and the cell numbers per cluster are shown below.
# Clusters 9 (actually 10) and 10 (actually 11) are the smallest

cell.num.riboFilt <- table(filt.noRiboMito.cbo.clusters@active.ident)
cell.num.riboFilt
sum(cell.num.riboFilt)
```

```
# save 'cbo.cluster.markers.noRiboMito' as a csv file. This object has NO mitochondrial or ribosomal genes.
# The object has already been filtered to keep cells with mitoch gene expression < 10%.

write.csv(cbo.cluster.markers.noRiboMito, "./outputs50K/CBO/cbo.cluster.markers.NoRiboNoMito.csv", quote = F)
```

```{r}
# load 'cbo.cluster.markers.NoRiboNoMito.csv'

cbo.cluster.markers.noRiboMito <- read.csv("outputs50K/CBO/cbo.cluster.markers.NoRiboNoMito.csv", 
                                       header = TRUE, 
                                       stringsAsFactors = FALSE)
```

```{r}
# cluster 0 top10 markers - csv file is gene and cell filtered, NO mitoch genes, NO ribo genes
# cluster 0 = roof plate/choroid plexus
cluster0.markers <- cbo.cluster.markers.noRiboMito[cbo.cluster.markers.noRiboMito$cluster == 0, ] %>%
  head(n=10)

cluster0.markers
```

```{r}
# cluster 1 top10 markers - csv file is gene and cell filtered, NO mitoch genes, NO ribo genes
# cluster 1 = 4th ventricle derivatives/choroid plexus
cluster1.markers <- cbo.cluster.markers.noRiboMito[cbo.cluster.markers.noRiboMito$cluster == 1, ] %>%
  head(n=10)

cluster1.markers
```

```{r}
# cluster 2 top10 markers - csv file is gene and cell filtered, NO mitoch genes, NO ribo genes
# cluster 2 = inhibitory cerebellar nuclei/GABAergic lineages
cluster2.markers <- cbo.cluster.markers.noRiboMito[cbo.cluster.markers.noRiboMito$cluster == 2, ] %>%
  head(n=10)

cluster2.markers
```

```{r}
# cluster 3 top10 markers - csv file is gene and cell filtered, NO mitoch genes, NO ribo genes
# cluster 3 = hard to classify, glial and PN markers. Could these be progenitors which express radial
# glia markers
cluster3.markers <- cbo.cluster.markers.noRiboMito[cbo.cluster.markers.noRiboMito$cluster == 3, ] %>%
  head(n=20)

cluster3.markers
```

```{r}
# cluster 4 top10 markers - csv file is gene and cell filtered, NO mitoch genes, NO ribo genes
# cluster 4 = could be glia??
cluster4.markers <- cbo.cluster.markers.noRiboMito[cbo.cluster.markers.noRiboMito$cluster == 4, ] %>%
  head(n=20)

cluster4.markers
```

```{r}
# cluster 5 top10 markers - csv file is gene and cell filtered, NO mitoch genes, NO ribo genes
# cluster 5 = probable glia
cluster5.markers <- cbo.cluster.markers.noRiboMito[cbo.cluster.markers.noRiboMito$cluster == 5, ] %>%
  head(n=20)

cluster5.markers
```

```{r}
# cluster 6 top10 markers - csv file is gene and cell filtered, NO mitoch genes, NO ribo genes
# cluster 6 = GCP/granule neuron
cluster6.markers <- cbo.cluster.markers.noRiboMito[cbo.cluster.markers.noRiboMito$cluster == 6, ] %>%
  head(n=20)

cluster6.markers
```

```{r}
# cluster 7 top10 markers - csv file is gene and cell filtered, NO mitoch genes, NO ribo genes
# cluster 7 = rhombic lip, GCP, proliferative granule cells
cluster7.markers <- cbo.cluster.markers.noRiboMito[cbo.cluster.markers.noRiboMito$cluster == 7, ] %>%
  head(n=10)

cluster7.markers
```

```{r}
# cluster 8 top10 markers - csv file is gene and cell filtered, NO mitoch genes, NO ribo genes
# cluster 8 = Glutamatergic (granule,DCN,UBC) lineage

cluster8.markers <- cbo.cluster.markers.noRiboMito[cbo.cluster.markers.noRiboMito$cluster == 8, ] %>%
  head(n=10)

cluster8.markers
```

```{r cluster9_cbo.cluster.markers.noRiboMito_top10markers}
# cluster 9 top10 markers - csv file is gene and cell filtered, NO mitoch genes, NO ribo genes
# cluster 9 = glial cells
cluster9.markers <- cbo.cluster.markers.noRiboMito[cbo.cluster.markers.noRiboMito$cluster == 9, ] %>%
  head(n=10)

cluster9.markers
```

```{r cluster10_cbo.cluster.markers.noRiboMito_top10markers}
# cluster 10 top10 markers - csv file is gene and cell filtered, NO mitoch genes, NO ribo genes
# cluster 10 = ciliated cells
cluster10.markers <- cbo.cluster.markers.noRiboMito[cbo.cluster.markers.noRiboMito$cluster == 10, ] %>%
  head(n=10)

cluster10.markers
```

```{r DotplotSeppTFmarkers_some}
# using the Sepp et al paper (Extended Fig 8g), test if markers from that paper are expressed 
# in the seurat object filtered to remove all ribosomal and mitochondrial genes, and where cells which
# have mitoch genes > 10% are excluded.
# MUST use the already clustered seurat object.
# I used just the transcription factor (TF) markers (sepp.TF.markers)

sepp.TF.markers <- c("PAX3", "RFX4", "TCFL71", "GLIS3", "SOX6", "TCF7L2", "HOPX", "TSC22D4", "ID4",
             "ESRRB", "FOXP2", "FOXP4", "MEF2C",
             "PAX2", "TFAP2A", "TFAP2B", "BHLHE22", "PRDM8",
             "PAX6", "BARHL1", "E2F3", "ST18", "SATB2", "ZBTB18", "NEUROD1", "ETV1", "KLF9", "NR3C1")


DotPlot(object = filt.noRiboMito.cbo.clusters, features = sepp.TF.markers) +
  coord_flip()
```

```{r DotplotSeppMarkers+cilia_many, fig.height=10}
# from the Sepp paper repeat Dotplot with TF and some key non-TF markers
# added markers of ciliated cells from cluster10 above.
# used the argument col.min = 0 so that average expression scale is from 0 to 2.5 in key

sepp.all.markers <- c("TFAP2A", "PAX3", "RFX4", "TCFL71", "TCF7L2", "GLIS3", "SOX6", "HOPX", "TSC22D4",  
                      "ID4", "NCKAP5", "SLIT2", "CLYBL", "SLC1A3", "SLC6A11", "AQP4",  #astroglial/progenitor
                      
                      "ESRRB", "FOXP1", "FOXP2", "FOXP4", "MEF2C", "RORB", "CDH9", "ETV1", "EBF1", "EBF2", 
                      "SKOR2", "DAB1", "IPTR",   # Purkinje
                      
                      "NOTCH1",    # VZ neuroblast
             
                      "PAX2", "NFIA", "NFIB", "TFAP2B", "BHLHE22", "PRDM8",  # interneuron
                      
                      "ZFHX3", "ZFHX4",   # GABAergic deep nuclei neurons
             
                      "TOP2A", "ATOH1", "LMX1A", "OTX2","ETV6", "PAX6", "BARHL1", "E2F3", "ST18", "SATB2", 
                      "ZBTB18", "NEUROD1", "ZIC1", "ZIC4", "GABAR6", "CNTN1", "CNTN2", "KLF9", "NR3C1", "LMO3", 
                      "SYNPR", "KCNIP4", "HCRTR2", 
                      "EOMES", "TRPC3", "GRM1")
                      # UBC/GCP/granule/glut deep cerebellar nuclei                                                                   
seppAll.cilia.stem.markers <- c(sepp.all.markers, "CCNO", "CFAP299", "ROPN1L", "C11orf88", "ANKRD66", "C20orf85", 
                                "STOML3", "TEKT4", "SNTN", "MEIS1", "POU3F2", "HOXB2", "MEIS2", "GRIA4")

DotPlot(object = filt.noRiboMito.cbo.clusters, features = seppAll.cilia.stem.markers, col.min = 0) +
  coord_flip()
```

```{r ViolinPlot_GranuleLineageByCluster, fig.height=8}
# plot of the clusters containing granule/RL lineage suggests RL-derived clusters are 6, 7, 8
# CDH5 (marker of ONS76 cells) is not expressed but CDH1 is (also a marker of ONS76 cells).
# CD44 is not expressed or expressed very little.
# ITGB1 is expressed at quite a high level across most cell types

VlnPlot(filt.noRiboMito.cbo.clusters, features = c("HOXB2", "RRM2", "NEUROD1", "CNTN1", "CNTN2", "MEIS1", "MEIS2", "GRIA4", "POU3F2", "ACTC1"))
```

```{r}
# attempt to use Clustree package to find the right number of clusters
# -----------------------------------------------------------------------------------------------

# Select a range of resolutions
resolution.range <- seq(from = 0, to = 1.6, by = 0.2)

# Find clusters using a range of resolutions, the clustering information is in the metadata slot
filt.noRiboMito.cbo.clusters <- FindClusters(object = filt.noRiboMito.cbo.clusters, 
                                             resolution = resolution.range,
                                             verbose = FALSE)
```

```{r ClustreeCBO, fig.height=12}
# the clustering tree finds 9 clusters before some nodes have multiple incoming edges.
# note that at res = 1 there are 14 clusters

clustree(filt.noRiboMito.cbo.clusters)
```

```{r}
# at res = 1.0 there are 14 clusters.
# repeat the clustering at res = 1 and try the Sepp classification again

filt.noRiboMito.cbo.clusters.res1 <- FindClusters(filt.noRiboMito.cbo, resolution = 1, verbose = FALSE)
```

```{r DotplotSeppMarkersCilia_Res1.0, fig.height=10}
# from the Sepp paper repeat Dotplot with TF and some key non-TF markers
# added markers of ciliated cells from cluster10 above.
# used the argument col.min = 0 so that average expression scale is from 0 to 2.5 in key

sepp.all.markers <- c("TFAP2A", "PAX3", "RFX4", "TCFL71", "TCF7L2", "GLIS3", "SOX6", "HOPX", "TSC22D4",  
                      "ID4", "NCKAP5", "SLIT2", "CLYBL", "SLC1A3", "SLC6A11", "AQP4",  #astroglial/progenitor
                      
                      "ESRRB", "FOXP1", "FOXP2", "FOXP4", "MEF2C", "RORB", "CDH9", "ETV1", "EBF1", "EBF2", 
                      "SKOR2", "DAB1", "IPTR",   # Purkinje
                      
                      "NOTCH1",    # VZ neuroblast
             
                      "PAX2", "NFIA", "NFIB", "TFAP2B", "BHLHE22", "PRDM8",  # interneuron
                      
                      "ZFHX3", "ZFHX4",   # GABAergic deep nuclei neurons
             
                      "TOP2A", "ATOH1", "LMX1A", "OTX2","ETV6", "PAX6", "BARHL1", "E2F3", "ST18", "SATB2", 
                      "ZBTB18", "NEUROD1", "ZIC1", "ZIC4", "GABAR6", "CNTN1", "CNTN2", "KLF9", "NR3C1", "LMO3", 
                      "SYNPR", "KCNIP4", "HCRTR2", 
                      "EOMES", "TRPC3", "GRM1")
                      # UBC/GCP/granule/glut deep cerebellar nuclei                                                                   
seppAll.cilia.stem.markers <- c(sepp.all.markers, "CCNO", "CFAP299", "ROPN1L", "C11orf88", "ANKRD66", "C20orf85", 
                                "STOML3", "TEKT4", "SNTN", "MEIS1", "POU3F2", "HOXB2", "MEIS2", "GRIA4") # stemness markers and ciliated cell markers

DotPlot(object = filt.noRiboMito.cbo.clusters.res1, features = seppAll.cilia.stem.markers, col.min = 0) +
  coord_flip()
```

```
# save the cluster 6 (granule/NEUROD1) markers as input for VennDiagrams comparing cluster 6 in CBO
# with clusters 2, 6, 9, 10 of ONS76-CBO sample
cluster6.markers <- cbo.cluster.markers.noRiboMito[cbo.cluster.markers.noRiboMito$cluster == 6, ]
write.csv(cluster6.markers, file = "./outputs50K/DEG_VennDiag/cbo_cluster6_granule.csv")
```











