---
title: "DAOY-CBO_Seurat_EXPLORE1_50K"
author: "JJ"
date: "12/08/2022"
output: html_document
---

```{r graphical_output}
options(bitmapType='cairo-png') # this line of code needed BEFORE knitr code in chunk below
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = "png",
                      dpi = 300,
                      echo = TRUE,   # prints the code (if FALSE then only code output is shown)
                      cache = TRUE)
```

```{r libraries}
suppressPackageStartupMessages({
library(Seurat)
library(glmGamPoi)
library(Matrix)
library(patchwork)
library(tidyverse)
})
```

```{r loadFilteredSeuratObject}
# load full filtered (genes and cells) seurat object

filt.cboDaoy.seurat <- readRDS("outputs50K/DAOY-CBO/filtCellsGenes.daoyCbo.seurat.rds")
```

```{r calcRibosomalpercent}
# calculate ribosomal percentage

filt.cboDaoy.seurat[["percent.ribo"]] <- PercentageFeatureSet(filt.cboDaoy.seurat, pattern = "^RP[SL]")
```

```{r}
summary(filt.cboDaoy.seurat@meta.data$percent.ribo)
```

```{r}
VlnPlot(filt.cboDaoy.seurat, features = c("percent.mt", "percent.ribo"), ncol = 3)
```


Remove ribosomal and mitochondrial genes as the clustering works better without these, and is then free of top 'markers' which are ribosome or mitochondrial genes.
```{r remove_ribo_genes}
# FILTER OUT ALL RIBOSOMAL GENES, THEN ALL MITOCHONDRIAL GENES (ON THE CELLS WHICH ARE REMAINING
# THAT HAVE PERCENT.MITO < 10%) BEFORE DOING THE CLUSTERING, AND DON'T REMOVE ANY MORE ACTUAL CELLS
#-------------------------------------------------------------------------------------

# remove ribosomal genes first
filt.noRibo.cboDaoy <- filt.cboDaoy.seurat[ ! grepl('^RP[SL]', rownames(filt.cboDaoy.seurat)), ]
```

```{r check_ribo_removed}
# this has worked - 97 ribosomal genes have been removed from the new Seurat object
# there are 3056 cells
dim(filt.cboDaoy.seurat)
dim(filt.noRibo.cboDaoy)
```

```{r remove_mitoch_genes}
# remove mitochondrial genes 
filt.noRiboMito.cboDaoy <- filt.noRibo.cboDaoy[ ! grepl("^MT-", rownames(filt.noRibo.cboDaoy)), ]
```

```{r check_filtering}
# check if ribosome and mitochondrial gene filtering works - it does!
dim(filt.noRibo.cboDaoy)
dim(filt.noRiboMito.cboDaoy)
```

```
# save the no ribosome and no mitochondrial genes seurat object
saveRDS(filt.noRiboMito.cboDaoy, file = "./outputs50K/DAOY-CBO/filt.NoRiboNoMito.cboDaoy.rds")
```

```{r DimPlotSCTransformUMAPclusteringNoCCdiffRegressn}
# Perform SCTransform, but THIS OBJECT HAS NOT UNDERGONE CELL CYCLE REGRESSION 
# Perform dimensionality reduction by PCA and UMAP embedding
# FindClusters is run with only the default resolution parameter (= 0.8)
# 13 clusters identified

daoyCbo.filt <- SCTransform(filt.noRiboMito.cboDaoy, method = "glmGamPoi", verbose = FALSE)

daoyCbo.filt <- RunPCA(daoyCbo.filt, verbose = FALSE)

daoyCbo.filt <- RunUMAP(daoyCbo.filt, dims = 1:30, verbose = FALSE)

daoyCbo.filt <- FindNeighbors(daoyCbo.filt, dims = 1:30, verbose = FALSE)

daoyCbo.filt.clusters <- FindClusters(daoyCbo.filt, verbose = FALSE)

DimPlot(daoyCbo.filt.clusters, label = TRUE) # '+ NoLegend()' is an option to insert here
```

```{r}
# SAVE THE 'filt.noRiboMito.cboDaoy.clusters' seurat object

saveRDS(daoyCbo.filt.clusters, file = "./outputs50K/DAOY-CBO/filt.NoRiboNoMito.cboDaoy.clusters.rds")
```

```{r load_filt.NoRiboNoMito.cboDaoy.clusters.rds}
filt.noRiboMito.cboDaoy.clusters <- readRDS("outputs50K/DAOY-CBO/filt.NoRiboNoMito.cboDaoy.clusters.rds")
```

```{r differential_genes_by_cluster_filt.noRiboMito.cbo.clusters}
# Find positive markers for each of the 11 clusters of 'filt.noRiboMito.cbo.clusters'

cboDaoy.cluster.markers.noRiboMito <- FindAllMarkers(filt.noRiboMito.cboDaoy.clusters, 
                                                     min.pct = 0.25, 
                                                     only.pos = TRUE)
```

```{r cellCountsPerCluster}
# there are 3056 cells in total, which is correct, and the cell numbers per cluster are shown below.

cell.num.riboFilt <- table(filt.noRiboMito.cboDaoy.clusters@active.ident)
cell.num.riboFilt
sum(cell.num.riboFilt)
```

```
# save 'cboDaoy.cluster.markers.noRiboMito' as a csv file. This object has NO mitochondrial or ribosomal genes.
# The object has already been filtered to keep cells with mitoch gene expression < 10%.

write.csv(cboDaoy.cluster.markers.noRiboMito, "./outputs50K/DAOY-CBO/cboDaoy.cluster.markers.NoRiboNoMito.csv", quote = F)
```

```{r loadCboDaoyClusterMarkersNoRiboMitoCSV}
cboDaoy.cluster.markers.noRiboMito <- read.csv("outputs50K/DAOY-CBO/cboDaoy.cluster.markers.NoRiboNoMito.csv", 
                                       header = TRUE, 
                                       stringsAsFactors = FALSE)
```


```{r Dotplot_all_seppMarkers+cilia, fig.height=12}
# from the Sepp paper repeat Dotplot with TF and some key non-TF markers
# added markers of ciliated cells from cluster10 above.
# used the argument col.min = 0 so that average expression scale is from 0 to 2.5 in key

sepp.all.markers <- c("TFAP2A", "PAX3", "RFX4", "TCFL71", "TCF7L2", "GLIS3", "SOX6", "HOPX", "TSC22D4",  
                      "ID4", "NCKAP5", "SLIT2", "CLYBL", "SLC1A3", "SLC6A11", "AQP4",  #astroglial/progenitor
                      
                      "ESRRB", "FOXP1", "FOXP2", "FOXP4", "MEF2C", "RORB", "CDH9", "ETV1", "EBF1", "EBF2", 
                      "SKOR2", "DAB1", "IPTR",   # Purkinje
                      
                      "NOTCH1",    # VZ neuroblast
             
                      "PAX2", "NFIA", "NFIB", "TFAP2B", "BHLHE22", "PRDM8",  # interneuron
                      
                      "ZFHX3", "ZFHX4",   # GABAergic deep nuclei neurons
             
                      "TOP2A", "ATOH1", "LMX1A", "OTX2","ETV6", "PAX6", "BARHL1", "E2F3", "ST18", "SATB2", 
                      "ZBTB18", "NEUROD1", "ZIC1", "ZIC4", "GABAR6", "CNTN1", "CNTN2", "KLF9", "NR3C1", "LMO3", 
                      "SYNPR", "KCNIP4", "HCRTR2", 
                      "EOMES", "TRPC3", "GRM1")
                      # UBC/GCP/granule/glut deep cerebellar nuclei                                                                   
seppAll.cilia.stem.markers <- c(sepp.all.markers, "CCNO", "CFAP299", "ROPN1L", "C11orf88", "ANKRD66", "C20orf85", 
                                "STOML3", "TEKT4", "SNTN", "MEIS1", "POU3F2", "HOXB2", "MEIS2", "GRIA4")


DotPlot(object = filt.noRiboMito.cboDaoy.clusters, features = seppAll.cilia.stem.markers, col.min = 0) +
  coord_flip()
```


---------------------------------------------------------------------------------------------------------
***Cluster the 'phase' seurat object in which the cell cycle difference has been regressed out***
```{r}
# load the object
phase.daoyCbo.sct <- readRDS("outputs50K/DAOY-CBO/phase.filtRiboMito.daoyCbo.SCT.regressedCCdiff.rds")
```

```{r}
# there are now 15 clusters!
phase.daoyCbo.sct <- phase.daoyCbo.sct %>% 
  RunPCA(verbose = FALSE) %>% 
  RunUMAP(dims=1:30, verbose=FALSE) %>% 
  FindNeighbors(dims=1:30, verbose=FALSE) %>% 
  FindClusters(verbose=FALSE)
DimPlot(phase.daoyCbo.sct)
```

```{r}
# find markers of each cluster
daoyCbo.sct.POSdegs <- FindAllMarkers(object = phase.daoyCbo.sct,
                                      min.pct = 0.25,
                                      only.pos = TRUE)
```

```
# save the POSITIVE DEGs of 'daoyCbo.sct.POSdegs'
saveRDS(daoyCbo.sct.POSdegs, file = "outputs50K/DAOY-CBO/POS.DEGs.Daoy.Cbo.phase.sct.rds")
```

```{r}
# get the number of cells per cluster
table(Idents(phase.daoyCbo.sct))
```

```{r}
head(daoyCbo.sct.POSdegs)
```


```{r}
daoyCbo.sct.POSdegs %>% 
  group_by(cluster) %>% 
  slice_head(n = 30)
```

Looking at the above table, there are 15 clusters:
cluster 0 = roof plate/CP
cluster 1 = GABA-ergic cluster
cluster 2 = could be glia
cluster 3 = VZ-derived neurons, PN, 4th ventricle derivatives
cluster 4 = granule lineage
cluster 5 = could be glia
cluster 6 = radial glia and glial markers, uncertain
cluster 7 = DAOY cells
cluster 8 = uncertain, but SOX2-OT is expressed in OPCs and astrocytes
cluster 9 = proliferative cell,may be GCP lineage
cluster 10 = ciliated cells
cluster 11 = glia/radial glia
cluster 12 = uncertain (pct.1 and pct.2 are more similar)
cluster 13 = DAOY cells
cluster 14 = DAOY cells 

***Which clusters are the malignant clusters?***
Comparing the expression of marker genees of each cluster in DAOY-CBO and CBO control there are striking differences which indicate: 1. there are comparatively immature clusters (n=2) of granule cells expressing CNTN2 whereas these cells are not present in CBO control. 2. The number of CNTN1 clusters and NEUROD1 clusters is the same as in CBO control. 3. Cluster 7 expresses IRX3/5, PAX3, MIR124-2HG, POU3F2, MEIS1. None of these factors are expressed to the same level in CBO control. 4. Cluster 13 expresses MIR124-2HG, GRIA4, POU3F2, MEIS1 and LHX9. It also expresses markers of differentiation, e.g. GAP43, DCX, NCAM1 so THIS CLUSTER HAS MIXED CELLS AND SHOULD BE SUB-CLUSTERED/RECLUSTERED (unlike cluster 7 and 14) 5. Cluster 14 expresses TWIST1, PAX3, MEIS1, CHCHD2, and PRRX1 5. Overall, the DAOY-CBO sample has three more clusters than CBO control, and gene expression indicates these are DAOY cell clusters.
```{r VlnPlotGranuleGenes, fig.height=10}
# cluster 4 is the only NeuroD1 cluster, and there are 2 CNTN1 clusters (cluster 1 and 4 as in CBO control)
# malignant clusters are cluster 7 (n=137 cells), cluster 13 (n=77 cells), and cluster 14 (n=55 cells)
VlnPlot(phase.daoyCbo.sct, features = c("NEUROD1", "CNTN1", "CNTN2", "RBFOX3", "RTN1"))
```

```{r VlnPlotDaoyCells, fig.height=10}
# visualise gene expression indicating DAOY cell clusters
VlnPlot(phase.daoyCbo.sct, features = c("IRX3", "GRIA4", "IRX5", "TWIST1", "PAX3", "MIR124-2HG", 
                                        "POU3F2", "MEIS1", "LHX9", "CHCHD2", "PRRX1"))
```

Extract the DAOY cells and save as a separate object

```{r}
daoyOnly.daoyCbo <- subset(phase.daoyCbo.sct, idents = c("7", "13", "14"))
```


```{r}
saveRDS(daoyOnly.daoyCbo, file = "outputs50K/DAOY-CBO/DaoyOnlyCells.phaseDaoyCbo.SCT.CCdiffRegressed.rds")
```

Get the cell IDs of the 'daoyOnly.daoyCbo' object
```{r}
daoyCellID <- WhichCells(phase.daoyCbo.sct, idents = c("7", "13", "14"))
length(daoyCellID)
```

```{r}
# subset the 'filt.noRiboMito.cboDaoy' seurat object for just daoyCellID vector of cancer cells
daoy <- subset(filt.noRiboMito.cboDaoy, cells = daoyCellID)
```

```{r}
dim(daoy)
```

```
# save the 'daoy' object
saveRDS(daoy, file = "./outputs50K/DAOY-CBO/dooyOnlyCells.subset.filt.noRiboMito.cboDaoy.rds")
```


---------------------------------------------------------------------------------------------
***Examine cell cycle and proliferating cells***
```{r}
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
prolif.genes <- list(c(s.genes, g2m.genes))
```

```{r FeaturePlotProlifernDaoyCbo}
phase.daoyCbo.sct <- AddModuleScore(phase.daoyCbo.sct,
                        features = prolif.genes,
                        name = "prolif.genes")
FeaturePlot(phase.daoyCbo.sct, features = 'prolif.genes1', label = TRUE, repel = TRUE
            ) +
   scale_colour_gradientn(colours = rev(brewer.pal(n = 11, name = "RdBu")))

```


Unused code that is retained as it may prove useful
```
# In https://kb.10xgenomics.com/hc/en-us/articles/218169723-What-fraction-of-reads-map-to-ribosomal-proteins- on the 10X Genomics scRNA-seq website there is a table of percent ribosomal genes in different cell types. For neurons the upper limit is 20%. So try filtering 'filt.cboDaoy.seurat' for ribosomal genes and repeat the clustering.

# View metadata data frame, stored in object@meta.data
metadata <- filt.cboDaoy.seurat[[]]

# number of cells with percent.ribo > 20%: 1460 cells
# number of cells with percent.ribo < 20%: 1596 cells

nrow(metadata[metadata$percent.ribo > 20, ])

nrow(metadata[metadata$percent.ribo < 20, ])
```

Code which was not used, may prove useful later
```
# cluster 0 manual = probable glia
cluster0.markers <- cboDaoy.cluster.markers.noRiboMito[cboDaoy.cluster.markers.noRiboMito$cluster == 0, ] %>%
  head(n=10)

cluster0.markers
```

























