---
title: "INTEGRATE_DAOY-CBO_DAOY-SPHER_DAOY-2D_50K.Rmd"
author: "JJ"
date: "04/11/2022"
output: html_document
---


```{r graphical_output}
options(bitmapType='cairo-png') # this line of code needed BEFORE knitr code in chunk below
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = "png",
                      dpi = 300,
                      echo = TRUE,   # prints the code (if FALSE then only code output is shown)
                      cache = TRUE,
                      fig.path = "./outputs50K/INTEGRATE/figures_DAOY/")
```

```{r load_libraries}
# library(presto)
suppressPackageStartupMessages({
library(Seurat)
library(glmGamPoi)
library(Matrix)
library(patchwork)
library(clustree)
library(EnhancedVolcano)
library(tidyverse)
library(reshape2)
library(plyr)
})
```

```{r}
load("RDataFiles/INTEGRATE_DAOY-CBO_DAOY-SPHER_DAOY-2D_50K.Rmd.RData")
```

-------------------------------------------------------------------------------------------------------------
***Merge the DAOY-2D, DAOY-2Drep, DAOY-SPHER, DAOY-CBO, CBO objects***

```{r}
# load the objects
# load the original DAOY-2D dataset which was sequenced to 40-50K read depth
phase.dy2d.orig <- 
  readRDS("../analysis20K/outputs20K/filt.CellsGenesRiboMito.daoy2d.seurat.rds")
# load the repeat DAOY-2D dataset which was sequenced at much greater depth
phase.dy2d.rep <- readRDS("outputs50K/DAOY_2D/filt.CellsGenesRiboMito.daoy2dRep.seurat.rds")
# load the DAOY-SPHER dataset
dy3d <- readRDS("outputs50K/DAOY_SPHER/filt.RiboMitoCellsGenes.daoySpher.rds")
# load the DAOY-CBO and CBO datasets
dyCbo <- readRDS("outputs50K/DAOY-CBO/filtRiboMitoCellsGenes.daoyCbo.rds")
cbo <- readRDS("outputs50K/CBO/filt.RiboMitoCellsGenes.seurat.cbo.rds")
```

```{r}
daoy.big <- merge(phase.dy2d.orig, 
                  y = c(phase.dy2d.rep, dy3d, dyCbo, cbo), 
                  add.cell.ids = c("dy2dorig", "dy2drep", "dy3d", "dyCBO", "CBO"), 
                  project = "dyBig.merge")
daoy.big
```

```{r}
unique(sapply(X = strsplit(colnames(daoy.big), split = "_"), FUN = "[", 1))
```

```{r}
table(daoy.big$orig.ident)
```

```{r}
# Create metadata dataframe
metadata <- daoy.big@meta.data
# Add cell IDs to metadata
metadata$cells <- rownames(metadata)
# Create sample column
metadata$sample <- NA
metadata$sample[which(str_detect(metadata$cells, "^dy2dorig_"))] <- "monolayer_orig"
metadata$sample[which(str_detect(metadata$cells, "^dy2drep_"))] <- "monolayer_rep"
metadata$sample[which(str_detect(metadata$cells, "^dy3d_"))] <- "spheroid"
metadata$sample[which(str_detect(metadata$cells, "^dyCBO_"))] <- "coculture"
metadata$sample[which(str_detect(metadata$cells, "^CBO_"))] <- "organoid"
```

```{r}
head(metadata)
tail(metadata)
```

```{r}
daoy.big@meta.data <- metadata
```

```{r}
dyAll.df <- metadata 
dyAll.df$sample <- factor(dyAll.df$sample, levels = c("monolayer_orig", "spheroid",      "monolayer_rep", "coculture", "organoid"))
```

```{r BarChartCellCountsDaoyMonoSpherCocultOrgMerged, fig.height=6, fig.width=5}
ggplot(dyAll.df, aes(x=sample, fill=sample)) + 
  	geom_bar() +
  	theme_classic() +
   theme(axis.title = element_blank()) +
   NoLegend()
```

```{r}
daoy.big <- NormalizeData(daoy.big)
daoy.big <- FindVariableFeatures(daoy.big)
daoy.big <- ScaleData(daoy.big)
daoy.big <- RunPCA(daoy.big, features = VariableFeatures(object = daoy.big))
```

```{r PCAPlotMergedDaoyCboVsCboVsSpher, fig.height=6}
Idents(daoy.big) <- "sample"
DimPlot(daoy.big, reduction = "pca") +
  theme_gray()
```


```{r}
daoy.big <- RunUMAP(daoy.big, dims = 1:5, reduction = "pca")
```

```{r DimPlotMergedDyCboMonoOrigMonoRepSpherCboCNTRL, fig.height=6}
DimPlot(daoy.big, reduction = "umap", group.by = "sample")
```

```{r VlnPlotGenesDyCboSpherCboCNTRLmerged, fig.height=10, fig.width=12}
# look at some stemness markers
VlnPlot(daoy.big, features = c("POU3F2", "GRIA4", "MEIS1", "PRRX1", "IRX3", "CHCHD2"))
```

-----------------------------------------------------------------------------------------------
***Merge the DAOY-2D repeat sample, the DAOY-SPHER, DAOY-CBO samples and CBO control***
No longer need the original DAOY-2D sample in this merge

```{r}
# load datasets
# repeat DAOY-2D dataset which was sequenced at much greater depth
dy2d.rep <- readRDS("outputs50K/DAOY_2D/filt.CellsGenesRiboMito.daoy2dRep.seurat.rds")
# DAOY-SPHER dataset
dy3d <- readRDS("outputs50K/DAOY_SPHER/filt.RiboMitoCellsGenes.daoySpher.rds")
# DAOY-CBO and CBO datasets
dyCbo <- readRDS("outputs50K/DAOY-CBO/filtRiboMitoCellsGenes.daoyCbo.rds")
cbo <- readRDS("outputs50K/CBO/filt.RiboMitoCellsGenes.seurat.cbo.rds")
```

```{r}
dy.4samples <- merge(dy2d.rep, 
                  y = c(dy3d, dyCbo, cbo), 
                  add.cell.ids = c("dy2drep", "dy3d", "dyCBO", "cbo"), 
                  project = "dy4samples.merge")
dy.4samples
```

```{r}
unique(sapply(X = strsplit(colnames(dy.4samples), split = "_"), FUN = "[", 1))
```

```{r}
table(dy.4samples$orig.ident)
```

```{r}
# Create metadata dataframe
metadata <- dy.4samples@meta.data
# Add cell IDs to metadata
metadata$cells <- rownames(metadata)
# Create sample column
metadata$sample <- NA
metadata$sample[which(str_detect(metadata$cells, "^dy2drep_"))] <- "monolayer"
metadata$sample[which(str_detect(metadata$cells, "^dy3d_"))] <- "spheroid"
metadata$sample[which(str_detect(metadata$cells, "^dyCBO_"))] <- "coculture"
metadata$sample[which(str_detect(metadata$cells, "^cbo_"))] <- "organoid"
```

```{r}
dy.4samples@meta.data <- metadata
```

```{r BarPlotMergedCellCountsDaoyAll}
dy4samples.df <- metadata 
dy4samples.df$sample <- factor(dy4samples.df$sample, 
                               levels = c("spheroid", "monolayer", "coculture", "organoid"))
```

```{r BarChartDaoyCellCounts4conditions, fig.height=7, fig.width=5}
ggplot(dy4samples.df, aes(x=sample, fill=sample)) +
  	geom_bar(fill = "grey") +
  	theme_classic() +
   theme(axis.title = element_blank(), axis.title.y = element_text(size = 18),
         axis.text.x = element_text(size = 14, angle = 45, hjust = 1), 
         axis.text.y = element_text(size = 14)) +
   NoLegend()
```

```{r}
dy.4samples <- NormalizeData(dy.4samples)
dy.4samples <- FindVariableFeatures(dy.4samples)
dy.4samples <- ScaleData(dy.4samples)
dy.4samples <- RunPCA(dy.4samples, features = VariableFeatures(object = dy.4samples))
dy.4samples <- RunUMAP(dy.4samples, dims = 1:5, reduction = "pca")
```

```{r}
saveRDS(dy.4samples,
        file = "./outputs50K/INTEGRATE/Daoy4samplesMerged2dRepSpherCocultCBOpcaUmap.rds")
```

```{r}
dy.4samples <- readRDS("outputs50K/INTEGRATE/Daoy4samplesMerged2dRepSpherCocultCBOpcaUmap.rds")
```


This result makes sense. Spheroid is more closely related to coculture and organoid than monolayer
```{r DimPlotMergedDyCboSpherCboCNTRL, fig.height=6, fig.width=10}
DimPlot(dy.4samples, reduction = "umap", group.by = "sample") +
  labs(title = "")
```

```{r}
# Set Idents to 'sample'
Idents(dy.4samples) <- "sample"
```

The DAOY-CBO and control CBO have the most NRG2, NRG3 and NRG4 ligands
```{r VlnPlotNRGLigandsMergedDaoy2DVs3dVsDaoyCBOvsCbo4samples, fig.height=7, fig.width=10}
VlnPlot(dy.4samples, features = c("NRG1", "NRG2", "NRG3", "NRG4", "CD4"),
        slot = "data",
        assay = "RNA")
```

Here I am plotting the expression of notch ligands in DAOY-2D, DAOY-spher, DAOY-CBO and CBO samples. The increased expression of notch ligands in the DAOY-CBO and CBO samples is consistent with the upregulated expression of Notch readouts, HES5 and HES6 in the extracted DAOY-only coculture cells. This supports the view that increased ligand stimulation of notch signaling in the co-culture condition is what drives notch signalling in the DAOY cells in coculture

```{r VlnPlotNotchLigandsMergedDoyy2dVs3dVsDaoyCboVsCbo, fig.height=7, fig.width=10}
VlnPlot(dy.4samples, features = c("DLL1", "DLL3", "DLL4", "JAG1", "JAG2"),
        slot = "data",
        assay = "RNA")
```


----------------------------------------------------------------------------------------------------
***INTEGRATED analysis of DAOY only cells from DAOY-CBO, DAOY-SPHER cells and DAOY-2D cells using SCTransform***

The problem with the above merge is I cannot tell which are the DAOY cells in the DAOY-CBO sample. So, integrate the extracted DAOY cells from the co-culture with DAOY-SPHER cells - go back to phase object, find the cells that are cancerous, and use those cell IDs to subset the earlier seurat object.
IMPORTANT NOTE - if the objects being integrated are already SCTransformed, DO NOT run SCTransform again on the integrated objects. Best NOT to use already SCTransformed objects when using SCTransform integration workflow

```{r}
dy2d <- readRDS("outputs50K/DAOY_2D/filt.CellsGenesRiboMito.daoy2dRep.seurat.rds")
dySpher <- readRDS("outputs50K/DAOY_SPHER/filt.RiboMitoCellsGenes.daoySpher.rds")
dyOnly.dyCbo <- readRDS("outputs50K/DAOY-CBO/dooyOnlyCells.subset.filt.noRiboMito.cboDaoy.rds")
```

```{r}
dim(dy2d)
dim(dySpher)
dim(dyOnly.dyCbo)
```


```{r}
# No need to re-run this code as the resulting integrated object has been saved

# run SCTransform and then prep integration
dy.3samples.list <- list(dy2d, dySpher, dyOnly.dyCbo)
dy.3samples.list.sct <- suppressWarnings(lapply(X = dy.3samples.list, FUN = SCTransform))
features <- SelectIntegrationFeatures(object.list = dy.3samples.list.sct, nfeatures = 3000)
dy.3samples.list.sct <- PrepSCTIntegration(object.list = dy.3samples.list.sct, anchor.features = features)
```

```{r}
# make the integrated object
dy.3samples.sct.anchors <- FindIntegrationAnchors(object.list = dy.3samples.list.sct, 
                                            normalization.method = "SCT",
                                            anchor.features = features)
dy.3samples.combined.sct <- IntegrateData(anchorset = dy.3samples.sct.anchors, normalization.method = "SCT")
```

```{r}
# cluster the integrated object
dy.3samples.combined.sct <- RunPCA(dy.3samples.combined.sct, verbose = FALSE)
dy.3samples.combined.sct <- RunUMAP(dy.3samples.combined.sct, reduction = "pca", dims = 1:30)
dy.3samples.combined.sct <- FindNeighbors(dy.3samples.combined.sct, 
                                           reduction = "pca", 
                                           dims = 1:30)
dy.3samples.combined.sct <- FindClusters(dy.3samples.combined.sct, resolution = 0.2)
```

```{r}
# DAOY_2Drep = 1935; DAOY-CBO = 269; DAOY-spher = 660 cells
table(dy.3samples.combined.sct$orig.ident)
```

```{r}
# change the orig.ident notation for future ggplots
dy.3samples.combined.sct@meta.data[dy.3samples.combined.sct@meta.data == "DAOY_2Drep"] <- "DAOY_monolayer"
dy.3samples.combined.sct@meta.data[dy.3samples.combined.sct@meta.data == "Daoy-spher"] <- "DAOY_spheroid"
dy.3samples.combined.sct@meta.data[dy.3samples.combined.sct@meta.data == "DAOY-CBO"] <- "DAOY_coculture"
```


```{r}
# save the SCT integrated object, 'dy.3samples.combined.sct'
saveRDS(dy.3samples.combined.sct, 
        file = "outputs50K/INTEGRATE/DaoyOnlyCocult.DaoySpher.Daoy2d.clustered.integrated.SCT.rds")
```

```{r}
# load the above object
dy.3samples.combined.clust.sct <- 
   readRDS("outputs50K/INTEGRATE/DaoyOnlyCocult.DaoySpher.Daoy2d.clustered.integrated.SCT.rds")
```

```{r}
head(dy.3samples.combined.clust.sct@meta.data)
```

```{r}
table(dy.3samples.combined.clust.sct$orig.ident)
```


Nice overlap of DAOY cells grown in different conditions
```{r DimPlotIntegratedDaoyOnlyCocultVsDaoySpherVsDaoy2dRep, fig.height=6, fig.width=10}
DimPlot(dy.3samples.combined.clust.sct, reduction = "umap", group.by = "orig.ident") +
  labs(title = "")
```

The clustering of the different samples shows not very good separation between cells originating from different samples, which is fine, and what is expected after integration
```{r DimPlotIntegratedDaoyOnlyCocultVsDaoySpherVsDaoy2dCLUSTERS}
DimPlot(dy.3samples.combined.clust.sct, reduction = "umap", group.by = "seurat_clusters")
```

```{r}
Idents(dy.3samples.combined.clust.sct) <- "orig.ident"
levels(dy.3samples.combined.clust.sct)
```

```{r}
DefaultAssay(dy.3samples.combined.clust.sct) <- "RNA"
```

Here, use a subset of cells (maximum n=500 for any 1 sample) when running FindAllMarkers() otherwise this function is very slow for the DAOY_monolayer sample. On testing with the full number of cells, vs. downsampled cells there is no difference to the results
```{r}
# no need to rerun this code

daoy.cells.UPmarkers <- FindAllMarkers(object = dy.3samples.combined.clust.sct,
                                     min.pct = 0.25,
                                     logfc.threshold = log(2),
                                     max.cells.per.ident = 500,
                                     only.pos = TRUE
                                     )
```

```
# save the positive markers of each of the three conditions (2D, spher, DAOY-only cocult)
saveRDS(daoy.cells.UPmarkers, 
        file = "outputs50K/INTEGRATE/Pos.DEGs.DaoyCocult.2d.Spher.integrated.sct.rds")
```

```{r}
# load the above data (daoy.cells.UPmarkers)
daoy.cells.UPmarkers <- readRDS("outputs50K/INTEGRATE/Pos.DEGs.DaoyCocult.2d.Spher.integrated.sct.rds")
```


```{r}
# 12,607 UP-regulated markers
nrow(daoy.cells.UPmarkers)
```

```{r}
top20 <- daoy.cells.UPmarkers %>% 
  group_by(cluster) %>% 
  top_n(20, -p_val_adj) %>% 
   dplyr::slice(1:20)
```

```{r}
top20
```


```{r}
# apply ScaleData() before runnning DoHeatmap
dy.3samples.combined.clust.sct <- ScaleData(dy.3samples.combined.clust.sct,
                                         features = as.character(unique(top20$gene)), 
                                         assay = "RNA")
```

Upregulated genes:
COCULTURE:
CTNNA2 - coexpressed with migrating neurons, DCX, Tuj1 (doi: 10.1038/s41588-018-0166-0)
NNAT - expressed in brain, regulates ion channels (DOI: 10.1002/jcp.25498)
RMST - lncRNA
DCX - migrating neurons
NKAIN3 - Chd7 is required for activation of neuronal genes during GNP differentiation, and NKAIN3 is activated by CHD7 (doi: 10.1038/ncomms14758)
TOX3 - expressed in proliferating CNS cells (https://doi.org/10.1016/j.bbagrm.2016.04.005)
LRRN1 - expressed in mouse cerebellum, neuronal protein with an IgCAM domain which might indicate an intervention in neuronal migration (DOI: https://doi.org/10.1677/JME-06-0054)
ADGRB3 - C1ql1 and Bai3 (encoded by ADGRB3) immunoreactivities were mostly colocalized on climbing fibre terminals, not parallel fibres which are granule neuron axons (https://doi.org/10.1016/j.neuron.2014.12.020) and (https://doi.org/10.1038/s41431-018-0321-1). C1ql1 is the ligand for Bai3 receptor, from these mouse studies
MTAP - expressed in EGL (Gensat)
NLGN4X - autism gene, expressed in cerebral cortex
TTR - expressed in VZ, 4th ventricle
DCC - receptor for netrin-1, axon guidance
ELAVL2 - RNA bunding protein, whose co-expression networks are enriched for neurodev and synaptic genes (DOI: 10.1093/hmg/ddw110)
DACH1 - enriched in neuroepithelial and ventricular radial glia cells of the developing human neocortex (doi: 10.1093/cercor/bhy092)
ERBB4 - expressed in Granule neuron precursors and MB stem cells (doi: 10.3390/cancers12040997)

SPHEROID:
COL4A1 - collagen gene
GRP - gastrin-related peptide, secreted neuropeptide
MFAP4 -extracellular matrix protein
CNMD - transmembrane glycoprotein, undergoes cleavage, and that product is secreted into ECM
CFH - interacts with ECM proteins
GAS2 - GAS2 effects are dependent on the ECM (doi: 10.1242/jcs.140558)
```
{r HeatmapMarkersDaoyCocultSpherMono, fig.height=10}
# this heatmap looks odd, the upregulated genes for a single sample are not all grouped together
DoHeatmap(object = dy.3samples.combined.clust.sct, 
          features = as.character(unique(top20$gene)),
          group.by = "ident",
          assay = "RNA") +
  theme(axis.text.y = element_text(size = 12))
```

--------------------------------------------------------------------------------------------------
***Genes which show greater expression in DAOY-only coculture compared to monolayer and spheroid conditions***
Check the expression of various genes including stemness genes in ALL cells in each sample. Below are shown genes which are all increased in coculture condition. ERBB4 was identified in the Volcano Plot as the most significant upregulated gene in DAOY-only cocultured cells VS Daoy-spheroids.

See the Github issue (https://github.com/satijalab/seurat/issues/3334) for guidance on RNA assay and the importance of normalization of the data before VlnPlot, FeaturePlot (but not DoHeatMap())

THE VIOLIN PLOTS BELOW (NOT SHOWN) ARE THE SAME AS THE VIOLIN PLOTS IN THE LATER SECTION OF THIS FILE, SO THERE IS NO NEED TO PLOT THE SAME GENES SHOWING SIMILAR RESULTS TWICE!
```{r VlnPlotStemnessGenesIntegratedDaoy2dVs3dVsdyCocultAllCells, fig.height=20, fig.width=12}
# check the RNA slot is used when plotting integrated data!!! In this case I made the default assay = RNA

VlnPlot(dy.3samples.combined.clust.sct, 
        features = c("ERBB4", "SOX2", "POU3F2","FABP7", "IRX5", "TTYH1", "NRG3",                                                "NNAT","MIR124-2HG", "DCX", "CTNNA2", "MEIS1", "CD4"),
       slot = 'data',
       assay = 'RNA')
```


------------------------------------------------------------------------------------------------------
***Neuregulin signalling differs in the cells grown under monolayer, spheroid and co-culture conditions***
NRG1, 2, 4 expression are not elevated in the DAOY cells that have been cocultured relative to 2D and spheroid cultures. NRG 3 is higher in cocultured DAOY cells
```{r VlnPlotNeuregulinsIntegratedDaoyOnlyCocultVs3dVs2dAllCells, fig.height=10, fig.width=12}
VlnPlot(dy.3samples.combined.clust.sct, features = c("NRG1", "NRG2", "NRG3", "NRG4"))
```

Below are other interesting genes, but these do not show increased expressed in the DAOY-only coculture. There is not much evidence of increased SHH signalling. SMAD signalling is overall less than in ONS76 cells, with the possible exception of SMAD3 in DAOY-monolayer cells. Hypothesis is that the different level of SMAD signalling in ONS76 cells vs DAOY cells is responsible for the higher expression of SOX2 in ONS76 monolayer and spheroid cells. The targets of SMAD signalling, SERPINE1 and ID1 are expressed overall at lower level in DAOY cells vs ONS76 cells. Note however the expression of these targets is context-dependent (https://doi.org/10.1038/onc.2012.191)
```{r fig.height=4}
# check the RNA slot is used when plotting integrated data!!! In this case I made the default assay = RNA
dy.3samples.combined.clust.sct <- ScaleData(dy.3samples.combined.clust.sct)
Idents(dy.3samples.combined.clust.sct) <- "orig.ident"
```

```{r VlnPlotShhHoxB2GenesIntegratedDy2dVs3dVsDycocultAllCells, fig.height=7, fig.width=10}
VlnPlot(dy.3samples.combined.clust.sct, 
        features = c("HOXB2","IRX3", "TWIST1",
                    "PTCH1", "GLI1", "GLI2"),
        slot = "data",
        assay = "RNA")
```



```{r VlnPlotShhMycOtx2Zic1IntegratedDy2dVs3dVsDycocultAllCells, fig.height=7, fig.width=10}
VlnPlot(dy.3samples.combined.clust.sct, 
        features = c("GLI3", "HHIP", "MYC", "MYCN", "ZIC1", 
                     "OTX2"),
        slot = "data",
        assay = "RNA")
```

```{r VlnPlotMapkSmadIntegratedDy2dVs3dVsDycocultAllCells, fig.height=7, fig.width=10}
VlnPlot(dy.3samples.combined.clust.sct, 
        features = c("MAPK3","PROX1", "SMAD1", "SMAD2",
                      "SMAD3", "SMAD4"),
        slot = "data",
        assay = "RNA")
```

```{r VlnPlotNESId1SmadGenesIntegratedDy2dVs3dVsDycocultAllCells, fig.height=7, fig.width=10}
VlnPlot(dy.3samples.combined.clust.sct, 
        features = c("NES", "SMAD6", "SMAD7", "SERPINE1", "ID1"),
        slot = "data",
        assay = "RNA")
```


------------------------------------------------------------------------------------------------------
***Notch-delta signalling in DAOY-2D, DAOY-spheroid, DAOY-only coculture, DAOY-CBO (all cells) and CBO***
Check whether SOX2 upregulation in DAOY-only coculture is also due to increased notch signalling. It looks like there is greater HES5 and HES6 expression in coculture, suggesting increased notch signalling in this condition. Notch ligands are also shown but these genes are more relevant in the healthy CBO cells with which the DAOY cells are co-cultured. The take home message of notch ligand expression from 'dy.3samples.combined.clust.sct'  is that apart from DLL1, notch ligands are expressed at lower level in the DAOY cells grown in coculture
```{r VlnPlotIntegratedNotchDeltaDaoyOnlyCocultVsSpherVs2dAllCells, fig.height=7, fig.width=10}
# notch signalling readout genes and notch ligands
VlnPlot(dy.3samples.combined.clust.sct, features = c("HES5", "HES6",
                                        "DEC1", "DLL1"),
        slot = "data",
        assay = "RNA")
```

```{r VlnPlotIntegratedOtherNotchDeltaDaoyOnlyCocultVsSpherVs2dAllCells, fig.height=7, fig.width=10}
# notch signalling readout genes and notch ligands
VlnPlot(dy.3samples.combined.clust.sct, features = c("DLL3", "DLL4", "JAG1", "JAG2"),
        slot = "data",
        assay = "RNA")
```


--------------------------------------------------------------------------------------------------------
***Downsampling of DAOY-2D and DAOY-spher cells to generate Heatmap of DEGs in integrated DAOY-2D, spheroid and DAOY-only coculture samples***
The above heatmap has used all the cells in each sample. To get around the odd appearance of the heatmap, downsample the cells from each of the 'Idents' in the integrated seurat object so that DoHeatmap only shows n = 269 cells for each sample. 269 cells is the number of DAOY cells in the co-culture.
```{r}
# get the cell IDs of all 269 daoyOnly.daoycbo cells
dyIDs.cocult <- WhichCells(dy.3samples.combined.clust.sct, idents = "DAOY_coculture")
length(dyIDs.cocult)
```

```{r}
# get a random subset of DAOY_spheroid cells, with n=269
dyIDs.spher <- WhichCells(dy.3samples.combined.clust.sct, idents = "DAOY_spheroid")
length(dyIDs.spher)
set.seed(111)
dyIDs269.spher <- sample(dyIDs.spher, size = 269, replace = FALSE)
length(dyIDs269.spher)
```

```{r}
# get a random subset of DAOY_monolayer cells, with n=269
dyIDs.mono <- WhichCells(dy.3samples.combined.clust.sct, idents = "DAOY_monolayer")
length(dyIDs.mono)
set.seed(123)
dyIDs269.mono <- sample(dyIDs.mono, size = 269, replace = FALSE)
length(dyIDs269.mono)
```

Only use the below cell IDs as input cells for FindAllMarkers()
```{r}
# make a vector of cell identities to use
cellIDs.3samples <- c(dyIDs269.mono, dyIDs269.spher, dyIDs.cocult)
length(cellIDs.3samples)
```

```{r}
# subset 'dy.3samples.combined.clust.sct' to get only the 807 (269 x 3 = 807) cells
dy.3samples.small <- subset(dy.3samples.combined.clust.sct, cells = cellIDs.3samples)
dim(dy.3samples.small)
class(dy.3samples.small)
```

```{r}
# No need to rerun this code

# run FindAllMarkers on the smaller object - much quicker to run!
daoy.3samplesSmall.UPmarkers <- FindAllMarkers(object = dy.3samples.small,
                                     min.pct = 0.25,
                                     logfc.threshold = log(2),
                                     only.pos = TRUE
                                     )
# 12,385 UP-regulated markers
```

```{r}
# save these positive markers
saveRDS(daoy.3samplesSmall.UPmarkers, file = "outputs50K/INTEGRATE/Pos.DEGs.DaoyCocult.2d.Spher.downsampled269cells.integrated.rds")
```

```{r}
daoy.3samplesSmall.UPmarkers <- 
  readRDS("outputs50K/INTEGRATE/Pos.DEGs.DaoyCocult.2d.Spher.downsampled269cells.integrated.rds")
```


```{r}
top20.small <- daoy.3samplesSmall.UPmarkers %>% 
  group_by(cluster) %>% 
  top_n(20, -p_val_adj) %>% 
   dplyr::slice(1:20)
```

```{r}
top20.small %>% 
  head(10)
```

```{r}
dy.3samples.small <- ScaleData(dy.3samples.small, assay = "RNA")
```

Plot a heatmap of genes of interest that are differentially expressed between DAOY monolayer, spheroid and DAOY-only coculture
```{r HeatmapMarkersDaoyCocultSpherMonoDownSampled, fig.height=13, fig.width=14}
DoHeatmap(object = dy.3samples.small, 
          features = as.character(unique(top20.small$gene)),
          group.by = "ident",
          assay = "RNA") +
   guides(color = "none") +
   theme(axis.text = element_text(size = 14))
 
```

-------------------------------------------------------------------------------------------------------
***Violin plots of stemness markers, ERBB4, neuregulins, notch-delta and other genes expressed in the DAOY cells in coculture, DAOY-SPHER and DAOY-2D condition in DOWNSAMPLED cells***
These Violin plots are consistent with the violin plots using ALL cells above.
Here we are taking a few genes expressed in the DAOY-CBO sample and comparing their expression in DAOY cells grown under different conditions (mono, spher, coculture)
Note SOX2 positively regulates POU3F2 and FABP7 expression, SOX2 represses NEUROD1 transcription. The plots below are interesting! All the below genes, incl. stemness genes are only upregulated in the coculture condition! 
UNITS ARE LOG-NORMED VERSION OF COUNTS.
NNAT is one of the genes known to upregulated in MB.


```{r VlnPlotStemnessMarkerGenesIntegratedDaoyOnlyCocultVsSpherVsMonoDownSampled, fig.height=7, fig.width=10}
# make violin plots of key genes in the cancer cells grown under three conditions. The 'dy.3samples.small'
# seurat object should be normalised after loading, and then only generate the Violin plot!
# The y-axis is in log values

VlnPlot(dy.3samples.small, 
        features = c("SOX2", "POU3F2","FABP7", "GRIA4", "MYT1L"),
        slot = "data",
        assay = "RNA")
```

```{r VlnPlotMeis1IrxAndOthersIntegratedDaoyOnlyCocultVsSpherVsMonoDownSampled, fig.height=7, fig.width=10}
VlnPlot(dy.3samples.small, 
        features = c("MEIS1", "DCX", "IRX3", "IRX5", "TTYH1"),
        slot = "data",
        assay = "RNA")
```

```{r VlnPlotZIC1andOthersIntegratedDaoyOnlyCocultVsSpherVsMonoDownSampled, fig.height=7, fig.width=10}
VlnPlot(dy.3samples.small, 
        features = c("NNAT", "MIR124-2HG", "ZIC1", "CTNNA2"),
        slot = "data",
        assay = "RNA")
```

NEUREGULINS:
High level of ERBB4 suggests neuregulin signalling is active in the cocultures - checking for NRG family member expression in the CBO shows NRG3 is strongly expressed in the DAOY only cells in coculture, consistent with the ViolinPlot above. NRG1 and NRG2 are expressed most in the DAOY-2D cells. NRG2 activates SOX2 expression in rodent adult neurogenesis (doi: 10.1073/pnas.0510410103). These violin plots are consistent with the violin plots using ALL cells.
```{r VlnPlotNeuregulinsDaoyOnlyCocultVsSpherVsMono, fig.height=7, fig.width=10}
VlnPlot(dy.3samples.small, 
        features = c("ERBB4", "ADAM17", "EGFR", "NRG1", "NRG2"),
        slot = "data",
        assay = "RNA")
```

```{r VlnPlotNeuregulinPtchGli2DaoyOnlyCocultVsSpherVsMono, fig.height=7, fig.width=10}
VlnPlot(dy.3samples.small, 
        features = c("NRG3", "NRG4", "PTCH1", "PTCH2", "GLI2"),
        slot = "data",
        assay = "RNA")
```

SOX2 and notch-delta signalling:
Sox2 is activated by Notch signalling and this relationship is well-recognised in multiple different cellular/tissue contexts (e.g.  https://doi.org/10.18632/oncotarget.10954 , doi: 10.1016/j.cellsig.2020.109537 , DOI: https://doi.org/10.1523/JNEUROSCI.3150-12.2013 , https://doi.org/10.1073/pnas.100308910). Readouts of Notch signalling are shown and HES5/6 are strongly expressed in the DAOY-coculture cells, while the Notch ligand DLL1 is expressed in the DAOY-only cells in coculture and DLL3 and JAG1 are expressed in the CBO control (same as Violin plots above).
```{r VlnPlotNotchDeltaIntegratedDaoyOnlyCocultVsSpherVsMonoDownsampled, fig.height=7, fig.width=10}
# notch signalling readout genes and notch ligands
VlnPlot(dy.3samples.small, features = c("HES5", "HES6", "GFAP", "FABP7", "OLIG1", "OLIG2"), 
        slot = "data",
        assay = "RNA")
```

Examine "stemness" markers in the different conditions. Only FABP7 is expressed
```{r VlnPlotFabp7PdgfraGenesIntegratedDaoyOnlyCocultVsSpherVsMonoDownsampled, fig.height=7, fig.width=10}
VlnPlot(dy.3samples.small, features = c("OLIG1", "OLIG2", "FABP7", "PDGFRA"), 
        slot = "data",
        assay = "RNA")
```


------------------------------------------------------------------------------------------
***Find genes differentially expressed between spheroid and monolayer condition***

```{r}
# find DEGs and plot in a volcano plot
# just use the two datasets merged
dy.3d2d.merge <- merge(dy2d.rep, 
                  y = c(dy3d), 
                  add.cell.ids = c("dy2drep", "dy3d"), 
                  project = "dy2samples.merge")
dy.3d2d.merge
```

```{r}
# Create metadata dataframe
metadata <- dy.3d2d.merge@meta.data
# Add cell IDs to metadata
metadata$cells <- rownames(metadata)
# Create sample column
metadata$sample <- NA
metadata$sample[which(str_detect(metadata$cells, "^dy2drep_"))] <- "monolayer"
metadata$sample[which(str_detect(metadata$cells, "^dy3d_"))] <- "spheroid"
```

```{r}
dy.3d2d.merge@meta.data <- metadata
```

```{r}
dy.3d2d.merge <- NormalizeData(dy.3d2d.merge)
dy.3d2d.merge <- FindVariableFeatures(dy.3d2d.merge)
dy.3d2d.merge <- ScaleData(dy.3d2d.merge)
dy.3d2d.merge <- RunPCA(dy.3d2d.merge, features = VariableFeatures(object = dy.3d2d.merge))
dy.3d2d.merge <- RunUMAP(dy.3d2d.merge, dims = 1:5, reduction = "pca")
```

```{r}
Idents(dy.3d2d.merge) <- "sample"
dy.3d2d.merge.markers <- FindMarkers(dy.3d2d.merge, ident.1 = "spheroid", ident.2 = "monolayer",
                               verbose = FALSE,
                               min.pct = 0.5,
                               logfc.threshold = 0.25)
```

```{r}
dy.3d2d.merge.markers.gene <- dy.3d2d.merge.markers %>% 
  rownames_to_column(var = "gene")
```

```{r}
dy.3d2d.merge.markers.gene %>% 
  arrange(desc(avg_log2FC)) %>% 
  tail(20)
```

```{r}
# save the DEGs which are both UP and DOWN regulated

write.csv(dy.3d2d.merge.markers.gene, 
          file = "./outputs50K/INTEGRATE/DEGsAllDaoySpherVSMonoMergedDoay2drepDaoySpher.csv", 
          quote = FALSE)
```

```{r}
# load the DEGs above

dy.3d2d.merge.markers.gene <- 
  read.csv("outputs50K/INTEGRATE/DEGsAllDaoySpherVSMonoMergedDoay2drepDaoySpher.csv",
           header = TRUE,
           stringsAsFactors = FALSE)
```


DOWNREGULATED
SERPINE1 - Downregulation of PAI-1 significantly reduced cellular proliferation through an inability to progress from the G(0-G1) phase of the cell cycle. Accordingly, overexpression of PAI-1 augmented proliferation by encouraging S-phase entry (DOI: 10.1158/1541-7786.MCR-13-0543)
MT2A - expression of the proliferation marker Ki67 was inversely correlated with MT3 at relapse
PRKCA - PRKCA kinase domain induces phosphorylation of RAF-1, which in turn activates the extracellular signal-regulated kinase/mitogen-activated protein kinase (ERK/MAPK) cascade, a key signaling pathway in melanocyte growth and survival.13 Furthermore, PRKCA fusions have been identified as putative pathogenic events in other tumors, such as papillary glioneuronal tumor and lung squamous cell carcinoma.  
EDIL3 -identification of 17 genes associated with enhanced medulloblastoma formation. We show that these genes are enriched for neuronal transcription factors defining a novel gene network which, when mutagenised by SB, is associated with increased cell proliferation and reduced neuronal differentiation (doi: 10.1186/2051-5960-1-35)
TENM3 - Evidence of its deregulation in different types of cancer has suggested that TENM3 may have a possible functional role in tumorigenesis. The in silico analysis of RNA sequencing data from the TCGA and GTEx projects demonstrates that TENM3 is significantly more highly expressed, as compared to normal tissues, in head and neck squamous cell carcinoma, in pancreatic adenocarcinoma, in thymoma (Figure 4), and in neuroblastoma. High TENM3 expression is observed in the human neuroblastoma cell line SH-SY5Y, in which TENM3 is one of the most abundant membrane proteins (doi: 10.3390/ijms22052321)

UPREGULATED in Volcano Plot - many ECM proteins
For example:
MDK - has been linked to EMT (https://doi.org/10.1038/s41388-019-1124-8)
COL14A1 -  is a collagen gene
COL1A2 -  is a collagen gene
CRIP1 - its upregulation is associated with REDUCED proliferation in breast cancer (doi: 10.1186/1476-4598-12-28)
SFRP2 - ECM remodelling (doi: 10.1152/ajpcell.00137.2016)
GPC3 - glypican
PCOLCE - This gene encodes a glycoprotein which binds and drives the enzymatic cleavage of type I procollagen and heightens C-proteinase activity
IFI6 - associated with ECM remodelling in oesophageal Cancer (https://doi.org/10.1007/s00262-022-03288-0)
VCAN - is an ECM proteoglycan (https://doi.org/10.3389/fimmu.2020.00512)
PTN - pleiotrophin is a secreted cell signaling cytokine that acts as growth factor associated with the extracellular matrix, which has recently started to come to the fore as a significant neuromodulator with multiple neuronal functions (https://doi.org/10.3389/fncel.2014.00443)

```{r VolcanoPlotDAOYSpherVSMono, fig.height=10, fig.width=15}
genes.to.label.daoySpherVsMono <- c("MDK", "COL14A1", "SFRP2", "IFI6", "VCAN", "PTN", "COL1A2",
                                    "CRIP1", "GPC3", "PCOLCE", "SERPINE1", "MT2A", "PKRCA",
                                    "EDIL3", "NEGR1", "TENM3")

EnhancedVolcano(dy.3d2d.merge.markers.gene,
                lab = dy.3d2d.merge.markers.gene$gene,
                x = "avg_log2FC",
                y = "p_val_adj",
                selectLab = genes.to.label.daoySpherVsMono,
                title = NULL,
                subtitle = NULL,
                pointSize = 5.0,
                labSize = 6.0,
                cutoffLineWidth = 1.0,
                axisLabSize = 30,
                legendLabSize = 25,
                legendIconSize = 10,
                legendPosition = "right",
                legendLabels = c("NS", "Log2FC", "p-value", "p-value and Log2FC"),
                drawConnectors = TRUE,
                widthConnectors = 0.5,
                gridlines.major = FALSE,
                gridlines.minor = FALSE)
```

-------------------------------------------------------------------------------------------------------
***Find Genes differentially upregulated in DAOY-only cocultures vs DAOY-spher***
Some of these genes are plotted in the violin plots above and can be found by inspecting genes recovered from the FindAllMarkers() function.
A volcano plot is shown here:
```{r}
# find DEGs and plot in a volcano plot
# use the two datasets merged
dy.cocult.3d.merge <- merge(dy3d, 
                  y = c(dyOnly.dyCbo), 
                  add.cell.ids = c("dy3d", "dyOnlyCbo"), 
                  project = "dy3dcocult.merge")
dy.cocult.3d.merge
```

```{r}
# Create metadata dataframe
metadata <- dy.cocult.3d.merge@meta.data
# Add cell IDs to metadata
metadata$cells <- rownames(metadata)
# Create sample column
metadata$sample <- NA
metadata$sample[which(str_detect(metadata$cells, "^dy3d_"))] <- "spheroid"
metadata$sample[which(str_detect(metadata$cells, "^dyOnlyCbo_"))] <- "coculture"
```

```{r}
dy.cocult.3d.merge@meta.data <- metadata
```

```{r}
dy.cocult.3d.merge <- NormalizeData(dy.cocult.3d.merge)
dy.cocult.3d.merge <- FindVariableFeatures(dy.cocult.3d.merge)
dy.cocult.3d.merge <- ScaleData(dy.cocult.3d.merge)
dy.cocult.3d.merge <- RunPCA(dy.cocult.3d.merge, features = VariableFeatures(object = dy.cocult.3d.merge))
dy.cocult.3d.merge <- RunUMAP(dy.cocult.3d.merge, dims = 1:5, reduction = "pca")
```

```{r}
saveRDS(dy.cocult.3d.merge,
        file = "./outputs50K/INTEGRATE/Daoy2samplesMergedSpherCocultPCAandUMAP.rds")
```

```{r}
dy.cocult.3d.merge <- readRDS("outputs50K/INTEGRATE/Daoy2samplesMergedSpherCocultPCAandUMAP.rds")
```


This shows the DAOY cells upregulate SOX2 expression (shown in 4th PC) when grown in the coculture condition versus spheroid condition
```{r VizDimLoadingsDAOYonlyCocultVsSpher, fig.height=5, fig.width=4}
VizDimLoadings(dy.cocult.3d.merge, dims = 4, reduction = "pca", balanced = TRUE, nfeatures = 20)
```

```{r}
Idents(dy.cocult.3d.merge) <- "sample"
dy.cocult.3d.merge.markers <- FindMarkers(dy.cocult.3d.merge, ident.1 = "coculture", 
                                          ident.2 = "spheroid",
                                          verbose = FALSE,
                                          min.pct = 0.5,
                                          logfc.threshold = 0.25)
```

```{r}
dy.cocult.3d.merge.markers.gene <- dy.cocult.3d.merge.markers %>% 
  rownames_to_column(var = "gene")
```

```{r}
dy.cocult.3d.merge.markers.gene %>% 
  arrange(desc(avg_log2FC)) %>% 
  head(10)
```

UPREGULATED in Volcano Plot:
ERBB4 - top gene, stands out as strongly upregulated
NNAT - NNAT promotes MB growth (DOI:10.1215/15228517-2008-038
CRABP1 - high CRABP1 expression levels are associated with poor patient prognosis, high tumor grade in breast cancer, MB modelling in cerebellar organoids (doi: 10.1038/s41467-019-13989-3 - Ballabio et al, Nat Comm). Overexpressed in Gp 3 and 4 MB (DOI: 10.1371/journal.pone.0112909)
NPAS3 - We found that the internodular MBEN (exclusively SHH-MB) compartment comprised proliferating early cerebellar granular neuronal precursors (CGNP)-like tumor cells as well as stromal, vascular, and immune cells. In contrast, the nodular compartment consisted of postmitotic, neuronally differentiated MBEN cells..... Astrocyte marker genes, such as NPAS3, CD44 and LAMA2, were also expressed in a fraction of early CGNP-like cells, which were identified as potential precursors of tumor astrocytes in our snRNA-seq trajectory analysis (https://doi.org/10.1101/2022.09.02.506321).
NR2F1 - enriched in Gp 3 MB (doi: https://doi.org/10.1101/2022.08.17.504304)
TUBB3 - neuronal differentiation marker
GPHN - upregulated in infant SHH-MB (DOI 10.1007/s00401-011-0846-7)
MEIS1 - upregulated in infant SHH-MB (DOI 10.1007/s00401-011-0846-7)
FEZ1 - upregulated in infant SHH-MB (DOI 10.1007/s00401-011-0846-7)
PRDM1 - upregulated in infant SHH-MB (DOI 10.1007/s00401-011-0846-7)

```{r VolcanoPlotDaoyOnlyCocultVsSpher, fig.height=10, fig.width=15}

genes.to.label.daoyCocultVsSpher <- c("ERBB4", "NNAT", "CRABP1", "NPAS3", "TUBB3", "NR2F1", "GPHN")

EnhancedVolcano(dy.cocult.3d.merge.markers.gene,
                lab = dy.cocult.3d.merge.markers.gene$gene,
                selectLab = genes.to.label.daoyCocultVsSpher,
                x = "avg_log2FC",
                y = "p_val_adj",
                title = NULL,
                subtitle = NULL,
                pointSize = 5.0,
                labSize = 6.0,
                cutoffLineWidth = 1.0,
                axisLabSize = 30,
                legendLabSize = 25,
                legendIconSize = 10,
                legendPosition = "right",
                legendLabels = c("NS", "Log2FC", "p-value", "p-value and Log2FC"),
                drawConnectors = TRUE,
                widthConnectors = 0.5,
                gridlines.major = FALSE,
                gridlines.minor = FALSE)
```

----------------------------------------------------------------------------------------------------
***Merge DAOY-2D and DAOY-spheroid samples, then use presto to find enriched genesets using g:Profiler (web app)***
I did not run fgsea as it does not work well with merged seurat objects (did not work with the equivalent ONS76 samples)
Merge the following: 
1) DAOY-2D single cluster 
2) DAOY-spher cluster 0
3) DAOY-spher cluster 1 
The files used are:
"phase.daoySpher.sct.clust.res0.2" and "phase.filtRiboMito.SCT.regressedCCdiff.daoy.2d"
Find the cell IDs of the cluster 0,1,2 cells in "phase.daoySpher.sct.clust.res0.2" and extract those cells
```{r}
phase.daoySpher.sct.clust.res0.2 <- 
readRDS("outputs50K/DAOY_SPHER/phase.daoySpher.SCT.clustered.res0.2.rds")
```

```{r}
# make Idents the 'seurat_clusters
Idents(phase.daoySpher.sct.clust.res0.2) <- "seurat_clusters"
```

```{r}
# extract the cell clusters
cl0cellID.daoySpher <- WhichCells(phase.daoySpher.sct.clust.res0.2, idents = "0")
cl1cellID.daoySpher <- WhichCells(phase.daoySpher.sct.clust.res0.2, idents = "1")

length(cl0cellID.daoySpher)
length(cl1cellID.daoySpher)

```

For DAOY-spher: use the vector of cell IDs to subset the object "phase.daoySpher.sct.clust.res0.2" and pull out clusters 0 - 1 as separate clusters
```{r}
# get ONS76 cells in cluster 0 of ONS76-spher using cell IDs
clus0.daoySpher <- subset(phase.daoySpher.sct.clust.res0.2, cells = cl0cellID.daoySpher)
clus1.daoySpher <- subset(phase.daoySpher.sct.clust.res0.2, cells = cl1cellID.daoySpher)
```

```{r}
# load the DAOY-2D sample (this is one big cluster of cells)
daoyMono.sct <- readRDS("outputs50K/DAOY_2D/phase.filtRiboMito.SCT.regressedCCdiff.daoy.2d.rds")
```

```{r}
# merge the above datasets
daoy.mono.spher.merged <- merge(daoyMono.sct,
                                 y = c(clus0.daoySpher, clus1.daoySpher),
                                 add.cell.ids = c("dyMono", "clus0.dySpher", 
                                                  "clus1.dySpher"),
                                 project = "daoy.mono.spher.merge")
```

```{r}
head(daoy.mono.spher.merged@meta.data)
```

```{r}
# assign cells to "seurat_clusters"
metadata <- daoy.mono.spher.merged@meta.data
metadata$cells <- rownames(metadata)
metadata$seurat_clusters[which(str_detect(metadata$cells, "^dyMono_"))] <- "mono"
metadata$seurat_clusters[which(str_detect(metadata$cells, "^clus0.dySpher_"))] <- "0"
metadata$seurat_clusters[which(str_detect(metadata$cells, "^clus1.dySpher_"))] <- "1"

head(metadata)
```

```{r}
daoy.mono.spher.merged@meta.data <- metadata
```

```{r}
daoy.mono.spher.genes <- wilcoxauc(daoy.mono.spher.merged, "seurat_clusters")
```

```{r}
head(daoy.mono.spher.genes)
```

```{r}
nrow(daoy.mono.spher.genes)
```

```{r}
# we have all the genes for each cluster
dplyr::count(daoy.mono.spher.genes, group)
```

```{r}
# order the genes for input to gProfiler (web app!)
daoy.mono.UPgenes <- daoy.mono.spher.genes %>% 
  dplyr::filter(group == "mono" & logFC > 0.25) %>%
  dplyr::arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)

nrow(daoy.mono.UPgenes) # I used 2129 genes in g:Profiler
```

```{r}
# this will be used as input for UPregulated genes in g:Profiler. Transfer file via Filezilla to
# Excel on Mac
write.csv(daoy.mono.UPgenes, 
          "./outputs50K/INTEGRATE/DAOY.monoSpher.merged.monoUPgenes.PRESTO.gProfilerInput.csv", 
          quote = F)
```

```{r}
# order the genes for input to gProfiler (web app!)
daoy.spher.clus0.UPgenes <- daoy.mono.spher.genes %>% 
  dplyr::filter(group == "0" & logFC > 0.25) %>%
  dplyr::arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)

nrow(daoy.spher.clus0.UPgenes) # I used 618 genes in g:Profiler
```

```{r}
# this will be used as input for UPregulated genes in g:Profiler. Transfer file via Filezilla to
# Excel on Mac
write.csv(daoy.spher.clus0.UPgenes, 
          "./outputs50K/INTEGRATE/DAOY.monoSpher.merged.cluster0spherUPgenes.PRESTO.gProfilerInput.csv", 
          quote = F)
```

```{r}
# order the genes for input to gProfiler (web app!)
daoy.spher.clus1.UPgenes <- daoy.mono.spher.genes %>% 
  dplyr::filter(group == "1" & logFC > 0.25) %>%
  dplyr::arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)

nrow(daoy.spher.clus1.UPgenes) # I used 346 genes in g:Profiler
```

```{r}
# this will be used as input for UPregulated genes in g:Profiler. Transfer file via Filezilla to
# Excel on Mac
write.csv(daoy.spher.clus1.UPgenes, 
          "./outputs50K/INTEGRATE/DAOY.monoSpher.merged.cluster1spherUPgenes.PRESTO.gProfilerInput.csv", 
          quote = F)
```



----------------------------------------------------------------------------------------------------
CODE RETAINED IN CASE IT PROVES USEFUL LATER
***Get AverageExpression of genes in the downsampled integrated object with 269 cells in each sample*** 
The samples are DaoyOnly-coculture, Daoy2D(rep) and DaoySpher

```
{r}
# save the downsampled seurat object with 809 cells (269 cells/sample)
saveRDS(dy.3samples.small, 
        file = "outputs50K/INTEGRATE/daoyOnlyCocult.daoy2D.daoySpher.downsampled.integrated.SCT.rds")
```

```
{r}
# load the above seurat object
dy.3samples.small <- readRDS("outputs50K/INTEGRATE/daoyOnlyCocult.daoy2D.daoySpher.downsampled.integrated.SCT.rds")

# data must be re-normalized after subsetting or AverageExpression will not work
dy.3samples.small <- NormalizeData(dy.3samples.small) 
```

```
{r}
class(dy.3samples.small)
```


```
{r}
table(dy.3samples.small$orig.ident)
```


```
{r}
# set the DefaultAssay to RNA and subset the integrated object
DefaultAssay(dy.3samples.small) <- "RNA"
dy2d <- subset(dy.3samples.small, subset = (orig.ident == "DAOY_monolayer"))
dy3d <- subset(dy.3samples.small, subset = (orig.ident == "DAOY_spheroid"))
dyCocult <- subset(dy.3samples.small, subset = (orig.ident == "DAOY_coculture"))
```

```
{r}
av.dy2d <- AverageExpression(dy2d)$RNA
av.dy3d <- AverageExpression(dy3d)$RNA
av.dyCocult <- AverageExpression(dyCocult)$RNA
```

```
{r}
# make the column names of the matrix objects meaningful
colnames(av.dy2d) <- "DAOY_monolayer"
colnames(av.dy3d) <- "DAOY_spheroid"
colnames(av.dyCocult) <- "DAOY_coculture"
```

```
{r}
# convert the matrices of Average Expression to dataframes
# df = dataframe
av.dy2d.df <- as.data.frame(av.dy2d) %>% 
  rownames_to_column("genes")
av.dy3d.df <- as.data.frame(av.dy3d) %>% 
  rownames_to_column("genes")
av.dyCocult.df <- as.data.frame(av.dyCocult) %>% 
  rownames_to_column("genes")
```

```
{r}
# left join all three dataframes in one go
dy.3dfs <- join_all(list(av.dyCocult.df, av.dy3d.df, av.dy2d.df), 
                       by = "genes", type = "left")
```

```
{r}
# identify any NA values - there are none
colSums(is.na(dy.3dfs))
```

```
# save the combined dataframe of Average Expression in subsets of the 
# integrated object (dy.3dfs)
saveRDS(dy.3dfs, file = "outputs50K/INTEGRATE/AvgExpressn.daoy.2d.3d.Cocult.df.downsampled.SCT.rds")
```

```
{r}
# load the above integrated seurat object
dy.3dfs <- readRDS("outputs50K/INTEGRATE/AvgExpressn.daoy.2d.3d.Cocult.df.downsampled.SCT.rds")
```


```
{r}
dy.3dfs[dy.3dfs$genes %in% c("ERBB4", "MEIS1", "IRX3", "IRX5", "SOX2", "POU3F2","FABP7",
                                        "MIR124-2HG", "PTCH1", "GLI2", "DCX", "CTNNA2", "ZIC1"), ]

```

```
{r}
genes.ord <- dy.3dfs[dy.3dfs$genes %in% c("ERBB4", "MEIS1", "IRX3", "IRX5", "SOX2", "POU3F2","FABP7",
                                        "MIR124-2HG", "PTCH1", "GLI2", "DCX", "CTNNA2", "ZIC1"), ]
genes.ord %>% 
   arrange(DAOY_coculture)
```

```
{r}
# plot the changes in gene expression for different samples.
# change the format of the data to long format for ggplot

genes.to.plot <- c("ERBB4", "MEIS1", "IRX3", "IRX5", "SOX2", "POU3F2","FABP7",
                                        "MIR124-2HG", "PTCH1", "GLI2", "DCX", "CTNNA2", "ZIC1")
diff.stem.genes <- dy.3dfs[dy.3dfs$genes %in% genes.to.plot, ]
diff.stem.long.df <- diff.stem.genes %>%
  pivot_longer(cols = -genes,
               names_to = "sample",
               values_to = "expression")
diff.stem.long.df
``` 

```
{r}
diff.stem.long.df$genes <- as.factor(diff.stem.long.df$genes)
levels(diff.stem.long.df$genes)
```

```
{r}
# reorder the legend by specifying the correct order of the genes
diff.stem.long.df$genes <- factor(diff.stem.long.df$genes, 
                                  levels = c("ERBB4", "CTNNA2", "DCX", "SOX2", "POU3F2", "MEIS1",
                                             "ZIC1", "MIR124-2HG", "FABP7", "IRX5", "GLI2",
                                             "IRX3", "PTCH1"))
levels(diff.stem.long.df$genes)
```

```
{r LineplotGenesOfInterestAvgExpressnDaoyOnlyCocultVsSpherVsMono, fig.height = 5}
# plot the genes as a line graph
diff.stem.long.df %>% 
  ggplot(aes(x = sample, y = expression, group = genes, colour = genes)) +
  labs(x = NULL, colour = NULL) +
  geom_point() +
  geom_line(size = 1) +
  scale_x_discrete(limits = c("DAOY_monolayer", "DAOY_spheroid", "DAOY_coculture")) +
  theme_classic() +
  theme(axis.title.y = element_text(vjust = 3),
        text = element_text(size = 15),
        axis.text.x = element_text(angle = 45, hjust = 1)) 
```

To plot just the low expression values, exclude ERBB4, CTNNA2, DCX and SOX2
```
{r}
# subset the dataframe excluding high expressing genes
small.genes <- diff.stem.long.df[diff.stem.long.df$genes %in% c("POU3F2", "MEIS1", "ZIC1", "MIR124-2HG",
                                                                "FABP7", "IRX5", "GLI2", "IRX3", "PTCH1"), ]
```

Maybe show this graph as an inset in the bigger graph above
```
{r LineplotGenesOfInterestLowExpressnGenesAvgExpressnOns76cocultSpherMono}
# plot only the genes in 'small.genes'
small.genes %>% 
  ggplot(aes(x = sample, y = expression, group = genes, colour = genes)) +
  labs(x = NULL, colour = NULL) +
  geom_point() +
  geom_line(size = 1) +
  scale_x_discrete(limits = c("DAOY_monolayer", "DAOY_spheroid", "DAOY_coculture")) +
  theme_classic() +
  theme(axis.title.y = element_text(vjust = 3),
        text = element_text(size = 15),
        axis.text.x = element_text(angle = 45, hjust = 1)) 
```

```
{r}
# load the above dataframe of combined Average Expression of the three samples
av.dy.3dfs <- readRDS("outputs50K/INTEGRATE/AvgExpressn.daoy.2d.3d.Cocult.df.downsampled.SCT.rds")
```

```
{r}
# run correlation function
# select the columns with numeric values only
RNAcounts <- av.dy.3dfs %>% 
  select(-genes)
corr.dy.3samples <- cor(RNAcounts)
corr.dy.3samples
```

Refer to https://dk81.github.io/dkmathstats_site/rvisual-corrplots.html to plot this upper triangular correlation plot
```
{r}
# prepare data for an upper triangular plot
upper.dy.3samples <- corr.dy.3samples

# Make upper triangular matrix by setting NA to lower triangular part:
upper.dy.3samples[lower.tri(upper.dy.3samples)] <- NA
upper.dy.3samples
```

```
{r}
# Melt this upper triangular matrix (package reshape2 needed) and remove NA values:
up.m.dy.3samples <- melt(upper.dy.3samples, na.rm = TRUE)
head(up.m.dy.3samples)
nrow(up.m.dy.3samples)
```

```
# save the above correlation dataframe
saveRDS(up.m.dy.3samples, file = "outputs50K/INTEGRATE/CorrTableAllgenesDAOYCocultSpherMono.downsampled.rds")
```

```
{r}
up.m.dy.3samples <- readRDS("outputs50K/INTEGRATE/CorrTableAllgenesDAOYCocultSpherMono.downsampled.rds")
```

```
{r PlotCorrAllGenesDaoyOnlyCocultVsSpherVsMono}
# ggplot upper triangular correlation matrix:
ggplot(data = up.m.dy.3samples, aes(x = Var1, y = Var2, fill = value)) +
  geom_tile() +
  theme_classic() +
  labs(x = "", y = "", fill = "Correlation \n") +
  scale_x_discrete(labels = c("DAOY_monolayer", 
                              "DAOY_spheroid",
                              "DAOY_coculture")) +
  scale_y_discrete(labels = c("DAOY_monolayer", 
                              "DAOY_spheroid",
                              "DAOY_coculture")) +
  theme(axis.text.x = element_text(face = "bold", angle = 45, hjust = 1),
        axis.text.y = element_text(face = "bold", vjust = 1),
        axis.ticks = element_blank(),
        axis.line = element_line(linetype = "blank"),
        panel.background = element_rect(fill = "gray87"),
        panel.grid.major = element_line(colour = "gray47", linetype = "dotted", size = 0.2)
        ) +
  geom_text(aes(label = round(value, 3)), color = "white", fontface = "bold", size = 4) +
  coord_fixed()
```

---------------------------------------------------------------------------------------------------
***Get a correlation matrix of all 7 samples, for DAOY and ONS76 cells***
Here each sample was used without extracting MB cells from the coculture condition

```
# NO NEED TO RE-RUN THIS CODE CHUNK

# load datasets
# repeat DAOY-2D dataset which was sequenced at much greater depth
dy2d.rep <- readRDS("outputs50K/DAOY_2D/filt.CellsGenesRiboMito.daoy2dRep.seurat.rds")
# DAOY-SPHER dataset
dy3d <- readRDS("outputs50K/DAOY_SPHER/filt.RiboMitoCellsGenes.daoySpher.rds")
# DAOY-CBO and CBO datasets
o76.2d <- readRDS("outputs50K/ONS76_2D/filt.NoRiboNoMito.ons76.2d.rds")
o76.3d <- readRDS("outputs50K/ONS76_SPHER/filt.NoRiboNoMito.ons76spher.rds")
dyCbo <- readRDS("outputs50K/DAOY-CBO/filtRiboMitoCellsGenes.daoyCbo.rds")
o76Cbo <- readRDS("outputs50K/ONS76-CBO/filtRiboMitoCellsGenes.ons76cbo.rds")
cbo <- readRDS("outputs50K/CBO/filt.RiboMitoCellsGenes.seurat.cbo.rds")
```

```
all.samples <- merge(dy2d.rep, 
                  y = c(dy3d, o76.2d, o76.3d, dyCbo, o76Cbo, cbo), 
                  add.cell.ids = c("dy2d", "dy3d", "ons2d", "ons3d", "dyCbo", "onsCbo", "Cbo"), 
                  project = "all.merge")
all.samples
```

```
# save the 'all.samples' object
saveRDS(all.samples, file = "./outputs50K/INTEGRATE/SevenSamplesMerged.rds")
```

```
{r}
# load the 'all.samples' object

all.samples <- readRDS("outputs50K/INTEGRATE/SevenSamplesMerged.rds")
```


```
{r}
unique(sapply(X = strsplit(colnames(all.samples), split = "_"), FUN = "[", 1))
```

```
{r}
table(all.samples$orig.ident)
```

```
{r}
# Create metadata dataframe
metadata <- all.samples@meta.data
# Add cell IDs to metadata
metadata$cells <- rownames(metadata)
# Create sample column
metadata$sample <- NA
metadata$sample[which(str_detect(metadata$cells, "^dy2d_"))] <- "DAOY_monolayer"
metadata$sample[which(str_detect(metadata$cells, "^dy3d_"))] <- "DAOY_spheroid"
metadata$sample[which(str_detect(metadata$cells, "^ons2d_"))] <- "ONS76_monolayer"
metadata$sample[which(str_detect(metadata$cells, "^ons3d_"))] <- "ONS76_spheroid"
metadata$sample[which(str_detect(metadata$cells, "^dyCbo_"))] <- "DAOY_organoid"
metadata$sample[which(str_detect(metadata$cells, "^onsCbo_"))] <- "ONS76_organoid"
metadata$sample[which(str_detect(metadata$cells, "^Cbo_"))] <- "organoid"
```

```
{r}
all.samples@meta.data <- metadata
```

```
{r}
all.samples.df <- metadata 
all.samples.df$sample <- factor(all.samples.df$sample, 
                                levels = c("ONS76_monolayer", "DAOY_spheroid",
                                           "ONS76_spheroid", "DAOY_monolayer", 
                                           "ONS76_organoid", "DAOY_organoid", "organoid"))
```

```
{r BarChartCellCounts7samples, fig.height=6, fig.width=5}
ggplot(all.samples.df, aes(x=sample)) + 
  	geom_bar(color = "black", fill = "white") +
  	theme_classic() +
   theme(axis.title = element_blank(), axis.text.x = element_text(angle = 45,
                                                                  hjust = 1)) +
   NoLegend()
```

```
# NO NEED TO RE-RUN THIS CODE CHUNK

all.samples <- NormalizeData(all.samples)
all.samples <- FindVariableFeatures(all.samples)
all.samples <- ScaleData(all.samples)
all.samples <- RunPCA(all.samples, features = VariableFeatures(object = all.samples))
all.samples <- RunUMAP(all.samples, dims = 1:5, reduction = "pca")
```

```
# save the CLUSTERED 'all.samples' merged object
saveRDS(all.samples, file = "./outputs50K/INTEGRATE/SevenSamplesMergedClustered.rds")
```

```
{r}
all.samples <- readRDS("outputs50K/INTEGRATE/SevenSamplesMergedClustered.rds")
```

```
{r DimPlot7samplesMerged, fig.height=3}
DimPlot(all.samples, reduction = "umap", group.by = "sample")
```


Get the AverageExpression with a view to correlation plot in ggplot
```
# NO NEED TO RE-RUN BELOW CODE CHUNKS

dy.2d <- subset(all.samples, subset = (sample == "DAOY_monolayer"))
dy.3d <- subset(all.samples, subset = (sample == "DAOY_spheroid"))
ons76.2d <- subset(all.samples, subset = (sample == "ONS76_monolayer"))
ons76.3d <- subset(all.samples, subset = (sample == "ONS76_spheroid"))
dyCbo <- subset(all.samples, subset = (sample == "DAOY_organoid"))
ons76Cbo <- subset(all.samples, subset = (sample == "ONS76_organoid"))
Cbo <- subset(all.samples, subset = (sample == "organoid"))
```

```
DefaultAssay(all.samples) <- "RNA"
```

```
av.dy.2d <- AverageExpression(dy.2d)$RNA
av.dy.3d <- AverageExpression(dy.3d)$RNA
av.ons76.2d <- AverageExpression(ons76.2d)$RNA
av.ons76.3d <- AverageExpression(ons76.3d)$RNA
av.dyCbo <- AverageExpression(dyCbo)$RNA
av.ons76Cbo <- AverageExpression(ons76Cbo)$RNA
av.Cbo <- AverageExpression(Cbo)$RNA
```

```
colnames(av.dy.2d)
```


```
# make the column names of the matrix objects meaningful
colnames(av.dy.2d) <- "DAOY_monolayer"
colnames(av.dy.3d) <- "DAOY_spheroid"
colnames(av.ons76.2d) <- "ONS76_monolayer"
colnames(av.ons76.3d) <- "ONS76_spheroid"
colnames(av.dyCbo) <- "DAOY_organoid"
colnames(av.ons76Cbo) <- "ONS76_organoid"
colnames(av.Cbo) <- "organoid"
```

```
# convert the matrices of Average Expression to dataframes
# df = dataframe
av.dy.2d.df <- as.data.frame(av.dy.2d) %>% 
  rownames_to_column("genes")
av.dy.3d.df <- as.data.frame(av.dy.3d) %>% 
  rownames_to_column("genes")
av.ons76.2d.df <- as.data.frame(av.ons76.2d) %>% 
  rownames_to_column("genes")
av.ons76.3d.df <- as.data.frame(av.ons76.3d) %>% 
  rownames_to_column("genes")
av.dyCbo.df <- as.data.frame(av.dyCbo) %>% 
  rownames_to_column("genes")
av.ons76Cbo.df <- as.data.frame(av.ons76Cbo) %>% 
  rownames_to_column("genes")
av.Cbo.df <- as.data.frame(av.Cbo) %>% 
  rownames_to_column("genes")
```


```
# left join all 7 dataframes in one go
av.7dfs <- join_all(list(av.dyCbo.df, av.ons76Cbo.df, av.Cbo.df, av.ons76.3d.df,
                         av.dy.3d.df, av.ons76.2d.df, av.dy.2d.df), 
                         by = "genes", type = "left")
dim(av.7dfs)
```

```
head(av.7dfs)
```

```
# save the above correlation dataframe
saveRDS(av.7dfs, 
        file = "outputs50K/INTEGRATE/AvgExpressn.All7Samples.rds")
```

```
{r}
# load the above dataframe of combined Average Expression of all 7 samples
av.7dfs <- readRDS("outputs50K/INTEGRATE/AvgExpressn.All7Samples.rds")
```


```
{r}
# run correlation function
# select the columns with numeric values only
RNAcounts <- av.7dfs %>% 
  select(-genes)
corr.7samples <- cor(RNAcounts)
corr.7samples
```

Refer to https://dk81.github.io/dkmathstats_site/rvisual-corrplots.html to plot this upper triangular correlation plot
```
{r}
# prepare data for an upper triangular plot
upper.7samples <- corr.7samples

# Make upper triangular matrix by setting NA to lower triangular part:
upper.7samples[lower.tri(upper.7samples)] <- NA
upper.7samples
```

```
{r}
# Melt this upper triangular matrix (package reshape2 needed) and remove NA values:
up.m.7samples <- melt(upper.7samples, na.rm = TRUE)
head(up.m.7samples)
nrow(up.m.7samples)
```

```
# save the above correlation dataframe
saveRDS(up.m.7samples, file = "outputs50K/INTEGRATE/CorrTableAllgenes.7samples.rds")
```

```
{r}
up.m.7samples <- readRDS("outputs50K/INTEGRATE/CorrTableAllgenes.7samples.rds")
```

```
{r PlotCorrAllGenes7samples, fig.height=6}
# ggplot upper triangular correlation matrix:
ggplot(data = up.m.7samples, aes(x = Var1, y = Var2, fill = value)) +
  geom_tile() +
  theme_classic() +
  labs(x = "", y = "", fill = "Correlation \n") +
  scale_x_discrete(labels = c("organoid", 
                              "DAOY_organoid",
                              "ONS76_organoid", "DAOY_spheroid",
                              "ONS76_spheroid", "DAOY_monolayer",
                              "ONS76_monolayer")) +
  scale_y_discrete(labels = c("organoid", 
                              "DAOY_organoid",
                              "ONS76_organoid", "DAOY_spheroid",
                              "ONS76_spheroid", "DAOY_monolayer",
                              "ONS76_monolayer")) +
  theme(axis.text.x = element_text(face = "bold", angle = 45, hjust = 1, size = 10),
        axis.text.y = element_text(face = "bold", vjust = 1, size = 10),
        axis.ticks = element_blank(),
        axis.line = element_line(linetype = "blank"),
        panel.background = element_rect(fill = "gray87"),
        panel.grid.major = element_line(colour = "gray47", linetype = "dotted", size = 0.2)
        ) +
  geom_text(aes(label = round(value, 3)), color = "white", fontface = "bold", size = 2) +
  coord_fixed()
```




```
{r}
save.image("./RDataFiles/INTEGRATE_DAOY-CBO_DAOY-SPHER_DAOY-2D_50K.Rmd.RData")
```





