---
title: "INTEGRATE_ONS76-CBO_ONS76-2D_ONS76-SPHER_CBO_seurat_50K"
author: "JJ"
date: "29/08/2022"
output: html_document
---

```{r graphical_output}
options(bitmapType='cairo-png') # this line of code needed BEFORE knitr code in chunk below
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = "png",
                      dpi = 300,
                      echo = TRUE,   # prints the code (if FALSE then only code output is shown)
                      cache = TRUE,
                      fig.path = "./outputs50K/INTEGRATE/figures_ONS76/")
```

```{r load_libraries}
# library(presto)
suppressPackageStartupMessages({
library(Seurat)
library(glmGamPoi)
library(msigdbr)
library(Matrix)
library(patchwork)
library(clustree)
library(EnhancedVolcano)
library(tidyverse)
library(reshape2)
library(plyr)
})
```

```{r}

load("RDataFiles/INTEGRATE_ONS76-CBO_ONS76-2D_ONS76-SPHER_CBO_seurat_50K.RData")
```

-----------------------------------------------------------------------------------------------------
***Perform merge of ONS76-2D and ONS76-SPHER, then PCA/UMAP to compare the samples***
This seems to work best, far better than integration then PCA/RPCA.
```{r}
# load the data

ons76mono <- readRDS("outputs50K/ONS76_2D/filt.NoRiboNoMito.ons76.2d.rds")
ons76spher <- readRDS("outputs50K/ONS76_SPHER/filt.NoRiboNoMito.ons76spher.rds")
common.features <- intersect(rownames(ons76mono), rownames(ons76spher))
length(x = common.features)
```

```{r}
# Merge 'ons76mono' and 'ons76spher' based on common features only
ons76.merge <- merge(ons76spher[common.features, ], 
                     ons76mono[common.features, ],
                     add.cell.id = c("spher", "mono"))
ons76.merge
```

```{r}
head(ons76.merge@meta.data)
tail(ons76.merge@meta.data)
```

```{r}
# Create metadata dataframe
metadata <- ons76.merge@meta.data
# Add cell IDs to metadata
metadata$cells <- rownames(metadata)
# Create sample column
metadata$sample <- NA
metadata$sample[which(str_detect(metadata$cells, "^spher_"))] <- "spheroid"
metadata$sample[which(str_detect(metadata$cells, "^mono_"))] <- "monolayer"
head(metadata)
```

```{r}
ons76.merge@meta.data <- metadata
```

```{r BarPlotCellCountsONS76SpherVs2D, fig.height=4, fig.width=4}
metadata %>% 
  	ggplot(aes(x=sample, fill=sample)) + 
  	geom_bar(fill = "grey") +
  	theme_classic() + NoLegend()
```

In the first PC are numerous proteins with localisation or function related to the extracellular matrix.
```{r}
ons76.merge <- NormalizeData(ons76.merge)
ons76.merge <- FindVariableFeatures(ons76.merge)
ons76.merge <- ScaleData(ons76.merge)
ons76.merge <- RunPCA(ons76.merge, features = VariableFeatures(object = ons76.merge))
print(ons76.merge[["pca"]], dims = 1:5, nfeatures = 10)
```


```{r VizDimLoadingsONS76spherVsMono, fig.height=2, fig.width=2}
VizDimLoadings(ons76.merge, dims = 1, reduction = "pca", balanced = TRUE, nfeatures = 20)
```

```{r VizDimLoadingsONS76spherVsMono, fig.height=2, fig.width=2}
VizDimLoadings(ons76.merge, dims = 2, reduction = "pca", balanced = TRUE, nfeatures = 20)
```

```{r VizDimLoadingsONS76spherVsMono, fig.height=2, fig.width=2}
VizDimLoadings(ons76.merge, dims = 3, reduction = "pca", balanced = TRUE, nfeatures = 20)
```

```{r VizDimLoadingsONS76spherVsMono, fig.height=2, fig.width=2}
VizDimLoadings(ons76.merge, dims = 4, reduction = "pca", balanced = TRUE, nfeatures = 20)
```


```{r PCAPlotONS76spherVsMono}
Idents(ons76.merge) <- "sample"
DimPlot(ons76.merge, reduction = "pca") +
  theme_gray()
```


```{r PCAHeatmapONS76SpherVs2D, fig.height=4, fig.width=9}
# heatmap of genee expression by principal component
DimHeatmap(ons76.merge, dims = 1:3, balanced = TRUE)
```

```{r}
Idents(ons76.merge) <- "sample"
ons76.spher.markers <- FindMarkers(ons76.merge, ident.1 = "spheroid", ident.2 = "monolayer",
                               verbose = FALSE,
                               min.pct = 0.5,
                               logfc.threshold = 0.25)
```

```{r}
# SAVE the list of up and down regulated genes in ONS76_SPHER vs ONS76_2D
saveRDS(ons76.spher.markers, file = "outputs50K/INTEGRATE/ons76SpherVs2D.Up&DownDEGs.rds")
```

```{r}
# load the marker genes of 'ons76.spher.markers' object
ons76.spher.markers <- readRDS("outputs50K/INTEGRATE/ons76SpherVs2D.Up&DownDEGs.rds")
```

```{r}
head(ons76.spher.markers)
```


```{r}
ons76.spher.markers.gene <- ons76.spher.markers %>% 
  rownames_to_column(var = "gene")
```

```{r}
ons76.spher.markers.gene[ons76.spher.markers.gene$gene 
                              %in% c("IFITM3", "IGFBP2", "IGFBP5", "TUBA1A", "STC1", 
                              "BIRC5", "MKI67", "MCM7", "CENPF", "PPA1"), ]
```

```{r}
o76SpherVSmono.up <- ons76.spher.markers.gene %>% 
  arrange(desc(avg_log2FC)) %>% 
  head(20)
o76SpherVSmono.up
```

Use RNA assay when determining gene expression in a merged object
```{r}
ons76spherVsMono.avg <- as.data.frame(log1p(AverageExpression(ons76.merge, verbose = FALSE)$RNA))
```

```{r}
head(ons76spherVsMono.avg)
```


```{r ScatterPlotGeneExpONS76SpherVs2D}
# (MUST USE BACK-TICKS FOR NON-STANDARD COLUMN NAMES, e.g. the dash in ONS76-CBO will throw an error!!!)
  
ons76spherVsMono.avg$gene <- rownames(ons76spherVsMono.avg)
genes.to.label = c("IFITM3", "IGFBP2", "IGFBP5", "MGP", "TUBA1A", "STC1", 
                              "BIRC5", "MKI67", "MCM7", "CENPF", "PPA1")
d1 <- ggplot(ons76spherVsMono.avg, aes(monolayer, spheroid)) + geom_point(colour = "red") + 
  xlab("ONS76_monolayer") + ylab("ONS76_spheroid")
d1 <- LabelPoints(plot = d1, points = genes.to.label, repel = TRUE) +
  theme_classic()
d1
```

```{r}
head(ons76.spher.markers.gene, 20)
```


```{r VolcanoPlotONS76SpherVs2D, fig.height=10, fig.width=15}

genes.to.label.o76SpherVS2d <- c("TINAGL1", "HMGA1", "NRG1", "FST", "DKK1", "KRT18", 
                                 "BIRC5", "CTNNAL1", "CD24", "TK1", "PCLAF", "FRMD5",
                                 "CAV1", "MKI67", "CENPF", "MCM7", "IFITM3", "IGFBP2", 
                                 "SPARC", "IGFBP5", "TUBA1A", "STC1", "PTN", "FTH1", 
                                 "TIMP1", "IGFBP4", "ITM2B", "NTRK2", "IFITM1", "CLCA2", "APOE")

EnhancedVolcano(ons76.spher.markers,
                lab = rownames(ons76.spher.markers),
                x = "avg_log2FC",
                y = "p_val_adj",
                selectLab = genes.to.label.o76SpherVS2d,
                title = NULL,
                subtitle = NULL,
                pointSize = 5.0,
                labSize = 6.0,
                cutoffLineWidth = 1.0,
                axisLabSize = 30,
                legendLabSize = 25,
                legendIconSize = 10,
                legendPosition = "right",
                legendLabels = c("NS", "Log2FC", "p-value", "p-value and Log2FC"),
                drawConnectors = TRUE,
                widthConnectors = 0.5,
                gridlines.major = FALSE,
                gridlines.minor = FALSE)
```

-----------------------------------------------------------------------------------------------
***Merge the ONS76-2D sample, the ONS76-SPHER, ONS76-CBO samples and CBO control***
This is the best way to see how similar/different the samples are
```{r}
# load datasets
# ONS76-2D dataset
o76.2d <- readRDS("outputs50K/ONS76_2D/filt.NoRiboNoMito.ons76.2d.rds")
# ONS76-SPHER dataset
o76.3d <- readRDS("outputs50K/ONS76_SPHER/filt.NoRiboNoMito.ons76spher.rds")
# ONS-CBO and CBO datasets
o76.cbo <- readRDS("outputs50K/ONS76-CBO/filtRiboMitoCellsGenes.ons76cbo.rds")
cbo <- readRDS("outputs50K/CBO/filt.RiboMitoCellsGenes.seurat.cbo.rds")
```

```{r}
o76.4samples <- merge(o76.2d, 
                  y = c(o76.3d, o76.cbo, cbo), 
                  add.cell.ids = c("o762d", "o763d", "o76cbo", "cbo"), 
                  project = "dy4samples.merge")
o76.4samples
```

```{r}
unique(sapply(X = strsplit(colnames(o76.4samples), split = "_"), FUN = "[", 1))
```

```{r}
table(o76.4samples$orig.ident)
```

```{r}
# Create metadata dataframe
metadata <- o76.4samples@meta.data
# Add cell IDs to metadata
metadata$cells <- rownames(metadata)
# Create sample column
metadata$sample <- NA
metadata$sample[which(str_detect(metadata$cells, "^o762d_"))] <- "monolayer"
metadata$sample[which(str_detect(metadata$cells, "^o763d_"))] <- "spheroid"
metadata$sample[which(str_detect(metadata$cells, "^o76cbo_"))] <- "coculture"
metadata$sample[which(str_detect(metadata$cells, "^cbo_"))] <- "organoid"
```

```{r}
# assign metadata object back to the original seurat object
o76.4samples@meta.data <- metadata
```

```{r}
# create a dataframe of metadata as a means to generate a barplot of cell numbers per sample
# in the merged object
o76.4samples.df <- metadata 
o76.4samples.df$sample <- factor(o76.4samples.df$sample, 
                               levels = c("monolayer", "spheroid", "coculture", "organoid"))
```

```{r BarChartONS76CellCounts2d3dONS76CboAndCbo4conditions, fig.height=7, fig.width=5}
ggplot(o76.4samples.df, aes(x=sample, fill=sample)) + 
  	geom_bar(fill = "grey") +
  	theme_classic() +
   theme(axis.title = element_blank(), axis.title.y = element_text(size = 18),
         axis.text.x = element_text(size = 14, angle = 45, hjust = 1), 
         axis.text.y = element_text(size = 14)) +
   NoLegend()
```

```{r}
o76.4samples <- NormalizeData(o76.4samples)
o76.4samples <- FindVariableFeatures(o76.4samples)
o76.4samples <- ScaleData(o76.4samples)
o76.4samples <- RunPCA(o76.4samples, features = VariableFeatures(object = o76.4samples))
o76.4samples <- RunUMAP(o76.4samples, dims = 1:5, reduction = "pca")
```

```{r}
saveRDS(o76.4samples, file = "./outputs50K/INTEGRATE/Ons76-4samplesMerged2dRepSpherCocultCBOpcaUmap.rds")
```

```{r}
o76.4samples <- readRDS("outputs50K/INTEGRATE/Ons76-4samplesMerged2dRepSpherCocultCBOpcaUmap.rds")
```


This result makes sense - separation of the monolayer and spheroid samples but strong overlap of ONS76-CBO and CBO (organoid) samples
```{r DimPlotMergedDyCboSpher2dCboCNTRL4samples, fig.height=6, fig.width=10}
DimPlot(o76.4samples, reduction = "umap", group.by = "sample")
```

```{r VlnPlotONS76StemnessGenes4samplesMergedMonoSpherONS76CBOandCBO, fig.height=10, fig.width=12}
# NO NEED TO RUN THESE CODE CHUNKS

# look at some stemness markers
Idents(o76.4samples) <- "sample"
VlnPlot(o76.4samples, features = c("SOX2", "POU3F2", "GRIA4", "MEIS1", "PRRX1", "IRX3", "CHCHD2"),
        slot = "data", assay = "RNA")
```

```{r VlnPlotONS76OtherGenes4samplesMergedMonoSpherONS76CBOandCBO, fig.height=20, fig.width=12}
VlnPlot(o76.4samples, features = c("IRX5","FABP7", "TTYH1", "ERBB4", "NRG3", "NNAT",
                                        "MIR124-2HG", "DCX", "CTNNA2", "IRX3", "MEIS1", "TWIST1",
                                        "PTCH1", "GLI2", "MYC", "MYCN", "ZIC1", "OTX2", "LHX9"),
        slot = "data", assay = "RNA")
```

-------------------------------------------------------------------------------------------------
***Compare gene expression in ONS76-CBO ONS76 cells vs ONS76-SPHER***
As a first step, subset the ONS76-CBO clustered seurat object for the ONS76 cells. Find the suspected cancer clusters in ONS76-CBO - cluster 3,8,9. Remember in CBO there is only 1 Neurod1+ cluster and there are three in ONS76-CBO. Cluster 9 expresses Hoxb2, Meis1, Gria4 and Pou3f2, whereas there is no expression of Hoxb2, and no significant expression of Meis1, Gria4 or Pou3f2 in the CBO control. See below.
```{r}
# load the CLUSTERED 'phase.ons76cbo.sct' object
phase.ons76cbo.sct.clust <- 
  readRDS("outputs50K/ONS76-CBO/phase.ons76cbo.sct.CCdiffRegressed.filtRiboMito.clustered.rds")

```

```{r DimplotONS76cboSCT, fig.height=6, fig.width=10}
DimPlot(phase.ons76cbo.sct.clust, label = TRUE) # '+ NoLegend()' is an option to insert here
```

Clusters of interest (cancer clusters) are 3,8,9 on comparing ONS76-CBO with CBO. Note SOX2 is also expressed in cluster 9. GRIA4, MEIS1, HOXB2, ZIC1, POU3F2, LHX9 are all expressed in cluster 9
```{r ViolinPlotONS76CboSox2network, fig.width=14}
VlnPlot(phase.ons76cbo.sct.clust, 
        features = c("HOXB2", "SOX2", "POU3F2", "FABP7"),
        slot = "data", assay = "RNA")
```

```{r ViolinPlotONS76CboGranuleDiffGenes, fig.width=14}
VlnPlot(phase.ons76cbo.sct.clust, 
        features = c("GRIA4", "MEIS1", "LHX9", "NEUROD1", "ZIC1"),
        slot = "data", assay = "RNA")
```

Next look at the CBO control
```{r}
# load the SCTransformed, fully filtered seurat object
phase.cbo.sct.clusters <- 
  readRDS("outputs50K/CBO/phase.filtRiboMito.cbo.SCT.CCdiffRegressed.clustered.rds")
```

There is no equivalent to cluster 9 in the ONS76-CBO clustering above. Therefore, cluster 9 of ONS76-CBO must be a cluster of ONS76 cells
```{r ViolinPlotCboSox2network, fig.width=14}
# HOXB2 not expressed
VlnPlot(phase.cbo.sct.clusters, 
        features = c("SOX2", "POU3F2", "FABP7", "GRIA4"), 
        slot = "data", assay = "RNA")
```

```{r ViolinPlotCboGenesOfInterest, fig.width=14}
# HOXB2 not expressed
VlnPlot(phase.cbo.sct.clusters, 
        features = c("MEIS1", "LHX9", "NEUROD1", "ZIC1"), 
        slot = "data", assay = "RNA")
```

----------------------------------------------------------------------------------------
***Extract the ONS76 cells which were grown in the CBO***
```{r}
# No need to rerun this code
# extract the ONS76 cancer cells which are in cluster 3,8,9 of ONS76-CBO

Idents(phase.ons76cbo.sct.clust) <- "seurat_clusters"
ons76cboCells <- subset(phase.ons76cbo.sct.clust, subset = (seurat_clusters %in% c(3, 8, 9)))
```

There are 472 ONS76 cells in the ONS76-CBO coculture
```{r}
# No need to rerun this code
# number of cells = 472. These are ONS76 cells in the coculture
length(Cells(x = ons76cboCells))
```

```{r}
# calculating the number of cells in different ways shows the same result
phase.ons76cbo.sct.clust@meta.data %>% 
  filter(seurat_clusters %in% c("3", "8", "9")) %>% 
  nrow()
```


```{r}
# No need to rerun this code
# pass the extracted cells to a new object, which is a character vector
ons76CellID <- Cells(x = ons76cboCells)
class(ons76CellID)
length(ons76CellID)
```

```{r}
# NO NEED TO RE-RUN THESE CODE CHUNKS
# get the unclustered seurat object - this will include both ONS76 cells and CBO cells
ons76cboAll <- readRDS("outputs50K/ONS76-CBO/filtRiboMitoCellsGenes.ons76cbo.rds")
```

```{r}
# Create a new separate metadata dataframe with a new column of Cell IDs
metadata <- ons76cboAll@meta.data
# Add cell IDs to metadata
metadata$cellIDs <- rownames(metadata)
# add the metadata with cellID column back to original object
ons76cboAll@meta.data <- metadata
```


Sucessfully used the ONS76 cell IDs (from cluster 3,8,9) in ONS76-CBO seurat object, 'filtRiboMitoCellsGenes.ons76cbo.rds', to subset this seurat object so that it only contains the 472 ONS76 cells
```{r}
# No need to rerun this code
# subset this Seurat object according to ons76CellID (IDs of only ONS76 cells in co-culture)
ons76cbo.onsOnly <- subset(ons76cboAll, subset = cellIDs %in% ons76CellID)
dim(ons76cbo.onsOnly)
```

```{r}
# SAVE the 'ons76cbo.onsOnly' seurat object which contains the subset of ONS76-CBO cells that is
# only cancer cells - there are 472 cells

saveRDS(ons76cbo.onsOnly, file = "outputs50K/INTEGRATE/ons76only.ons76cbo.RiboMitoFiltered.rds")
```

----------------------------------------------------------------------------------------------
***Integrate the objects - 1. ONS76 cells grown in coculture, 2. ONS76-SPHER cells. *** 
Conventional workflow
```{r}
# objects to do an integrated analysis on:
# 'ons76cbo.onsOnly' and 'ons76spher'

ons76only <- readRDS("outputs50K/INTEGRATE/ons76only.ons76cbo.RiboMitoFiltered.rds")
dim(ons76only)
ons76spher <- readRDS("outputs50K/ONS76_SPHER/filt.NoRiboNoMito.ons76spher.rds")
dim(ons76spher)
```

```{r}
ons76only@meta.data <- ons76only@meta.data %>% 
  select(-cellID)
head(ons76only@meta.data)
```

Prepare the integration of ONS76only cells in coculture and ONS76-SPHER
```{r}
ons76.2samples.list <- list(ons76only, ons76spher)

# normalize and identify variable features for each dataset independently (from https://satijalab.org/seurat/articles/integration_rpca.html)
ons76.2samples.list <- lapply(ons76.2samples.list, FUN = function(x) {
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})
```

```{r}
# select features that are repeatedly variable across datasets for integration' run PCA on each
# dataset using these features
features <- SelectIntegrationFeatures(object.list = ons76.2samples.list)

ons76.2samples.list <- lapply(ons76.2samples.list, FUN = function(x){
  x <- ScaleData(x, features = features, verbose = FALSE)
  x <- RunPCA(x, features = features, verbose = FALSE)
})
```

```{r}
ons76.2samples.anchors <- FindIntegrationAnchors(object.list = ons76.2samples.list,
                                          anchor.features = features)
ons76only.integrated <- IntegrateData(anchorset = ons76.2samples.anchors)
```

```{r}
# this step is necessary as ggplot will not work with 'hyphens'
ons76only.integrated@meta.data[ons76only.integrated@meta.data == "ONS76-CBO"] <- "ONS76_CBO"
```

```{r}
# cell cycle scoring
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

ons76only.integrated <- CellCycleScoring(ons76only.integrated,
                                 g2m.features = g2m.genes,
                                 s.features = s.genes)
```

```{r}
# get the cell cycle difference (CC difference)
ons76only.integrated$CC.difference <- ons76only.integrated$S.Score - ons76only.integrated$G2M.Score 
```

```{r}
# specify that we will perform downstream analysis on the corrected data. Note that the
# original unmodified data still resides in the 'RNA' assay. Regress the CC difference

DefaultAssay(ons76only.integrated) <- "integrated"

# Run the standard workflow for visualization and clustering
ons76only.integrated <- ScaleData(ons76only.integrated, vars.to.regress = "CC.difference", verbose = FALSE)
ons76only.integrated <- RunPCA(ons76only.integrated, npcs = 30, verbose = FALSE)
ons76only.integrated <- RunUMAP(ons76only.integrated, reduction = "pca", dims = 1:30)
ons76only.integrated <- FindNeighbors(ons76only.integrated, reduction = "pca", dims = 1:30)
ons76only.integrated <- FindClusters(ons76only.integrated, resolution = 0.5)
```

These samples are not well integrated looking at the UMAP plot
```{r DimplotUMAPIntegrationOns76onlyCboVSspher, fig.height=3}
DimPlot(ons76only.integrated, reduction = "umap", group.by = "orig.ident")
```


```{r DimPlotUMAPclustersOns76cocultSpher}
DimPlot(ons76only.integrated, reduction = "umap", group.by = "seurat_clusters")
```

```{r}
# Pull Info for 20 features, balanced positive/negative, for PC1
TopFeatures(object = ons76only.integrated[["pca"]], dim = 1, nfeatures = 40, balanced = TRUE)
```

```{r}
# No need to re-run this code chunk

DefaultAssay(ons76only.integrated) <- "RNA"
Idents(ons76only.integrated) <- "orig.ident"
ons76only.integrated.markers <- FindMarkers(ons76only.integrated, 
                                                ident.1 = "ONS76_CBO", 
                                                ident.2 = "ons76spher",
                                                verbose = FALSE,
                                                min.pct = 0.25,
                                                logfc.threshold = 0.25)
ons76only.integrated.markers.gene <- ons76only.integrated.markers %>% 
  rownames_to_column(var = "gene")
```

```{r}
# SAVE the list of up and down regulated genes in ONS76 only cells in coculture (ONS76-CBO) vs 
# ONS76_SPHER
saveRDS(ons76only.integrated.markers.gene, file = "outputs50K/INTEGRATE/ons76onlyInCocultVsSpher.Up&DownDEGs.rds")
```


```{r}
# load the above data
ons76only.integrated.markers.gene <- 
  readRDS("outputs50K/INTEGRATE/ons76onlyInCocultVsSpher.Up&DownDEGs.rds")
```


```{r}
ons76only.integrated.markers.gene %>% 
arrange(desc(avg_log2FC)) %>% 
  slice_head(n=10)
```

```{r}
ons76only.integrated.markers.gene %>% 
  filter(gene %in% c("NEUROD1", "ZIC1"))
```

```{r VolcanoPlotONS76cocultExtractVsSpher, fig.height=10, fig.width=15}
# plot of ONS76 cells extracted from coculture vs. ONS76-Spheroid
genes.to.label.o76SpherVS2d <- c("NNAT", "AFF3", "NR2F1", "EYA2", "NEUROD1", "KIAA1211", "TRHDE",
                                 "KCNB2", "TUBB2B", "THSD7B", "NKAIN3", "RUNX1T1", "GREB1L",
                                 "KCNT2", "PPP1R17", "LGALS1", "CHCHD2", "IFITM3", "PLP2",
                                 "S100A2", "CAST", "ID1", "MGP", "IGFBP5", "APOE", "FABP4",
                                 "VCAN",  "MIR205HG", "CHI3L1")

EnhancedVolcano(ons76only.integrated.markers.gene,
                lab = ons76only.integrated.markers.gene$gene,
                x = "avg_log2FC",
                y = "p_val_adj",
                selectLab = genes.to.label.o76SpherVS2d,
                title = NULL,
                subtitle = NULL,
                pointSize = 5.0,
                labSize = 6.0,
                cutoffLineWidth = 1.0,
                axisLabSize = 30,
                legendLabSize = 25,
                legendIconSize = 10,
                legendPosition = "right",
                legendLabels = c("NS", "Log2FC", "p-value", "p-value and Log2FC"),
                drawConnectors = TRUE,
                widthConnectors = 0.5,
                gridlines.major = FALSE,
                gridlines.minor = FALSE)
```


Another way of extracting the PCs
```{r}
print(ons76only.integrated[["pca"]], dims = 1:5, nfeatures = 10)
```

```{r VizDimLoadingsONS76cocultVsSpher, fig.height=10, fig.width=10}
VizDimLoadings(ons76only.integrated, dims = 1:4, reduction = "pca", balanced = TRUE, nfeatures = 30)
```

```{r PCAHeatmapONS76SpherVs2D}
# this does not work - raised issue with Seurat team in Satija lab
DimHeatmap(ons76only.integrated, dims = 1, balanced = TRUE, slot = "scale.data")
```

-------------------------------------------------------------------------------------------------------------------
***Integrate ONS76-2D, ONS76-SPHER and ONS76 cells in ONS76-CBO as a step to generate a correlation matrix of gene expression***
Get the correlation between ONS76-2D, ONS76-SPHER and ONS76 cells grown in ONS76-CBO. Show as a correlation matrix

```{r}
# load the objects
ons76mono <- readRDS("outputs50K/ONS76_2D/filt.NoRiboNoMito.ons76.2d.rds")
ons76spher <- readRDS("outputs50K/ONS76_SPHER/filt.NoRiboNoMito.ons76spher.rds")
ons76only.ons76cbo <- readRDS("outputs50K/INTEGRATE/ons76only.ons76cbo.RiboMitoFiltered.rds")
```

```{r}
dim(ons76mono)
dim(ons76spher)
dim(ons76only.ons76cbo)
```

```{r}
# start the integration (use  standard PCA)
ons76.3samples.list <- list(ons76mono, ons76spher, ons76only.ons76cbo)

# normalize and identify variable features for each dataset independently (from https://satijalab.org/seurat/articles/integration_rpca.html)
ons76.3samples.list <- lapply(ons76.3samples.list, FUN = function(x) {
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})
```

```{r}
# select features that are repeatedly variable across datasets for integration' run PCA on each
# dataset using these features
features <- SelectIntegrationFeatures(object.list = ons76.3samples.list)

ons76.3samples.list <- lapply(ons76.3samples.list, FUN = function(x){
  x <- ScaleData(x, features = features, verbose = FALSE)
  x <- RunPCA(x, features = features, verbose = FALSE)
})
```

```{r}
# continue integration of 3 samples, use PCA
ons76.3samples.anchors <- FindIntegrationAnchors(object.list = ons76.3samples.list,
                                          anchor.features = features)
ons76.3samples.integrated <- IntegrateData(anchorset = ons76.3samples.anchors)
```

```{r}
# this step is necessary as ggplot will not work with 'hyphens'
ons76.3samples.integrated@meta.data[ons76.3samples.integrated@meta.data == "ONS76-CBO"] <- "ONS76_coculture"
ons76.3samples.integrated@meta.data[ons76.3samples.integrated@meta.data == "ONS76-2D"] <- "ONS76_monolayer"
ons76.3samples.integrated@meta.data[ons76.3samples.integrated@meta.data == "ons76spher"] <- "ONS76_spheroid"
```

```{r}
# cell cycle scoring
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

ons76.3samples.integrated <- CellCycleScoring(ons76.3samples.integrated,
                                 g2m.features = g2m.genes,
                                 s.features = s.genes)
```

```{r}
# get the cell cycle difference (CC difference)
ons76.3samples.integrated$CC.difference <- ons76.3samples.integrated$S.Score - ons76.3samples.integrated$G2M.Score 
```

```{r}
# specify that we will perform downstream analysis on the corrected data. Note that the
# original unmodified data still resides in the 'RNA' assay. Regress the CC difference

DefaultAssay(ons76.3samples.integrated) <- "integrated"

# Run the standard workflow for visualization and clustering
ons76.3samples.integrated <- ScaleData(ons76.3samples.integrated, vars.to.regress = "CC.difference", verbose = FALSE)
ons76.3samples.integrated <- RunPCA(ons76.3samples.integrated, npcs = 30, verbose = FALSE)
ons76.3samples.integrated <- RunUMAP(ons76.3samples.integrated, reduction = "pca", dims = 1:30)
ons76.3samples.integrated <- FindNeighbors(ons76.3samples.integrated, reduction = "pca", dims = 1:30)
ons76.3samples.integrated <- FindClusters(ons76.3samples.integrated, resolution = 0.2)
```

```{r}
saveRDS(ons76.3samples.integrated, 
        file = "./outputs50K/INTEGRATE/ons76.mono.spher.extractCocult.integrated.CCdiffRegressed.rds")
```

```{r}
ons76.3samples.integrated <- 
  readRDS("outputs50K/INTEGRATE/ons76.mono.spher.extractCocult.integrated.CCdiffRegressed.rds")
```


```{r DimplotUMAPons76ExtractCocultSpherMono}
DimPlot(ons76.3samples.integrated, reduction = "umap", group.by = "orig.ident")
```


```{r}
head(ons76.3samples.integrated@meta.data)
```

```{r}
table(ons76.3samples.integrated$orig.ident)
```

```{r}
# subset the integrated object to extract each sample on its own prior to calculating
# Average Expression. This avoids problems with infinite values in results of AverageExpression
# when an attempt is made to calculate pseudobulk values on each sample separately!!

ons76.2d <- subset(ons76.3samples.integrated, subset = (orig.ident == "ONS76_monolayer"))
ons76.3d <- subset(ons76.3samples.integrated, subset = (orig.ident == "ONS76_spheroid"))
ons76cocult <- subset(ons76.3samples.integrated, subset = (orig.ident == "ONS76_coculture")) 
```

```{r}
# compute average expression according to sample identity, but first make the Default Assay = RNA
# very important that NormaliseData(seur_obj) is run BEFORE computing AverageExpression!!

DefaultAssay(ons76.3samples.integrated) <- "RNA"     # not sure this is needed here!
av.ons76.2d <- AverageExpression(ons76.2d)$RNA
av.ons76.3d <- AverageExpression(ons76.3d)$RNA
av.ons76cocult <- AverageExpression(ons76cocult)$RNA
```

```{r}
# compute lengths of objects
length(av.ons76.2d)
length(av.ons76.3d)
length(av.ons76cocult)
```

```{r}
# make the column names of the matrix objects meaningful
colnames(av.ons76.2d) <- "ONS76_monolayer"
colnames(av.ons76.3d) <- "ONS76_spheroid"
colnames(av.ons76cocult) <- "ONS76_coculture"
```

```{r}
# convert the matrices of Average Expression to dataframes
# df = dataframe
av.ons76.2d.df <- as.data.frame(av.ons76.2d) %>% 
  rownames_to_column("genes")
av.ons76.3d.df <- as.data.frame(av.ons76.3d) %>% 
  rownames_to_column("genes")
av.ons76cocult.df <- as.data.frame(av.ons76cocult) %>% 
  rownames_to_column("genes")
```

```{r}
# left join all three dataframes in one go
av.ons76.3dfs <- join_all(list(av.ons76.2d.df, av.ons76.3d.df, av.ons76cocult.df), 
                       by = "genes", type = "left")
```

```{r}
# identify an NA values - there are none
colSums(is.na(av.ons76.3dfs))
```

```{r}
head(av.ons76.3dfs)
```

```{r}
# save the combined dataframe of Average Expression in subsets of the 
# integrated object (av.ons76.3samples.integrated)
saveRDS(av.ons76.3dfs, file = "outputs50K/INTEGRATE/AvgExpressn.ons76.2d.3d.Cocult.df.rds")
```

```{r}
# load the above dataframe of combined Average Expression of the three samples
av.ons76.3dfs <- readRDS("outputs50K/INTEGRATE/AvgExpressn.ons76.2d.3d.Cocult.df.rds")
```

Expression of differentiation markers, and 'stemness' markers found in the ONS76 cells in co-culture shows that differentiation markers are progressively expressed and maximal in ONS76 coculture. The 'stemness' markers MEIS1, POU3F2 and HOXB2 are expressed highest in ONS76 coculture.
```{r}
# find expression of genes related to differentiation and stemness in the 3 samples:
# 1. ONS76-2D 2. ONS76-SPHER 3. ONS76-coculture

av.ons76.3dfs[av.ons76.3dfs$genes %in% c("NEUROD1", "POU3F2", "MEIS1", "DCX", "ZIC1", "PTCH1", "SOX2"), ]
```


```{r}
# plot the changes in gene expression for different samples.
# change the format of the data to long format for ggplot
diff.stem.genes <- av.ons76.3dfs[av.ons76.3dfs$genes %in% 
                                   c("NEUROD1", "POU3F2", "MEIS1", "DCX", "ZIC1", "PTCH1", "SOX2"), ]
diff.stem.long.df <- diff.stem.genes %>%
  pivot_longer(cols = -genes,
               names_to = "sample",
               values_to = "expression")
``` 
  
```{r}
diff.stem.long.df
```

```{r}
# the genes are ordered alphabetically in the legend which does not correspond to their order in 
# the plot, where NEUROD1 has the highest value and GLI2 the lowest value
diff.stem.long.df$genes <- as.factor(diff.stem.long.df$genes)
levels(diff.stem.long.df$genes)
```

```{r}
# reorder the legend by specifying the correct order of the genes
diff.stem.long.df$genes <- factor(diff.stem.long.df$genes, 
                                  levels = c("NEUROD1", "DCX", "MEIS1", "ZIC1", "SOX2", "POU3F2", "PTCH1"))
levels(diff.stem.long.df$genes)
```

The expression of high and low expressing genes in a single plot makes the low expression values hard to tell apart
```{r LineplotGenesOfInterestAvgExpressnOns76cocultSpherMono, fig.height = 8}
# plot the genes as a line graph
diff.stem.long.df %>% 
  ggplot(aes(x = sample, y = expression, group = genes, colour = genes)) +
  labs(x = NULL, colour = NULL) +
  geom_point() +
  geom_line(size = 1) +
  scale_x_discrete(limits = c("ONS76_monolayer", "ONS76_spheroid", "ONS76_coculture")) +
  theme_classic() +
  theme(axis.title.y = element_text(vjust = 3),
        text = element_text(size = 15),
        axis.text.x = element_text(angle = 45, hjust = 1)) 
```

To plot just the low expression values, exclude NEUROD1 and DCX
```{r}
# subset the dataframe excluding high expressing genes
small.genes <- diff.stem.long.df[diff.stem.long.df$genes %in% c("MEIS1", "ZIC1", "SOX2", "POU3F2", "PTCH1"), ]
```

```{r}
small.genes
```

Maybe show this graph as an inset in the bigger graph above. Note SOX2 is decreased in ONS76-only cells extracted from coculture because I used all the ONS76 cell clusters in the ONS76-CBO sample. There is only 1 SOX2 cluster in this group of cells.
```{r LineplotGenesOfInterestLowExpressnGenesAvgExpressnOns76cocultSpherMono}
# plot only the genes in 'small.genes'
small.genes %>% 
  ggplot(aes(x = sample, y = expression, group = genes, colour = genes)) +
  labs(x = NULL, colour = NULL) +
  geom_point() +
  geom_line(size = 1) +
  scale_x_discrete(limits = c("ONS76_monolayer", "ONS76_spheroid", "ONS76_coculture")) +
  theme_classic() +
  theme(axis.title.y = element_text(vjust = 3),
        text = element_text(size = 15),
        axis.text.x = element_text(angle = 45, hjust = 1)) 
```


```{r}
# run correlation function
# select the columns with numeric values only
RNAcounts <- av.ons76.3dfs %>% 
  select(-genes)
corr.ons76.3samples <- cor(RNAcounts)
corr.ons76.3samples
```

Refer to https://dk81.github.io/dkmathstats_site/rvisual-corrplots.html to plot this upper triangular correlation plot
```{r}
# prepare data for an upper triangular plot
upper.ons76.3samples <- corr.ons76.3samples

# Make upper triangular matrix by setting NA to lower triangular part:
upper.ons76.3samples[lower.tri(upper.ons76.3samples)] <- NA
upper.ons76.3samples
```

```{r}
# Melt this upper triangular matrix (package reshape2 needed) and remove NA values:
up.m.ons76.3samples <- melt(upper.ons76.3samples, na.rm = TRUE)
head(up.m.ons76.3samples)
nrow(up.m.ons76.3samples)
```

```{r}
# save the above correlation dataframe
saveRDS(up.m.ons76.3samples, file = "outputs50K/INTEGRATE/CorrTableAllgenesONS76CocultSpherMono.rds")
```

```{r}
up.m.ons76.3samples <- readRDS("outputs50K/INTEGRATE/CorrTableAllgenesONS76CocultSpherMono.rds")
```


```{r PlotCorrAllGenesONS76CocultSpherMono}
# ggplot upper triangular correlation matrix:
ggplot(data = up.m.ons76.3samples, aes(x = Var1, y = Var2, fill = value)) +
  geom_tile() +
  theme_classic() +
  labs(x = "", y = "", fill = "Correlation \n") +
  scale_x_discrete(labels = c("ONS76_monolayer", 
                              "ONS76_spheroid",
                              "ONS76_coculture")) +
  scale_y_discrete(labels = c("ONS76_monolayer", 
                              "ONS76_spheroid",
                              "ONS76_coculture")) +
  theme(axis.text.x = element_text(face = "bold", angle = 45, hjust = 1),
        axis.text.y = element_text(face = "bold", vjust = 1),
        axis.ticks = element_blank(),
        axis.line = element_line(linetype = "blank"),
        panel.background = element_rect(fill = "gray87"),
        panel.grid.major = element_line(colour = "gray47", linetype = "dotted", size = 0.2)
        ) +
  geom_text(aes(label = round(value, 3)), color = "white", fontface = "bold", size = 4) +
  coord_fixed()
```

-------------------------------------------------------------------------------------------------------
***Compare the ONS76 coculture cells, ONS76-2D and ONS76-SPHER cells by conventional integration***
This did not result in as much overlap of cells as with SCT based integration. Given we think the cell type (ONS76) in the 3 samples are all the same, the SCT integration clearly performs better

```{r}
# load the objects
ons76mono <- readRDS("outputs50K/ONS76_2D/filt.NoRiboNoMito.ons76.2d.rds")
ons76spher <- readRDS("outputs50K/ONS76_SPHER/filt.NoRiboNoMito.ons76spher.rds")
ons76only.ons76cbo <- readRDS("outputs50K/INTEGRATE/ons76only.ons76cbo.RiboMitoFiltered.rds")
```

```{r}
# start the integration (use RPCA)
ons76.3samples.list <- list(ons76mono, ons76spher, ons76only.ons76cbo)

# normalize and identify variable features for each dataset independently (from https://satijalab.org/seurat/articles/integration_rpca.html)
ons76.3samples.list <- lapply(ons76.3samples.list, FUN = function(x) {
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})
```

```{r}
# select features that are repeatedly variable across datasets for integration' run PCA on each
# dataset using these features
features <- SelectIntegrationFeatures(object.list = ons76.3samples.list)

ons76.3samples.list <- lapply(ons76.3samples.list, FUN = function(x){
  x <- ScaleData(x, features = features, verbose = FALSE)
  x <- RunPCA(x, features = features, verbose = FALSE)
})
```

```{r}
# continue integration of 3 samples, using PCA
ons76.3samples.anchors.2 <- FindIntegrationAnchors(object.list = ons76.3samples.list,
                                          anchor.features = features)
ons76.3samples.combined <- IntegrateData(anchorset = ons76.3samples.anchors)
```

```{r}
# this step is necessary as ggplot will not work with 'hyphens'
ons76.3samples.combined@meta.data[ons76.3samples.combined@meta.data == "ONS76-CBO"] <- "ONS76_CBO"
ons76.3samples.combined@meta.data[ons76.3samples.combined@meta.data == "ONS76-2D"] <- "ONS76_2D"
```

```{r}
# cell cycle scoring
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

ons76.3samples.combined <- CellCycleScoring(ons76.3samples.combined,
                                 g2m.features = g2m.genes,
                                 s.features = s.genes)
```

```{r}
# get the cell cycle difference (CC difference)
ons76.3samples.combined$CC.difference <- ons76.3samples.combined$S.Score - ons76.3samples.combined$G2M.Score 
```

```{r}
# specify that we will perform downstream analysis on the corrected data. Note that the
# original unmodified data still resides in the 'RNA' assay. Regress the CC difference

DefaultAssay(ons76.3samples.combined) <- "integrated"

# Run the standard workflow for visualization and clustering
ons76.3samples.combined <- ScaleData(ons76.3samples.combined, vars.to.regress = "CC.difference", verbose = FALSE)
ons76.3samples.combined <- RunPCA(ons76.3samples.combined, npcs = 30, verbose = FALSE)
ons76.3samples.combined <- RunUMAP(ons76.3samples.combined, reduction = "pca", dims = 1:30)
ons76.3samples.combined <- FindNeighbors(ons76.3samples.combined, reduction = "pca", dims = 1:30)
ons76.3samples.combined <- FindClusters(ons76.3samples.combined, resolution = 0.2)
```

```{r}
# save the clustered 'ons76.3samples.combined' that underwent conventional integration by PCA,
# and resolution = 0.2
saveRDS(ons76.3samples.combined, 
        file = "outputs50K/INTEGRATE/ons76cocult.2d.spher.integrated.clustered.conventionalPCA.rds")
```

```{r}
ons76.3samples.combined <- 
  readRDS("outputs50K/INTEGRATE/ons76cocult.2d.spher.integrated.clustered.conventionalPCA.rds")
```

```{r}
DimPlot(ons76.3samples.combined, reduction = "umap", group.by = "orig.ident")
```

```{r}
DimPlot(ons76.3samples.combined, reduction = "umap", group.by = "seurat_clusters")
```

-------------------------------------------------------------------------------------------------------
***Compare the ONS76 coculture cells, ONS76-2D and ONS76-SPHER cells by SCT integration***

This results in greater overlap of the cells from the different conditions than with conventional integration.
```{r}
# No need to re-run this code chunk

# load the objects
ons76mono <- readRDS("outputs50K/ONS76_2D/filt.NoRiboNoMito.ons76.2d.rds")
ons76spher <- readRDS("outputs50K/ONS76_SPHER/filt.NoRiboNoMito.ons76spher.rds")
ons76only.ons76cbo <- readRDS("outputs50K/INTEGRATE/ons76only.ons76cbo.RiboMitoFiltered.rds")
```

```{r}
# run SCTransform and then prep integration
ons76.3samples.list <- list(ons76mono, ons76spher, ons76only.ons76cbo)
ons76.3samples.list.sct <- suppressWarnings(lapply(X = ons76.3samples.list, FUN = SCTransform))
features <- SelectIntegrationFeatures(object.list = ons76.3samples.list.sct, nfeatures = 3000)
ons76.3samples.list.sct <- PrepSCTIntegration(object.list = ons76.3samples.list.sct, anchor.features = features)
```

```{r}
# No need to re-run this code chunk

ons76.3samples.sct.anchors <- FindIntegrationAnchors(object.list = ons76.3samples.list.sct, 
                                            normalization.method = "SCT",
                                            anchor.features = features)
ons76.3samples.combined.sct <- IntegrateData(anchorset = ons76.3samples.sct.anchors, normalization.method = "SCT")
```

```{r}
# No need to re-run this code chunk

ons76.3samples.combined.sct <- RunPCA(ons76.3samples.combined.sct, verbose = FALSE)
ons76.3samples.combined.sct <- RunUMAP(ons76.3samples.combined.sct, reduction = "pca", dims = 1:30)
ons76.3samples.combined.sct <- FindNeighbors(ons76.3samples.combined.sct, 
                                           reduction = "pca", 
                                           dims = 1:30)
ons76.3samples.combined.sct <- FindClusters(ons76.3samples.combined.sct, resolution = 0.2)
```

```{r}
# ONS76_2D = 1935; ONS76-CBO = 269; ONS76-spher = 660 cells
table(ons76.3samples.combined.sct$orig.ident)
```

```{r}
# change the orig.ident notation for future ggplots
ons76.3samples.combined.sct@meta.data[ons76.3samples.combined.sct@meta.data == "ONS76-2D"] <- "ONS76_monolayer"
ons76.3samples.combined.sct@meta.data[ons76.3samples.combined.sct@meta.data == "ons76spher"] <- "ONS76_spheroid"
ons76.3samples.combined.sct@meta.data[ons76.3samples.combined.sct@meta.data == "ONS76-CBO"] <- "ONS76_coculture"
```

```{r}
# save the SCT integrated object, 'ons76.3samples.combined.sct'
saveRDS(ons76.3samples.combined.sct, 
        file = "outputs50K/INTEGRATE/ons76cocult.2d.spher.clustered.integrated.SCT.pca.rds")
```

```{r}
# load the SCT integrated dataset
ons76.3samples.combined.sct <- 
  readRDS("outputs50K/INTEGRATE/ons76cocult.2d.spher.clustered.integrated.SCT.pca.rds")
```

```{r DimPlotONS76Integrated3samplesONS76monoSpherExtractCoculture, fig.height=6, fig.width=10}
DimPlot(ons76.3samples.combined.sct, reduction = "umap", group.by = "orig.ident")
```

-----------------------------------------------------------------------------------------------------
***Find upregulated DEGs in each condition and plot (Do-)Heatmap of top genes for each condition***

For this procedure, use the SCT integrated samples
```{r}
# No need to re-run this code chunk

# this step is necessary as ggplot will not work with 'hyphens'
ons76.3samples.combined.sct@meta.data[ons76.3samples.combined.sct@meta.data == "ONS76-CBO"] <- "ONS76_coculture"
ons76.3samples.combined.sct@meta.data[ons76.3samples.combined.sct@meta.data == "ONS76-2D"] <- "ONS76_monolayer"
ons76.3samples.combined.sct@meta.data[ons76.3samples.combined.sct@meta.data == "ons76spher"] <- "ONS76_spheroid"
```

```{r}
DefaultAssay(ons76.3samples.combined.sct) <- "RNA"
Idents(ons76.3samples.combined.sct) <- "orig.ident"
```

```{r}
# No need to re-run this code chunk

ons76.all.UPmarkers <- FindAllMarkers(object = ons76.3samples.combined.sct,
                                     min.pct = 0.25,
                                     only.pos = TRUE)
```

```{r}
# save the positive markers of each of the three conditions (ONS76 cocult, 2D, spher)
saveRDS(ons76.all.UPmarkers, 
        file = "outputs50K/INTEGRATE/Pos.DEGs.ons76cocult2dSpher.integrated.sct.pca.rds")
```

```{r}
ons76.all.UPmarkers <- readRDS("outputs50K/INTEGRATE/Pos.DEGs.ons76cocult2dSpher.integrated.sct.pca.rds")
```


Nice to see NEUROD1 and EYA2 in the list of top markers by p-values for ONS76 coculture
```{r}
test <- ons76.all.UPmarkers[ons76.all.UPmarkers$cluster == "ONS76_coculture", ]
test %>% 
  arrange(p_val_adj) %>% 
  head(10)
```

```{r}
# number of markers = 8867
nrow(ons76.all.UPmarkers)
```

```{r}
# sort the table on adjusted p-values, and choose the top 10 genes based on lowest ('-' sign) p_val_adj
top10 <- ons76.all.UPmarkers %>% 
 group_by(cluster) %>% 
  top_n(10, -p_val_adj)
```

```{r}
top10
```

```{r}
ons76.3samples.combined.sct <- ScaleData(ons76.3samples.combined.sct,
                                         features = as.character(unique(top10$gene)), 
                                         assay = "RNA")
```


```{r HeatmapMarkersONS76cocultSpherMono, fig.height=10, fig.width=12}
Idents(ons76.3samples.combined.sct) <- "orig.ident"
DoHeatmap(object = ons76.3samples.combined.sct, 
          features = as.character(unique(top10$gene)),
          assay = "RNA") +
  guides(color = "none") +
  theme(axis.text.y = element_text(size = 15))
```


```{r VlnPlotStemAndOtherGenesONS76ExtractCboMonoSpher3samples, fig.height=15, fig.width=12}
# this is the same set of genes as in the DAOY integration .Rmd file; genes are shown in the same 
# order for comparison
VlnPlot(ons76.3samples.combined.sct, 
        features = c("SOX2", "POU3F2","FABP7", "IRX5", "TTYH1", "NRG3",                                                                     "NNAT","MIR124-2HG", "DCX", "CTNNA2", "MEIS1", "CD4"),
        slot = 'data',
        assay = 'RNA')
```


This plot is mainly to show SHH readouts. Of these HHIP, GLI3 are strongly expressed in ONS76-2D cells that have high SOX2 expression
```{r  VlnPlotHoxB2ShhGenesONS76ExtractCboMonoSpher3samples, fig.height=12, fig.width=12}

VlnPlot(ons76.3samples.combined.sct, features = c("HOXB2","IRX3", "TWIST1",
                                        "PTCH1", "GLI1", "GLI2", "GLI3", "HHIP",
                                        "MYC", "MYCN", "ZIC1", "OTX2"),
        slot = "data", assay = "RNA")
```

------------------------------------------------------------------------------------------------------------
***Integration of ONS76-2D, ONS76-SPHEROID and ONS76-only extracted from coculture, cluster 9 and clusters3+8***

```{r}
Idents(phase.ons76cbo.sct.clust) <- "seurat_clusters"
cl9.ons76cboCells <- subset(phase.ons76cbo.sct.clust, subset = (seurat_clusters %in% c(9)))
```

Find the cell IDs of the cluster 9 cells in "phase.ons76cbo.sct.clust" and extract those cells
```{r}
cl9cellID <- WhichCells(phase.ons76cbo.sct.clust, idents = "9")
length(cl9cellID)
```

For ONS76-only cocultures: use the vector of cell IDs to subset the object "filtRiboMitoCellsGenes.ons76cbo.rds" and pull out cluster 9, and also get clusters 3 and 8 as separate clusters
```{r}
# get ONS76 cells in cluster 9 of ONS76-CBO using cell IDs
o76.cbo <- readRDS("outputs50K/ONS76-CBO/filtRiboMitoCellsGenes.ons76cbo.rds")
cl9.ons76Cbo <- subset(o76.cbo, cells = cl9cellID)

# get ONS76 cells in clusters 3 and 8 using subsetting based on cluster selection
Idents(phase.ons76cbo.sct.clust) <- "seurat_clusters"
cl3.8.ons76 <- subset(phase.ons76cbo.sct.clust, subset = (seurat_clusters %in% c(3, 8)))
dim(cl3.8.ons76)
```

```{r}
cl3.8.ons76$orig.ident <- "ONS76_clusts_3_8_coculture"
```

```{r}
head(cl3.8.ons76@meta.data)
```


```{r}
ons76mono <- readRDS("outputs50K/ONS76_2D/filt.NoRiboNoMito.ons76.2d.rds")
ons76spher <- readRDS("outputs50K/ONS76_SPHER/filt.NoRiboNoMito.ons76spher.rds")
```

Downsample the ONS76-2D and ONS76-spheroid samples so that they are around 400 cells each
```{r}
# get the cell IDs of all cells in DAOY-2D sample
o76.IDs.mono <- WhichCells(ons76mono)
length(o76.IDs.mono)
set.seed(162)
# pick a random sample of 400
o76.mono.small <- sample(o76.IDs.mono, size = 400, replace = FALSE)
```

```{r}
o76mono.400cells <- subset(ons76mono, cells = o76.mono.small)
dim(o76mono.400cells)
```

Now subsample the ONS76-spheroid sample
```{r}
# get the cell IDs of all cells in DAOY-spher sample
o76.IDs.spher <- WhichCells(ons76spher)
length(o76.IDs.spher)
set.seed(173)
# pick a random sample of 400
o76.spher.small <- sample(o76.IDs.spher, size = 400, replace = FALSE)
```

```{r}
o76spher.400cells <- subset(ons76spher, cells = o76.spher.small)
dim(o76spher.400cells)
```

```{r}
ons76.4samples.list2 <- list(o76mono.400cells, o76spher.400cells, cl9.ons76Cbo, cl3.8.ons76)
ons76.4samples.list2.sct <- suppressWarnings(lapply(X = ons76.4samples.list2, FUN = SCTransform))
features <- SelectIntegrationFeatures(object.list = ons76.4samples.list2.sct, nfeatures = 3000)
ons76.4samples.list2.sct <- PrepSCTIntegration(object.list = ons76.4samples.list2.sct, anchor.features = features)
```

```{r}
# k.filter = NA and k.weight = 77 as if there are fewer than ~ 200 cells in a sample, the below functions
# do not run and throw an error.
ons76.4samples.sct.anchors2 <- FindIntegrationAnchors(object.list = ons76.4samples.list2.sct, 
                                            normalization.method = "SCT",
                                            anchor.features = features,
                                            k.filter = NA)
ons76.4samples.combined.sct2 <- IntegrateData(anchorset = ons76.4samples.sct.anchors2, normalization.method = "SCT",
                                              k.weight = 77)
```

```{r}
ons76.4samples.combined.sct2 <- RunPCA(ons76.4samples.combined.sct2, verbose = FALSE)
ons76.4samples.combined.sct2 <- RunUMAP(ons76.4samples.combined.sct2, reduction = "pca", dims = 1:30)
ons76.4samples.combined.sct2 <- FindNeighbors(ons76.4samples.combined.sct2, 
                                           reduction = "pca", 
                                           dims = 1:30)
ons76.4samples.combined.sct2 <- FindClusters(ons76.4samples.combined.sct2, resolution = 0.2)
```

```{r}
# change the orig.ident notation for future ggplots, seurat plots
ons76.4samples.combined.sct2@meta.data[ons76.4samples.combined.sct2@meta.data == "ONS76-2D"] <- "ONS76_monolayer"
ons76.4samples.combined.sct2@meta.data[ons76.4samples.combined.sct2@meta.data == "ons76spher"] <- "ONS76_spheroid"
ons76.4samples.combined.sct2@meta.data[ons76.4samples.combined.sct2@meta.data == "ONS76-CBO"] <- "ONS76_clust9_coculture"

```

```{r}
saveRDS(ons76.4samples.combined.sct2,
        file = "outputs50K/INTEGRATE/ons76.clus9Cbo.clus3ANDclus8Cbo.Mono.Spher.4samples.DownsampledIntegrated.SCT.rds")
```

```{r}
# load the above object
ons76.4samples.combined.sct2 <- readRDS("outputs50K/INTEGRATE/ons76.clus9Cbo.clus3ANDclus8Cbo.Mono.Spher.4samples.DownsampledIntegrated.SCT.rds")
```

```{r}
table(ons76.4samples.combined.sct2$orig.ident)
```

```{r DimplotONS763samplesSCTintegrationMonoVsSperVsCluster9fromONS76CboSample, fig.height=3}
DimPlot(ons76.4samples.combined.sct2, reduction = "umap", group.by = "orig.ident")
```

SOX2 is marginally increased in the coculture condition, but overall is more evenly expressed at high level in the different ONS76 cell samples compared to expression in DAOY cells. POU3F2 is markedly increased in co-cultured ONS76 cells.
```{r VlnPlotSOX2networkErbb4ONS76CboClust9VsMonoVsSpher3samples, fig.height=10, fig.width=14}
Idents(ons76.4samples.combined.sct2) <- "orig.ident"
VlnPlot(ons76.4samples.combined.sct2, 
        features = c("SOX2", "POU3F2","FABP7", "TTYH1", "ERBB4", "NRG3"),
        slot = "data",
        assay = "RNA"
        )
```

```{r VlnPlotDcxMeis1Hes6ONS76CboClust9VsMonoVsSpher3samples, fig.height=9, fig.width=14}
Idents(ons76.4samples.combined.sct2) <- "orig.ident"
VlnPlot(ons76.4samples.combined.sct2, 
        features = c("NNAT","MIR124-2HG", "DCX", "CTNNA2", "MEIS1", "HES6"),
        slot = "data",
        assay = "RNA"
        )
```

MAPK3, PROX1 can regulate SOX2 expression. However, could be SMAD2/3/4/5 maintain higher levels of SOX2 in ONS76 cells vs DAOY cells. In addition, the inhibitory SMADs, SMAD6/7 are expressed at lower level than in DAOY cells. The targets of SMAD signaling SERPINE1 and ID1 are expressed at higher level than in DAOY cells, although it should be noted the regulation of these targets by SMADs is context-dependent (https://doi.org/10.1038/onc.2012.191)
```{r  VlnPlotHoxB2ShhSMADGenesONS76CboClust9VsMonoVsSpher3samples, fig.height=25, fig.width=12}

VlnPlot(ons76.3samples.combined.sct2, features = c("HOXB2","IRX3", "TWIST1",
                                        "PTCH1", "GLI1", "GLI2", "GLI3", "HHIP",
                                        "MYC", "MYCN", "ZIC1", "OTX2", "MAPK3","PROX1", "SMAD1", "SMAD2",
                                        "SMAD3", "SMAD4", "SMAD5", "SMAD6", "SMAD7", "SERPINE1", "ID1"))
```

-------------------------------------------------------------------------------------------------------
***Explore geneset enrichment of ONS76-2D and ONS76-spher using presto and gProfiler (web app)***
fgsea did not work with this dataset (multifgsea command crashed, not sure why, but gProfiler worked!)
from: https://nbisweden.github.io/workshop-scRNAseq/labs/compiled/seurat/seurat_05_dge.html#Gene_Set_Enrichment_Analysis_(GSEA) and https://crazyhottommy.github.io/scRNA-seq-workshop-Fall-2019/scRNAseq_workshop_3.html
Merge the following: 
1) ONS76-2D single cluster 
2) ONS76-spher cluster 0 
3) ONS76-spher cluster 1 
4) ONS76-spher cluster 2
The files used are:
"phase.ons76spher.SCT.clustered.res0.4.rds" and "phase.filtRiboMito.SCT.regressedCCdiff.ons762d.rds"
Find the cell IDs of the cluster 0,1,2,3 cells in "phase.ons76spher.SCT.clustered.res0.4.rds" and extract those cells
```{r}
phase.ons76spher.sct.clust.res0.4 <- 
readRDS("outputs50K/ONS76_SPHER/phase.ons76spher.SCT.clustered.res0.4.rds")
```

```{r}
# make Idents the 'seurat_clusters
Idents(phase.ons76spher.sct.clust.res0.4) <- "seurat_clusters"
```

```{r}
# extract the cell clusters
cl0cellID.ons76spher <- WhichCells(phase.ons76spher.sct.clust.res0.4, idents = "0")
cl1cellID.ons76spher <- WhichCells(phase.ons76spher.sct.clust.res0.4, idents = "1")
cl2cellID.ons76spher <- WhichCells(phase.ons76spher.sct.clust.res0.4, idents = "2")

length(cl0cellID.ons76spher)
length(cl1cellID.ons76spher)
length(cl2cellID.ons76spher)
```

For ONS76-spher: use the vector of cell IDs to subset the object "phase.ons76spher.sct.clust.res0.4" and pull out clusters 0 - 2 as separate clusters
```{r}
# get ONS76 cells in cluster 0 of ONS76-spher using cell IDs
clus0.ons76spher <- subset(phase.ons76spher.sct.clust.res0.4, cells = cl0cellID.ons76spher)
clus1.ons76spher <- subset(phase.ons76spher.sct.clust.res0.4, cells = cl1cellID.ons76spher)
clus2.ons76spher <- subset(phase.ons76spher.sct.clust.res0.4, cells = cl2cellID.ons76spher)
```

```{r}
# load the ONS76-2D sample (this is one big cluster of cells)
ons76mono.sct <- readRDS("outputs50K/ONS76_2D/phase.filtRiboMito.SCT.regressedCCdiff.ons762d.rds")
```

```{r}
# merge the above datasets
ons76.mono.spher.merged <- merge(ons76mono.sct,
                                 y = c(clus0.ons76spher, clus1.ons76spher, clus2.ons76spher),
                                 add.cell.ids = c("o76mono", "clus0.o76spher", 
                                                  "clus1.o76spher", "clus2.o76spher"),
                                 project = "ons76.mono.spher.merge")
```

```{r}
head(ons76.mono.spher.merged@meta.data)
```

```{r}
# assign cells to "seurat_clusters"
metadata <- ons76.mono.spher.merged@meta.data
metadata$cells <- rownames(metadata)
metadata$seurat_clusters[which(str_detect(metadata$cells, "^o76mono_"))] <- "mono"
metadata$seurat_clusters[which(str_detect(metadata$cells, "^clus0.o76spher_"))] <- "0"
metadata$seurat_clusters[which(str_detect(metadata$cells, "^clus1.o76spher_"))] <- "1"
metadata$seurat_clusters[which(str_detect(metadata$cells, "^clus2.o76spher_"))] <- "2"
head(metadata)
```

```{r}
ons76.mono.spher.merged@meta.data <- metadata
```

```{r}
ons76.mono.spher.genes <- wilcoxauc(ons76.mono.spher.merged, "seurat_clusters")
```

```{r}
head(ons76.mono.spher.genes)
```

```{r}
# we have all the genes for each cluster
dplyr::count(ons76.mono.spher.genes, group)
```

```{r}
# order the genes for input to gProfiler (web app!)
ons76.mono.UPgenes <- ons76.mono.spher.genes %>% 
  dplyr::filter(group == "mono" & logFC > 0.25) %>%
  arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)

nrow(ons76.mono.spher.UPgenes) # I used 1385 genes in g:Profiler
```

```{r}
# this will be used as input for UPregulated genes in g:Profiler. Transfer file via Filezilla to
# Excel on Mac
write.csv(ons76.mono.UPgenes, 
          "./outputs50K/INTEGRATE/ONS76.monoSpher.merged.monoUPgenes.PRESTO.gProfilerInput.csv", 
          quote = F)
```

```{r}
# order the genes for input to gProfiler (web app!)
ons76.spher.clus0.UPgenes <- ons76.mono.spher.genes %>% 
  dplyr::filter(group == "0" & logFC > 0.25) %>%
  arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)

nrow(ons76.spher.clus0.UPgenes) # I used 422 genes in g:Profiler
```

```{r}
# this will be used as input for UPregulated genes in g:Profiler web app. Transfer file via Filezilla to
# Excel on Mac
write.csv(ons76.spher.clus0.UPgenes, 
          "./outputs50K/INTEGRATE/ONS76.monoSpher.merged.cluster0spherUPgenes.PRESTO.gProfilerInput.csv", 
          quote = F)
```

```{r}
# order the genes for input to gProfiler (web app!)
ons76.spher.clus1.UPgenes <- ons76.mono.spher.genes %>% 
  dplyr::filter(group == "1" & logFC > 0.25) %>%
  arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)

nrow(ons76.spher.clus1.UPgenes) # I used 321 genes in g:Profiler
```

```{r}
# this will be used as input for UPregulated genes in g:Profiler web app. Transfer file via Filezilla to
# Excel on Mac
write.csv(ons76.spher.clus1.UPgenes, 
          "./outputs50K/INTEGRATE/ONS76.monoSpher.merged.cluster1spherUPgenes.PRESTO.gProfilerInput.csv", 
          quote = F)
```

```{r}
# order the genes for input to gProfiler (web app!)
ons76.spher.clus2.UPgenes <- ons76.mono.spher.genes %>% 
  dplyr::filter(group == "2" & logFC > 0.25) %>%
  arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)

nrow(ons76.spher.clus2.UPgenes) # I used 206 genes in g:Profiler
```

```{r}
# this will be used as input for UPregulated genes in g:Profiler web app. Transfer file via Filezilla to
# Excel on Mac
write.csv(ons76.spher.clus2.UPgenes, 
          "./outputs50K/INTEGRATE/ONS76.monoSpher.merged.cluster2spherUPgenes.PRESTO.gProfilerInput.csv", 
          quote = F)
```


------------------------------------------------------------------------------------------------------------
***Merge of downsampled 1. ONS76-2D, 2. ONS76-SPHEROID, 3. ONS76-only extracted from coculture, cluster 9 only and 4. clusters 3 and 8 combined, 5. DAOY-2D (rep), 6. DAOY-spheroid cells, 7. DAOY-only extracted from coculture***
To look at gene expression of SMADs, SOX2, notch, Shh, etc.. in a DoHeatmap of merged data
We use the cluster 9 ONS76 cells (extracted from coculture), clusters 3 and 8 ONS76 also extracted from co-culture, mono, spheroid, DAOY cells extracted from coculture, mono, spheroid. 7 samples total
DAOY-2D, DAOY-Spher, ONS76-2D, ONS76-Spher = 400 cells in each sample (downsampled)
DAOY-only extracted from coculture = 269 cells
Cell numbers in the ONS76-only cocultures.
Cancerous NEUROD1 expressing clusters in ONS76-CBO
Cluster 3 = 282 cells
Cluster 8 = 113 cells.
Total in cluster 3 AND 8 = 395 cells
```{r}
# get a random sample of 77 of the cell IDs of each of the ON76 and DAOY cell samples.
# objects for the ONS76 samples (cells)
# samples to merge for ONS76: o76mono.77cells, o76spher.77cells, cl9.ons76Cbo, cl3.8.ons76Cbo
# samples to merge for DAOY: 
```

```{r}
# create objects for the DAOY samples
dy2d <- readRDS("outputs50K/DAOY_2D/filt.CellsGenesRiboMito.daoy2dRep.seurat.rds")
dySpher <- readRDS("outputs50K/DAOY_SPHER/filt.RiboMitoCellsGenes.daoySpher.rds")
dyOnly.dyCbo <- readRDS("outputs50K/DAOY-CBO/dooyOnlyCells.subset.filt.noRiboMito.cboDaoy.rds")
```

Cell numbers in the ONS76-only cocultures.
Cancerous NEUROD1 expressing clusters in ONS76-CBO
Cluster 3 = 282 cells
Cluster 8 = 113 cells.
Total in cluster 3 AND 8 = 395 cells
```{r}
Idents(phase.ons76cbo.sct.clust) <- "seurat_clusters"
cl3.8.ons76cboCells <- subset(phase.ons76cbo.sct.clust, subset = (seurat_clusters %in% c(3, 8)))
```

Find the cell IDs of the cluster 3 and 8 cells in "phase.ons76cbo.sct.clust" and extract those cells
```{r}
cl3.8.cellID <- WhichCells(phase.ons76cbo.sct.clust, idents = c("3", "8"))
length(cl3.8.cellID)
```

Now use this vector of cell IDs to subset the object "filtRiboMitoCellsGenes.ons76cbo.rds"
```{r}
o76.cbo <- readRDS("outputs50K/ONS76-CBO/filtRiboMitoCellsGenes.ons76cbo.rds")
cl3.8.ons76Cbo <- subset(o76.cbo, cells = cl3.8.cellID)
```

```{r}
ons76mono <- readRDS("outputs50K/ONS76_2D/filt.NoRiboNoMito.ons76.2d.rds")
ons76spher <- readRDS("outputs50K/ONS76_SPHER/filt.NoRiboNoMito.ons76spher.rds")
```

```{r}
# get the cell IDs of all cells in DAOY-2D sample
dy2d.IDs <- WhichCells(dy2d)
length(dy2d.IDs)
set.seed(123)
# pick a random sample of 400 cell IDs
dy2d.small400 <- sample(dy2d.IDs, size = 400, replace = FALSE)
# Subset the seurat object for 400 cells
dy2d.400cells <- subset(dy2d, cells = dy2d.small400)
dim(dy2d.400cells)


# get the cell IDs of all cells in DAOY-spher sample
dySpher.IDs<- WhichCells(dySpher)
length(dySpher.IDs)
set.seed(134)
# pick a random sample of 400 cell IDs
dy3d.small400 <- sample(dySpher.IDs, size = 400, replace = FALSE)
# Subset the seurat object for 400 cells
dySpher.400cells <- subset(dySpher, cells = dy3d.small400)
dim(dySpher.400cells)

# use all cells in DAOY-CBO sample from which DAOY cells have been extracted 
dim(dyOnly.dyCbo)
```

```{r}
# get a sample of the ONS76 cells in each condition incl. cluster 9 from the ONS76-CBO condition

# get the cell IDs of all cells in ONS76-2D sample
o76.IDs.mono <- WhichCells(ons76mono)
length(o76.IDs.mono)
set.seed(162)
# pick a random sample of 400
o76.mono.small400 <- sample(o76.IDs.mono, size = 400, replace = FALSE)
o76mono.400cells <- subset(ons76mono, cells = o76.mono.small400)
dim(o76mono.400cells)

# get the cell IDs of all cells in ONS76-spher sample
o76.IDs.spher <- WhichCells(ons76spher)
length(o76.IDs.spher)
set.seed(173)
# pick a random sample of 400
o76.spher.small400 <- sample(o76.IDs.spher, size = 400, replace = FALSE)
o76spher.400cells <- subset(ons76spher, cells = o76.spher.small400)
dim(o76spher.400cells)

# choose all cells of cluster 9 of ONS76-CBO sample (n = 77 cells only)
dim(cl9.ons76Cbo)

# choose all cells of cluster 9 of ONS76-CBO sample (n = 77 cells only)
dim(cl3.8.ons76Cbo)

```

```{r}
# create the merged object of ONS76 cells and DAOY cells 
# (monolayer = 400 cells; spheroid = 400 cells, ONS76clus9 = 77 cells; ONS76clus3+8 = 395 cells)
daoy.o76.cells <- merge(o76mono.400cells, 
                  y = c(o76spher.400cells, cl3.8.ons76Cbo, cl9.ons76Cbo, 
                        dy2d.400cells, dySpher.400cells, dyOnly.dyCbo), 
                  add.cell.ids = c("o76mono", "o76spher", "o76clus38", "o76clus9", "dy2d", 
                                   "dySpher", "dyCocult"), 
                  project = "dy.ons.merge")
```

```{r}
head(daoy.o76.cells@meta.data)
```


```{r}
# this object is the correct size
dim(daoy.o76.cells)
```

```{r}
# Create metadata dataframe
metadata <- daoy.o76.cells@meta.data
# Add cell IDs to metadata
metadata$cells <- rownames(metadata)
# Create sample column
metadata$sample <- NA
metadata$sample[which(str_detect(metadata$cells, "^o76mono_"))] <- "ONS76_monolayer"
metadata$sample[which(str_detect(metadata$cells, "^o76spher_"))] <- "ONS76_spheroid"
metadata$sample[which(str_detect(metadata$cells, "^o76clus9_"))] <- "ONS76_cluster_9"
metadata$sample[which(str_detect(metadata$cells, "^o76clus38_"))] <- "ONS76_clusters_3_8"
metadata$sample[which(str_detect(metadata$cells, "^dy2d_"))] <- "DAOY_monolayer"
metadata$sample[which(str_detect(metadata$cells, "^dySpher_"))] <- "DAOY_spheroid"
metadata$sample[which(str_detect(metadata$cells, "^dyCocult_"))] <- "DAOY_coculture"
head(metadata)
```

```{r}
daoy.o76.cells@meta.data <- metadata
```

```{r}
daoy.o76.cells@meta.data %>% 
  filter(sample == "ONS76_clusters_3_8") %>% 
  head()
```


```{r}
daoy.o76.cells <- ScaleData(daoy.o76.cells)
```

```{r}
# change Idents from the default (which is cell IDs) to the sample
Idents(daoy.o76.cells) <- "sample"

# re-order the sample levels
my.levels <- c("ONS76_monolayer", "ONS76_spheroid", "ONS76_cluster_9", "ONS76_clusters_3_8",
               "DAOY_monolayer", "DAOY_spheroid", "DAOY_coculture")
levels(daoy.o76.cells) <- my.levels
levels(daoy.o76.cells)
```

```{r}
saveRDS(daoy.o76.cells, 
file = "./outputs50K/INTEGRATE/daoy.ons76.Merged.7samples.o76Mono.o76Spher.o76clus9.o76clus3AND8.dyMono.dySpher.dyOnlyCocult.rds")
```

```{r}
daoy.o76.cells <- 
  readRDS("outputs50K/INTEGRATE/daoy.ons76.Merged.7samples.o76Mono.o76Spher.o76clus9.o76clus3AND8.dyMono.dySpher.dyOnlyCocult.rds")
```


The results of VlnPlot are very similar in this merged object compared to the integrated object, with just ONS76 samples.
```{r VlnPlotSox2NesPou3F27samplesMergedDy2d3dcocultONS2d3dclust3AND8clust9SoxPou3F, fig.width=14}
VlnPlot(daoy.o76.cells,
        features = c("SOX2", "NES", "POU3F2", "MKI67", "TOP2A"),
        slot = "data",
        assay = "RNA")
```

```{r VlnPlotNeurod1Dcx7samplesMergedDy2d3dcocultONS2d3dclust3AND8clust9SoxPou3F, fig.width=14}
VlnPlot(daoy.o76.cells,
        features = c("IRX5","MIR124-2HG", "NEUROD1", "DCX", "CTNNA2"),
        slot = "data",
        assay = "RNA")
```


```{r VlnPlotSHHgenesOtx2Zic27samplesMergedDy2d3dcocultONS2d3dCocultClust3AND8AND9SOX2Network, fig.width=14}
VlnPlot(daoy.o76.cells,
        features = c("PTCH1", "OTX2", "ZIC2", "GLI1", "GLI2", "GLI3"),
        slot = "data",
        assay = "RNA")
```

Now view the same Violin Plots without the 'DAOY_coculture' condition by subsetting the seurat object
```{r}
table(daoy.o76.cells$sample)
```

```{r}
# THIS WAY OF SUBSETTING THE INTEGRATED OBJECT ALSO WORKS!
Idents(daoy.o76.cells) <- "sample"
dy.o76.small <- subset(daoy.o76.cells, 
                       subset = (sample %in% c("ONS76_monolayer", "ONS76_spheroid",
                                               "ONS76_cluster_9", "ONS76_clusters_3_8",
                                               "DAOY_monolayer", "DAOY_spheroid")))
```

```{r}
table(dy.o76.small$sample)
```

```{r}
dy.o76.small <- ScaleData(dy.o76.small)
```

```{r}
# re-order the sample levels
my.levels2 <- c("ONS76_monolayer", "ONS76_spheroid", "ONS76_cluster_9", "ONS76_clusters_3_8",
               "DAOY_monolayer", "DAOY_spheroid")
levels(dy.o76.small) <- my.levels2
levels(dy.o76.small)
```

```{r VlnPlot6samplesMergedDy2d3dONS2d3dCocultClust3AND8AND9SOX2Network, fig.width=14}
VlnPlot(dy.o76.small,
        features = c("SOX2", "POU3F2", "FABP7", "OTX2", "ZIC2", "GLI2"),
        slot = "data",
        assay = "RNA")
```


SOX2 network. POU3F2 is expressed in the ONS76 and DAOY co-cultured cells. The below heatmap is not that useful, earlier violin plots are more useful.
```{r Heatmap7samplesMergedDy2d3dcocultONS2d3dclust3AND8clust9Sox2Network, fig.height=10, fig.width=12}
sox.genes <- c("SOX2", "POU3F2", "FABP7")
DoHeatmap(daoy.o76.cells,
          features = sox.genes,
          group.by = levels("sample"),
          slot = "scale.data",
          assay = "RNA") +
  theme(axis.text = element_text(size = 12))
```

It is obvious that TGF-B/SMAD signalling is most active in the DAOY 2D-cultures which has the LOWEST Sox2 expression.
```{r Heatmap7samplesMergedDy2d3dcocultONS2d3dclust3AND8clust9SmadGenes, fig.height=5}
# choose the genes to display in the heatmap
# note that ERBB1 is not present in daoy.o76.cells@assays$RNA@counts! To check this, simply convert
# daoy.o76.cells@assays$RNA@counts to a data.frame (command is: 'as.data.frame(daoy.o76.cells@assays$RNA@counts)'),
# then find the gene of interest using, e.g. dplyr 'filter(gene == "ERBB1")' command.
smad.genes <- c("SOX2", "SMAD1", "SMAD2", "SMAD3", "SMAD4", "SMAD5", "SMAD6", "SMAD7",
                    "SERPINE1", "ID1")

# re-ordering sample order only works if group.by argument is used below!!
DoHeatmap(daoy.o76.cells,
          features = smad.genes,
          group.by = levels("sample"),
          slot = "scale.data",
          assay = "RNA") +
  guides(color = "none") +
  theme(axis.text = element_text(size = 15))
```

Investigate SHH signalling in the samples > multiple readouts of SHH signalling are expressed in DAOY-monolayer especially, but SHH itself is no more strongly expressed in these cells than in other samples. This therefore reflects cell-intrinsic overactivation of SHH signalling in DAOY-monolayer cells in particular.
```{r Heatmap7samplesMergedDy2d3dcocultONS2d3dclust3AND8clust9SHHgenes, fig.height=4}
shh.genes <- c("GLI1", "GLI2", "GLI3", "PTCH1", "PTCH2", "HHIP", "SHH")
DoHeatmap(daoy.o76.cells,
          features = shh.genes,
          group.by = levels("sample"),
          slot = "scale.data",
          assay = "RNA") +
  guides(color = "none") +
  theme(axis.text = element_text(size = 12))
```

Investigate Notch-delta signalling in all samples > of most interest is the expression of HES4, HES5 and HES6 in the ONS76 (cluster 9) or DAOY cells extracted from CBO coculture
```{r Heatmap7samplesMergedDy2d3dcocultONS2d3dclust3AND8clust9Notchgenes, fig.height=4}
notch.genes <- c("SOX2", "HEY1", "HEY2", "HEYL", "DEC1",
                  "HES1", "HES4", "HES6",  "HES2",
              "HES5", "HES7","DLL1", "DLL3", "DLL4", "JAG1", "JAG2")
DoHeatmap(daoy.o76.cells,
          features = notch.genes,
          group.by = levels("sample"),
          slot = "scale.data",
          assay = "RNA") +
  guides(color = "none") +
  theme(axis.text = element_text(size = 12))
```


Investigate Neuregulin signalling in all samples
```{r Heatmap7samplesMergedDy2d3dcocultONS2d3dclust3AND8clust9notchGenes, fig.height=4}
neureg.genes <- c("SOX2", "ERBB2", "ERBB3", "ERBB4", "NRG1", "NRG2", "NRG3", "NRG4")
DoHeatmap(daoy.o76.cells,
          features = neureg.genes,
          group.by = levels("sample"),
          slot = "scale.data",
          assay = "RNA") +
  theme(axis.text = element_text(size = 12))
```

```{r}
save.image("./RDataFiles/INTEGRATE_ONS76-CBO_ONS76-2D_ONS76-SPHER_CBO_seurat_50K.RData")
```


```{r}
sessionInfo()
```


Additional code not run but might be useful in future
#--------------------------------------------------------------------------------------------------
```
ons76.spher.cbo.clust <- subset(ons76.all.combined.sct.clust, subset = (orig.ident == "ONS76-CBO" | orig.ident == "ons76spher"))
```

```
# THIS WAY OF SUBSETTING THE INTEGRATED OBJECT ALSO WORKS!
subset(ons76.all.combined.sct.clust, subset = (orig.ident %in% c("ONS76-CBO", "ons76spher")))
```

Getting Idents
```
table(Idents(ons76.spher.cbo.clust))
```

How to load CSV files
```
# in general, no need to load the csv files. Here one is loaded just to remind me what arguments
# to use
clus0.ons76.cocult.vs.spher.all <- read.csv(
  "outputs50K/INTEGRATE/clus0.ons76.cocult.vs.spher.allDEgenes.csv",
         header = TRUE,
         stringsAsFactors = FALSE)
```
How to write CSV files
```
# reminder of how to use write.csv() function
write.csv(clus1.ons76.cocult.vs.spher.pos, 
          file = "./outputs50K/INTEGRATE/clus1.ons76.cocult.vs.spher.posDEgenes.csv", 
          quote = F) 
```

How to subset a Seurat object
```
ons76.all.combined.sct.clust <- NormalizeData(object = ons76.all.combined.sct.clust, assay = "RNA")
DefaultAssay(ons76.all.combined.sct.clust) <- "RNA"
ons76.cbo.2d.clust <- subset(ons76.all.combined.sct.clust, subset = (orig.ident == "ONS76-CBO" | orig.ident == "ONS76-2D"))
```


------------------------------------------------------------------------------------------------
***Use upregulated genes in ONS76 spheroids vs 2D as query list for AddModuleScore() to identify cells that are ONS76 cells in the ONS76-CBO sample***
Unfortunately this approach did not work!
Get lists of upregulated genes for each of the 4 clusters in ONS76-SPHER vs ONS76-2D - restrict the list to the top 30 in each cluster. The idea is to make lists of UP-regulated genes in each cluster of ONS76-SPHER, and then try searching for cells in the co-culture condition (ONS76-CBO) that express those 'modules'. If found, these will be ONS76 cancer cells. Use dplyr select() to get just the genes. Doing this does not work, because for one module (clus0), virtually all cells in the ONS76-CBO sample express the module at high level, but for other modules much fewer cells express it.

-----------------------------------------------------------------------------------------------
***Integrated analysis of the ONS76-CBO sample, ONS76-SPHEROID and CBO samples - conventional normalization***
The idea is to see whether ONS76 cells in the co-culture sample cluster with other cancer cells in the ONS76-spheroid sample. CONVENTIONAL NORMALIZATION was used. I used RECIPROCAL PCA instead of CCA as is commonly used when integrating datasets that are biologically different. This worked and showed that SOME ONS76-CBO cells cluster with the ONS876-spheroid cells and other ONS76-CBO cells resembling rhombic lip/GCP progenitors cluster with a separate cluster of ONS76-spheroid cells that might be more differentiated. Let's pull out these ONS76-CBO cells that are in cluster 0 and cluster 7

----------------------------------------------------------------------------------------------------------
***Subclustering of conventional integrated seurat object (cluster 7)***
Able to do this, but did not prove helpful for the analysis
Cluster 7 in the conventional integration workflow (seurat object: ons76.integrated) is an interesting cluster. Try to subcluster this cluster, and then do differential expression. However, it would be good to know the sample origins of the subclusters of cluster 7. This information should be in meta.data (orig.ident).
```
# get the graph names for subclustering
names(ons76.integrated@graphs)
```
The subclustering works. res=0.5 or res=0.4 does not make a difference to number of subclusters.
```
ons76.integrated <- FindSubCluster(ons76.integrated, 
                            cluster = "7",
                            graph.name = "integrated_nn",
                            subcluster.name = "GCP",
                            resolution = 0.5,
                            algorithm = 1)          # '1' is the original Louvain algorithm
```

I ran this code to check the column names of meta.data after 'subcluster.name=GCP' was added to
```
head(ons76.integrated@meta.data, n = 20)
```

```
# filter the rows that have seurat_clusters == 7
metadata.clus7 <- ons76.integrated@meta.data %>% 
  filter(seurat_clusters == 7)

head(metadata.clus7)
```

Do a ggplot of orig.ident vs GCP category. We can see that all three subclusters 7_0 to 7_2 are represented in all conditions (CBO, ONS76-CBO and ons76spher). So I can dismiss the hypothesis that the subclusters are specific to one condition only. However, subcluster 7_2 accounts for around 50% of ONS76-SPHER sample, whereas this subcluster forms a smaller proportion of the other two conditions. Not possible to draw firm conclusion as to whether these differences are meaningful as the number of cells in GCP cluster for ONS76-SPHER is only 63.
```
ggplot(data = metadata.clus7, aes(x = orig.ident, fill = GCP)) +
         geom_bar()
```

Visualise the subclusters on a UMAP plot. Notice the 'group.by' argument is set to 'GCP' which is the subclustered cluster 7.
```
DimPlot(ons76.integrated, reduction = "umap", group.by = "GCP", label = TRUE)
```

Get violin plots of NEUROD1 clusters in the integrated seurat object. This shows that compared to the analysis of a single sample and its control (ONS76-CBO vs CBO) the integration has resulted in reduced complexity of the NEUROD1 expressing cell subtype. In the INTEGRATED workflow there are two NEUROD1 clusters (cluster 3 and 9) which have cells from both CBO and ONS76-CBO. So it will not be helpful to compare DEGs between cluster 3 and 9.
```
VlnPlot(ons76.integrated, 
        features = c("NEUROD1", "CNTN1", "CNTN2", "GRIA4", "RRM2", "ACTC1"))
```

***Subclustering of conventional integrated seurat object (cluster 3 and cluster 9)***
Let's try subclustering cluster 3 first - this is the main NEUROD1+ cluster. We'll see if there are cells which distinguish wild-type cerebellar neurons from the ONS76-CBO sample. This is the bigger of two NEUROD1+ clusters.
```
ons76.integrated <- FindSubCluster(ons76.integrated, 
                            cluster = "3",
                            graph.name = "integrated_nn",
                            subcluster.name = "neurod1_3",
                            resolution = 0.4,
                            algorithm = 1)          # '1' is the original Louvain algorithm
```

The subclusters are present in both conditions, ONS76-CBO and the CBO control (none in spheroid). There are more ONS76-CBO NeuroD1+ cells in cluster 3 than in cluster 3 of CBO sample.
```
# filter the rows that have seurat_clusters == 3
metadata.clus3 <- ons76.integrated@meta.data %>% 
  filter(seurat_clusters == 3)

ggplot(data = metadata.clus3, aes(x = orig.ident, fill = neurod1_3)) +
         geom_bar()
```

Same result for cluster 9 as for cluster 3. Three subclusters found and all three are present in both conditions! There are more ONS76-CBO NeuroD1+ cells than CBO NeuroD1+ cells in cluster 9. The integrated object used here might not be the best way to look at this cluster as the cells are aligned and differences between cells are reduced compared to when individual samples are analysed (e.g. ONS76-CBO compared to CBO control).
```
ons76.integrated <- FindSubCluster(ons76.integrated, 
                            cluster = "9",
                            graph.name = "integrated_nn",
                            subcluster.name = "neurod1_9",
                            resolution = 0.4,
                            algorithm = 1)          # '1' is the original Louvain algorithm

# filter the rows that have seurat_clusters == 9
metadata.clus9 <- ons76.integrated@meta.data %>%
  filter(seurat_clusters == 9)

ggplot(data = metadata.clus9, aes(x = orig.ident, fill = neurod1_9)) +
         geom_bar()

```

-------------------------------------------------------------------------------------------------------
***Integrate the NEUROD1+ clusters in ONS76-CBO and CBO samples after SCT normalisation***
This approach did not work.
Instead of above approach, extract all the individual NeuroD1+ cells from the CBO cluster, and from the ONS76-CBO cluster. CBO cluster has 1 NeuroD1+ cluster and the ONS76-CBO cluster has 3 NeuroD1+ clusters


-------------------------------------------------------------------------------------------------------
***Integrate (by RPCA) the NEUROD1+ clusters in ONS76-CBO and CBO samples after conventional normalization***
This too doesn't work
First run conventional normalization on CBO sample at res=0.5 (higher res gives 2 NeuroD1 clusters!). There is only 1 NeuroD1+ cluster!

---------------------------------------------------------------------------------------------------
This code is for illustration only to show the workflow
***Running SCTransform based integration***
``` SCtransform_CBO_ONS76-CBO
# prepare the SCTransformed seurat objects
# the seurat object has NO mitochondrial and NO ribosomal genes
# 'outputs50K" folder contains the correct RDS seurat object
# when running this code the warning' "Warning in theta.ml(y = y, mu = fit$fitted) : iteration limit reached" will appear regularly. See this Stack Exchange answer (https://bioinformatics.stackexchange.com/questions/16439/sctransform-warning-in-theta-mly-y-mu-fitfitted-iteration-limit-reach) for the meaning. This behaviour of the code is expected.

filt.noRiboMito.ons76.2d <- readRDS("outputs50K/ONS76_2D/filt.NoRiboNoMito.ons76.2d.rds")

filt.noRiboMito.ons76spher <- readRDS("outputs50K/ONS76_SPHER/filt.NoRiboNoMito.ons76spher.rds")

ons76.cells.cocult <- readRDS("outputs50K/ONS76-CBO/ons76.cells.only.rds")

ons76.all.list <- list(filt.noRiboMito.ons76.2d, filt.noRiboMito.ons76spher, ons76.cells.cocult)

ons76.all.list <- suppressWarnings(lapply(X = ons76.all.list, FUN = SCTransform))
```

```
features <- SelectIntegrationFeatures(object.list = ons76.all.list, nfeatures = 3000)

ons76.all.list <- PrepSCTIntegration(object.list = ons76.all.list, anchor.features = features)
```

```
ons76.all.anchors <- FindIntegrationAnchors(object.list = ons76.all.list, normalization.method = "SCT",
    anchor.features = features)
ons76.all.combined.sct <- IntegrateData(anchorset = ons76.all.anchors, normalization.method = "SCT")
```

```
# save the combined (integrated) seurat object, which includes: ONS76-2D, ONS76-SPHER, and ONS76 cells extracted
# from ONS76-CBO sample.
# note this integrated seurat object has NOT been through PCA or UMAP commands

saveRDS(ons76.all.combined.sct, file = "./outputs50K/INTEGRATE/ons76.all.combined.sct.rds")
```

```
# here we read in the ons76 cells from monolayer, spheroid, and ons76-CBO coculture
ons76.all.combined.sct <- readRDS("outputs50K/INTEGRATE/ons76.all.combined.sct.rds")
```

```
ons76.all.combined.sct <- RunPCA(ons76.all.combined.sct, verbose = FALSE)
ons76.all.combined.sct <- RunUMAP(ons76.all.combined.sct, reduction = "pca", dims = 1:30)
```

```
ons76.all.combined.sct <- FindNeighbors(ons76.all.combined.sct, 
                                           reduction = "pca", 
                                           dims = 1:30)
```

```clustree, fig.height=8
# Select a range of resolutions
resolution.range <- seq(from = 0, to = 1, by = 0.2)

# Find clusters using a range of resolutions, the clustering information is in the metadata slot
ons76.all.combined.sct.clusters <- FindClusters(object = ons76.all.combined.sct, 
                                             resolution = resolution.range,
                                             verbose = FALSE)

# apply Clustree: this suggests there are 3 or possibly 4 clusters (4th cluster is smallest)
clustree(ons76.all.combined.sct.clusters)
```

```r findClusters_res0.2_res0.4

ons76.all.clusters.res0.2 <- FindClusters(ons76.all.combined.sct, resolution = 0.2)

ons76.all.clusters.res0.4 <- FindClusters(ons76.all.combined.sct, resolution = 0.4)
```

```DimplotsRes0.2Integrated3samples, fig.height=12, fig.width=8
# With res=0.2, there are 3 clusters in ONS76-2D and ONS76-CBO, but 4 clusters in ons76spher. 
# The 4th cluster in ons76spher is small.
# THE INTEGRATION OF 3 CONDTIONS: ONS76-2D, ONS76-SPHER AND ONS76 CELLS IN COCULTURE HAS WORKED VERY WELL -
      # there is great overlap of the UMAP clusters from the three conditions, which is probably 
      # because this type of integration forces dissimilar cells to be more similar!
p1 <- DimPlot(ons76.all.clusters.res0.2, reduction = "umap", group.by = "orig.ident")
p2 <- DimPlot(ons76.all.clusters.res0.2, reduction = "umap", split.by = "orig.ident")
p3 <- DimPlot(ons76.all.clusters.res0.2, reduction = "umap", label = TRUE, repel = TRUE)

p1 / p2 / p3
```

-----------------------------------------------------------------------------------------------------
***Perform integration of ONS76-2D and ONS76-SPHER, then PCA/UMAP to compare the samples***
This workflow makes the cells too similar in terms of gene expression. The same cells grown under different conditions (spheroids or monolayer cells) end up being intermingled, which is not useful
```
