---
title: "ONS76-CBO_Seurat_EXPLORE1_50K"
author: "JJ"
date: "11/08/2022"
output: html_document
---

```{r graphical_output}
options(bitmapType='cairo-png') # this line of code needed BEFORE knitr code in chunk below
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = "png",
                      dpi = 300,
                      echo = TRUE,   # prints the code (if FALSE then only code output is shown)
                      cache = TRUE,
                      fig.path = "./outputs50K/ONS76-CBO/figures/")
```

```{r libraries}
# library(reshape2) for formatting long table
suppressPackageStartupMessages({
library(Seurat)
library(glmGamPoi)
library(RColorBrewer)
library(fgsea)
library(msigdbr)
library(presto)
library(Matrix)
library(heatmap3)
library(patchwork)
library(tidyverse)
library(plyr)
})
```

```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK
load("RDataFiles/ONS76-CBO_Seurat_EXPLORE1_50K.RData")
```

```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK
# load full filtered (genes and cells) seurat object if needed

filt.cboONS76.seurat <- readRDS("outputs50K/ONS76-CBO/filtCellsGenes.ons76cbo.seurat.rds")
```

------------------------------------------------------------------------------------------
***Cluster the 'phase' object of ONS76-CBO sample and find NEUROD1 clusters (SCTransform)***
Repeat the clustering of ONS76-CBO that has undergone regression of the cell cycle difference.

```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK
# load the object - this object is a bit different from the ClusterMap markers object
phase.ons76cbo.sct <- readRDS("outputs50K/ONS76-CBO/phase.filtRiboMito.ons76cbo.SCT.regressedCCdiff.rds")
```

```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK
# run PCA and UMAP
phase.ons76cbo.sct <- phase.ons76cbo.sct %>% 
  RunPCA(verbose = FALSE) %>% 
  RunUMAP(dims=1:30, verbose=FALSE) %>% 
  FindNeighbors(dims=1:30, verbose=FALSE) %>% 
  FindClusters(verbose=FALSE)
```

```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK
# save the clustered 'phase.ons76cbo.sct' object
saveRDS(phase.ons76cbo.sct, file = "outputs50K/ONS76-CBO/phase.ons76cbo.sct.CCdiffRegressed.filtRiboMito.clustered.rds")
```


```{r}
# load the CLUSTERED 'phase.ons76cbo.sct' object
phase.ons76cbo.sct.clust <- 
  readRDS("outputs50K/ONS76-CBO/phase.ons76cbo.sct.CCdiffRegressed.filtRiboMito.clustered.rds")
```

```{r DimplotONS76cboSCTphaseCCdiffRegressed, fig.height=6, fig.width=10}
DimPlot(phase.ons76cbo.sct.clust, label = TRUE) # '+ NoLegend()' is an option to insert here
```


```{r ElbowPlotONS76cboPhaseSCTclust}
# elbow plot shows that there is an elbow at PC ~10, but the graph does not level off after
# that showing there is some additional information to ~ PC 40.
# The input object is the CLUSTERED saved seurat object

ElbowPlot(phase.ons76cbo.sct.clust, ndims = 50, reduction = "pca")
```

There are 2545 cells and the breakdown by cluster is shown
```{r}
# cluster 3 = 282 cells
# cluster 8 = 113 cells
# cluster 9 = 77 cells
cell.num <- table(phase.ons76cbo.sct.clust@active.ident)
cell.num
sum(cell.num)
```

-------------------------------------------------------------------------------------------
***Get the differentially expressed genes in the 11 ONS76-CBO clusters using a FOR loop***
Getting cluster markers this way is less efficient than using FindAllMarkers() command!
```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK

Idents(phase.ons76cbo.sct.clust) <- "seurat_clusters"
markers.pos <- data.frame()
for (i in levels(Idents(phase.ons76cbo.sct.clust))){
  markers <- FindMarkers(phase.ons76cbo.sct.clust, ident.1 = i, grouping.var = "ident", 
                         verbose = FALSE, only.pos = TRUE, logfc.threshold = 0.25)
  dat <- markers %>% rownames_to_column(var = "genes")
  dat$cluster_ID <- i
  dat$cluster_ID <- as.numeric(dat$cluster_ID)
  df <- data.frame(dat)
  markers.pos <- rbind(markers.pos, df)
}
head(markers.pos)
```

```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK
# save the postive markers of ONS76-CBO clusters generated from SCTransformed, CC difference,
# ribosomal and mitochondrial gene filtered 'phase.ons76cbo.sct.CCdiffRegressed.filtRiboMito.clustered.rds'
saveRDS(markers.pos, file = "outputs50K/ONS76-CBO/pos.markers.phase.filtRiboMito.CCdiffRegress.SCT.rds")
```

```{r}
# load the positive markers of clusters of ONS76-CBO
markers.pos <- 
  readRDS("outputs50K/ONS76-CBO/pos.markers.phase.filtRiboMito.CCdiffRegress.SCT.rds")
```

```{r}
table(markers.pos$cluster_ID)
```

```{r}
dim(markers.pos)
```

Check out the first few DEGs by group (cluster_ID) of 'markers.pos'
```{r}
markers.pos %>% 
  group_by(cluster_ID) %>% 
  slice_head(n = 3)
```

***CLUSTER ANNOTATION***
```{r}
# continue to replace the cluster numbers with the identity of each cluster
new.cluster.ids <- c("RP-choroid plexus", "glio-neuronal/purkinje cells", "choroid plexus", 
                     "ONS76_cluster_3", "granule precursors", "uncertain", "RL/DCN",
                     "oligodendrocytes", "ONS76_cluster_8", "ONS76_cluster_9",
                     "progenitors")

names(new.cluster.ids) <- levels(phase.ons76cbo.sct.clust)
phase.ons76cbo.sct.clust.2 <- RenameIdents(phase.ons76cbo.sct.clust, new.cluster.ids)
```

```{r, DimplotONS76CboCellAnnot, fig.height=6, fig.width=10}
DimPlot(phase.ons76cbo.sct.clust.2, reduction = "umap", label = TRUE, repel = TRUE, pt.size = 0.5)
```

```{r}
# Create metadata dataframe
metadata <- phase.ons76cbo.sct.clust.2@meta.data
# Add cell IDs to metadata
metadata$cells <- rownames(metadata)
# Create sample column
metadata$celltype <- NA
metadata$celltype[which(metadata$seurat_clusters == "0")] <- "RP-choroid plexus"
metadata$celltype[which(metadata$seurat_clusters == "1")] <- "glio-neuronal/purkinje cells"
metadata$celltype[which(metadata$seurat_clusters == "2")] <- "choroid plexus"
metadata$celltype[which(metadata$seurat_clusters == "3")] <- "ONS76_cluster_3"
metadata$celltype[which(metadata$seurat_clusters == "4")] <- "granule precursors"
metadata$celltype[which(metadata$seurat_clusters == "5")] <- "uncertain"
metadata$celltype[which(metadata$seurat_clusters == "6")] <- "RL/DCN"
metadata$celltype[which(metadata$seurat_clusters == "7")] <- "oligodendrocytes"
metadata$celltype[which(metadata$seurat_clusters == "8")] <- "ONS76_cluster_8"
metadata$celltype[which(metadata$seurat_clusters == "9")] <- "ONS76_cluster_9"
metadata$celltype[which(metadata$seurat_clusters == "10")] <- "progenitors"
```

```{r}
phase.ons76cbo.sct.clust.2@meta.data <- metadata
```

```{r}
head(phase.ons76cbo.sct.clust.2@meta.data)
```

```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK
saveRDS(phase.ons76cbo.sct.clust.2, 
  file = "./outputs50K/ONS76-CBO/phase.filtRiboMito.ons76Cbo.CellAnnot.SCT.CCdiffRegressed.clustered.rds")
```


--------------------------------------------------------------------------------------------------------
***Transcriptomic comparison of annotated DAOY-CBO and CBO datasets at CLUSTER level***
code from: https://github.com/quadbiolab/scRNAseq_analysis_vignette/blob/master/Tutorial.md
using Spearman correlation

```{r}
# load ONS76-CBO object
# load the phase.ONS76cbo.sct.clust.2 object (this has additional cell type annotation
# in metadata)
phase.ons76cbo.sct.clust.2 <- 
  readRDS("outputs50K/ONS76-CBO/phase.filtRiboMito.ons76Cbo.CellAnnot.SCT.CCdiffRegressed.clustered.rds")
```

```{r}
# load CBO object
# load the 'phase.cbo.sct.clusters.2' object
phase.cbo.sct.clusters.2 <- 
  readRDS("outputs50K/CBO/phase.filtRiboMito.cbo.CellAnnot.SCT.CCdiffRegressed.clustered.rds")
```


****Spearman correlation to compare clusters in ONS76-CBO and CBO control****
KNITTING THESE CODE CHUNKS FAILS!!
```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK
av.exp.cbo <- sapply(sort(unique(phase.cbo.sct.clusters.2$celltype)), function(ct) rowMeans(phase.cbo.sct.clusters.2$RNA@data[,which(phase.cbo.sct.clusters.2$celltype == ct)] ))
```

```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK
avg.exp.ons76Cbo <- sapply(levels(phase.ons76cbo.sct.clust.2@active.ident), function(ct) rowMeans(phase.ons76cbo.sct.clust.2$RNA@data[,which(phase.ons76cbo.sct.clust.2@active.ident == ct)]))
```

```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK
genes2cor <- intersect(VariableFeatures(phase.cbo.sct.clusters.2), 
                       rownames(phase.ons76cbo.sct.clust.2))
corr2ons76.cbo.cl <- cor(avg.exp.ons76Cbo[genes2cor,], av.exp.cbo[genes2cor,], method = "spearman")
```

```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK
corr2ons76.cbo.cl
```

My cell annotations based on the positive DEGs in each cluster turned out to be accurate
```{r HeatmapONS76CboVSCboClusterCorrelations, fig.height=4, fig.width=4}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK

 heatmap3(corr2ons76.cbo.cl, scale="none", margins=c(15,17),
          labRow = colnames(avg.exp.ons76Cbo), labCol = colnames(av.exp.cbo), cexRow=0.8, cexCol=0.8,
          col=colorRampPalette(rev(c("#b2182b","#d6604d","#f4a582","#fddbc7","#f7f7f7","#d1e5f0","#92c5de","#4393c3","#2166ac")))(30))

```


-----------------------------------------------------------------------------------------------------------
***Identification of healthy and cancerous NEUROD1 clusters in ONS76-CBO sample generated by SCTransform***
The Violin Plot shows that the NEUROD1 clustering using SCTransform() is consistent with the results of conventional normalisation and identifies clusters 3,4,8 as the NEUROD1 clusters. Also the other markers allow us to distinguish the healthy from cancerous cells. There are three NEUROD1 clusters in the ONS76-CBO sample, but only one in the CBO sample. The ONS76-CBO sample should contain both healthy and cancerous NEUROD1 cells.
Which ones are the NEUROD1 clusters? In the CBO sample, clustering shows there is only 1 NEUROD1 cluster and this cluster also expresses RBFOX3 and RTN1 (derived from  https://doi.org/10.7554/eLife.37551). We can see that the only cluster that expresses all three genes, as in the NEUROD1 cluster in CBO sample, is cluster 4. So cluster 4 NEUROD1+ cells must be the healthy cells and the other two clusters are medulloblastoma clusters that have differentiated. The SCTransform better distinguishes the healthy from the cancerous NEUROD1+ clusters because cluster 7, the cancer cluster, also expresses RBFOX3 in conventional normalization, but NOT in SCTransform normalization, so SCTransform is superior. 
```{r ViolinPlotNeurod1clustersONS76cboSCT, fig.width=14}
# only cluster 4 expresses NEUROD1, RBFOX3 and RTN1 and must be wild type cells

VlnPlot(phase.ons76cbo.sct.clust, features = c("NEUROD1", "RBFOX3", "RTN1", "NES"))
```

Extract the cells/clusters that are NeuroD1+. There are 634 NEUROD1 positive cells - both healthy and cancerous
```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK
neurod1.ons76cbo <- subset(phase.ons76cbo.sct.clust, idents = c("3", "4", "8"))
dim(neurod1.ons76cbo)
```

Find out how many healthy cells there are, and how many cancerous ones.
```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK
# the size of healthy cluster of NEUROD1 cells = 239 cells
subset(phase.ons76cbo.sct.clust, idents = c("4")) %>% 
  dim()

# the size of cancerous cluster of NEUROD1 cells = 395 cells
subset(phase.ons76cbo.sct.clust, idents = c("3", "8")) %>% 
  dim()
```

```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK

# save the neurod1.ons76cbo clusters 3,4,8 above - this was generated from the phase CC difference
# regressed SCTransformed seurat object

saveRDS(neurod1.ons76cbo, file = "./outputs50K/ONS76-CBO/neurod1ClustsAll.phaseCCDiffRegress.SCT.ons76cbo.rds")
```

```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK

# save the neurod1.ons76cbo clusters 3, and 8 only (ONS76 CANCER CELLS) - 
# this was generated from the phase CC difference regressed SCTransformed seurat object
neurod1.cancer.ons76cbo <- subset(phase.ons76cbo.sct, idents = c("3", "8"))
saveRDS(neurod1.cancer.ons76cbo, 
        file = "./outputs50K/ONS76-CBO/neurod1ClustsCancer.phaseCCDiffRegress.SCT.ons76cbo.rds")
```

--------------------------------------------------------------------------------------------------
***Clustering the CBO sample to find the NEUROD1+ clusters***
Before the correlation matrix can be generated, we need to load the SCTransformed, CC difference regressed seurat objects and isolate the individual NEUROD1+ clusters from ONS76-CBO and CBO samples
```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK
# simply load the clustered seurat object
# load the data by running PCA, UMAP, etc on 'phase.filtRiboMito.ons76cbo.SCT.regressedCCdiff.rds'
# which is saved in the outputs50K folder and can be accessed by readRDS above. I ran the code again
# in the section 'Cluster the 'phase' object and find NEUROD1 clusters' above.

# load the single CBO control sample (CCdifference regressed out!)
phase.cbo.sct <- readRDS("outputs50K/CBO/phase.cbo.filtRiboMito.SCT.CCdiffRegress.rds")

# run PCA, UMAP etc on 'phase.cbo.sct' object from the CBO sample
phase.cbo.sct <- phase.cbo.sct %>% 
  RunPCA(verbose = FALSE) %>% 
  RunUMAP(dims=1:30, verbose=FALSE) %>% 
  FindNeighbors(dims=1:30, verbose=FALSE) %>% 
  FindClusters(verbose=FALSE)
```

```{r}
# load the clustered phase CBO object with CCdifference regressed, filtered for ribosomal and
# mitochondrial genes, and SCTransformed
phase.cbo.sct.clust <- 
  readRDS("outputs50K/CBO/phase.filtRiboMito.cbo.SCT.CCdiffRegressed.clustered.rds")
```


```{r VlnPlotNeurod1RBFX3RTN1PhaseCBOsct, fig.width=14}
# CBO sample expression of NEUROD1, etc.
VlnPlot(phase.cbo.sct.clust, features = c("NEUROD1", "RBFOX3", "RTN1", "NES"))
```

```{r}
# there are 271 NEUROD1+ cells in cluster 5 of CBO control 
# sample (phase, SCTransformed, CCdifference, regressed)
nd1.clus5.cbo <- subset(phase.cbo.sct.clust, idents = c("5"))
dim(nd1.clus5.cbo)
```


------------------------------------------------------------------------------------------
***Gene expression in ONS76 cancer clusters 3, 8, 9***

```{r ViolinPlotSox2networkONS76cbo, fig.width=14}
# plot of the clusters containing granule/RL lineage - clusters 2, 4, 6, 9 and 10 contain RL/granule
# lineage.
# CDH1 is expressed (in WT cells and ONS76 cells), but CDH5 (marker of ONS76 cells) is not expressed.
# CD44 is not expressed or expressed very little.
# ITGB1 is also not good at discriminating WT CBO from ONS76-CBO samples

VlnPlot(phase.ons76cbo.sct.clust, 
        features = c("SOX2", "POU3F2", "FABP7", "OTX2", "HOXB2"))
```

```{r ViolinPlotERRB4nrg3Meis1ZicONS76cbo, fig.width=14}
VlnPlot(phase.ons76cbo.sct.clust, 
        features = c("ERBB4", "NRG3", "MEIS1", "ZIC1", "ZIC2"))
```

```{r ViolinPlotDcxOtherGenesONS76cbo, fig.width=14}
VlnPlot(phase.ons76cbo.sct.clust, 
        features = c("DCX", "GRIA4", "TTYH1",  "IRX5", "IRX3"))
```

```{r ViolinPlotOtherGenesONS76cbo, fig.width=14}
VlnPlot(phase.ons76cbo.sct.clust, 
        features = c("MIR124-2HG", "NNAT", "CTNNA2", "LHX9"))
```

```{r ViolinPlotSHHgenesMycONS76cbo, fig.width=14}
VlnPlot(phase.ons76cbo.sct.clust, 
        features = c("GLI2", "GLI3", "PTCH1", "MYC", "MYCN", "TWIST1"))
```


-------------------------------------------------------------------------------------------------
***Compare gene expression (SOX2 stemness network) in the different ONS76 clusters 3,8,9***
```{r}
# subset the ONS76 clusters
clust9 <- subset(phase.ons76cbo.sct.clust, idents = c("9"))
clust8 <- subset(phase.ons76cbo.sct.clust, idents = c("8"))
clust3 <- subset(phase.ons76cbo.sct.clust, idents = c("3"))
```

```{r}
o76.3clusters <- merge(clust9, 
                  y = c(clust8, clust3), 
                  add.cell.ids = c("o76_clus9", "o76_clus8", "o76_clus3"), 
                  project = "o76.3clustersMerge")
o76.3clusters
```

```{r}
# Create metadata dataframe
metadata <- o76.3clusters@meta.data
# Add cell IDs to metadata
metadata$cells <- rownames(metadata)
# Create sample column
metadata$sample <- NA
metadata$sample[which(str_detect(metadata$cells, "^o76_clus9_"))] <- "cluster_9"
metadata$sample[which(str_detect(metadata$cells, "^o76_clus8_"))] <- "cluster_8"
metadata$sample[which(str_detect(metadata$cells, "^o76_clus3_"))] <- "cluster_3"
```

```{r}
o76.3clusters@meta.data <- metadata
```

```{r}
table(o76.3clusters$sample)
```

```{r}
o76.3clusters <- ScaleData(o76.3clusters)
```

```{r}
Idents(o76.3clusters) <- "sample"
```

The Sox2 network in ONS76 cell clusters grown in co-culture is smaller than in DAOY cells in the same condition
```{r ViolinPlotSOX2GenesONS76cbo, fig.height=6, fig.width=10}
VlnPlot(o76.3clusters, features = c("SOX2", "POU3F2", "FABP7", "OTX2", "ZIC2", "GLI2"))
```


---------------------------------------------------------------------------------------------------------
***Gene set enrichment analysis (GSEA) using FGSEA and PRESTO for malignant clusters of ONS76-CBO sample***
from: https://nbisweden.github.io/workshop-scRNAseq/labs/compiled/seurat/seurat_05_dge.html#Gene_Set_Enrichment_Analysis_(GSEA) and https://crazyhottommy.github.io/scRNA-seq-workshop-Fall-2019/scRNAseq_workshop_3.html
PRESTO is good package to use, ranks genes very fast for Gene Set Enrichment analysis

The first website uses the output of Seurat FindMarkers() to generate the gene list, but this does not use all genes. Note that the arguments are different from the standard ones used to generate DEG list.

Get the genes for control CBO sample cluster 5, which I have called granule precursors first and compare the ONS76-CBO sample to the cells in cluster 5 CBO.
```{r}
# apply presto to CBO control
cbo.genes.presto <- wilcoxauc(phase.cbo.sct.clusters.2, 'seurat_clusters')
head(cbo.genes.presto)
```

```{r}
# apply presto to ONS76-CBO sample
ons76cbo.genes.presto <- wilcoxauc(phase.ons76cbo.sct.clust.2, 'seurat_clusters')
head(ons76cbo.genes.presto)
```

```{r}
# we have all the genes for each cluster
dplyr::count(cbo.genes.presto, group)
```

```{r}
# we have all the genes for each cluster
dplyr::count(ons76cbo.genes.presto, group)
```


****Cluster 5 CBO control (granule neurons) - GO:Biological Process****
Test the gene set enrichment of cluster 5 in CBO, the granule cell cluster
```{r}
# continue with preparing data for input to fgsea and getting bar charts of gene set 
# enrichment analysis
clust5.cbo.genes.presto <- cbo.genes.presto %>%
  dplyr::filter(group == "5" & !(feature == "MALAT1")) %>%
  dplyr::arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)
```

```{r}
ranks.clus5 <- deframe(clust5.cbo.genes.presto)
head(ranks.clus5)
```

```{r}
GOBP.geneset <- msigdbr(species = "Homo sapiens", category = "C5", subcategory = "GO:BP")
head(GOBP.geneset)
```

```{r}
fgsea_GOBPsets <- GOBP.geneset %>% split(x = .$gene_symbol, f = .$gs_name)
```

```{r}
# set eps to zero to get lower p-values
fgseaRes.GOBP.clus5 <- 
  fgseaMultilevel(fgsea_GOBPsets, eps = 0, stats = ranks.clus5, minSize = 10, scoreType = "pos", maxSize = 300)
```

```{r}
# top GO terms are related to proliferation not neuronal differentiation! Consistent with this
# granule cluster being composed of granule precursors
fgseaResTidy.GOBP.clus5 <- fgseaRes.GOBP.clus5 %>%
  as_tibble() %>%
  dplyr::arrange(desc(NES))

fgseaResTidy.GOBP.clus5 %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  dplyr::arrange(padj) %>% 
  head(10)
```

The GO:BP enrichments clearly show these granule 'neurons' are unlikely to be terminally differentiated and are still proliferative. This is why cluster 3 in ONS76-CBO (and perhaps cluster 8 in ONS76-CBO) have a gene expression profile of proliferative cells and that is correlated (Spearman correlation) with the gene expression profile of cluster 5 granule cells in the CBO control.

```{r BarplotFGSEAmsigdbGOBPclust5cbo, fig.height=5, fig.width=6}
# only plot the top 20 pathways, but note the 'padj' filter which will limit the number of plotted
# genesets to the top few

ggplot(fgseaResTidy.GOBP.clus5 %>% filter(padj < 0.001) %>% head(n=20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill = NES > 0)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="GO:Biological Pathways NES from GSEA") + 
  theme_minimal()
```

****Cluster 5 CBO control (granule precursors) - HALLMARK genesets****
```{r}
H.geneset <- msigdbr(species = "Homo sapiens", category = "H")
head(H.geneset)
```

```{r}
fgsea_Hsets <- H.geneset %>% split(x = .$gene_symbol, f = .$gs_name)
```

```{r}
# set eps to zero to get lower p-values
fgseaRes.H.clus5 <- 
  fgseaMultilevel(fgsea_Hsets, eps = 0, stats = ranks.clus5, minSize = 10, scoreType = "pos", maxSize = 300)
```

```{r}
fgseaResTidy.H.clus5 <- fgseaRes.H.clus5 %>%
  as_tibble() %>%
  dplyr::arrange(desc(NES))

fgseaResTidy.H.clus5 %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  dplyr::arrange(padj) %>% 
  head(10)
```


```{r BarplotFGSEAmsigdbHallmarkClust5cbo, fig.height=1, fig.width=4}
# only plot the top 20 pathways, but note the 'padj' filter which will limit the number of plotted
# genesets to the top few

ggplot(fgseaResTidy.H.clus5 %>% filter(padj < 0.001) %>% head(n=20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill = NES > 0)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="HALLMARK Pathways NES from GSEA") + 
  theme_minimal()
```

****Cluster 5 CBO control (granule neurons) - REACTOME****
```{r}
React.geneset <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:REACTOME")
head(React.geneset)
```

```{r}
fgsea_ReactomeSets <- React.geneset %>% split(x = .$gene_symbol, f = .$gs_name)
```

```{r}
# set eps to zero to get lower p-values
fgseaRes.Reactome.clus5 <- 
  fgseaMultilevel(fgsea_ReactomeSets, 
                  eps = 0, stats = ranks.clus5, 
                  minSize = 10, maxSize = 300,
                  scoreType = "pos")
```

```{r}
# top pathway hits for REACTOME
fgseaResTidy.reactome.clus5 <- fgseaRes.Reactome.clus5 %>%
  as_tibble() %>%
  arrange(desc(NES))

fgseaResTidy.reactome.clus5 %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  arrange(padj) %>% 
  head(20)
```

```{r BarplotFGSEAmsigdbREACTOMEclust5cbo, fig.height=4, fig.width=6}
ggplot(fgseaResTidy.reactome.clus5 %>% filter(padj < 0.001) %>% head(n=20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill = NES > 0)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="REACTOME Pathways NES from GSEA") + 
  theme_minimal()
```


****Cluster 3 ONS76-CBO - geneset enrichment (FGSEA) including input for g:Profiler and Cytoscape****
```{r}
# filter gene expression for cluster 3, the ONS76 cluster
# the presto gene list puts EYA2 and NEUROD1 near the top which is reassuring
ons76cbo.genes.presto %>%
  dplyr::filter(group == "3") %>%
  arrange(desc(logFC), desc(auc)) %>%
  head(10)
```

```{r}
# this code used to generate sorted gene list as input for gProfiler (web app)
ons76cbo.cl3.UPgenes.presto <- ons76cbo.genes.presto %>% 
  dplyr::filter(group == "3" & logFC > 0.25) %>%
  dplyr::arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)

nrow(ons76cbo.cl3.UPgenes.presto) # nrow=977, I used 977 genes in g:Profiler
```

```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK
# save as csv for input to g:Profiler (web app)
write.csv(ons76cbo.cl3.UPgenes.presto, 
          "./outputs50K/ONS76-CBO/ONS76cbo.cluster3.977UPgenes.PRESTO.gProfilerInput.csv", 
          quote = F)
```

```{r}
# FOR KNIT DO NOT RE-RUN THIS CODE CHUNK
# load the above file: this will be used as input for UPregulated genes in g:Profiler
ons76cbo.cl3.UPgenes.presto <- read.csv("outputs50K/ONS76-CBO/ONS76cbo.cluster3.UPgenes.PRESTO.gProfilerInput.csv", 
                                       header = TRUE, 
                                       stringsAsFactors = FALSE)
```

```{r}
# continue with preparing data for input to fgsea and getting bar charts of gene set 
# enrichment analysis
clust3.genes.presto <- ons76cbo.genes.presto %>%
  dplyr::filter(group == "3") %>%
  arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)
```

```{r}
ranks.clus3 <- deframe(clust3.genes.presto)
head(ranks.clus3)
```

****Cluster 3 ONS76-CBO - GO:Biological Process****
```{r}
GOBP.geneset <- msigdbr(species = "Homo sapiens", category = "C5", subcategory = "GO:BP")
head(GOBP.geneset)
```

```{r}
fgsea_GOBPsets <- GOBP.geneset %>% split(x = .$gene_symbol, f = .$gs_name)
```

```{r}
# set eps to zero to get lower p-values
fgseaRes.GOBP.clus3 <- 
  fgseaMultilevel(fgsea_GOBPsets, eps = 0, stats = ranks.clus3, minSize = 10, scoreType = "pos", maxSize = 300)
```

```{r}
# top GO terms are related to proliferation
fgseaResTidy.GOBP.clus3 <- fgseaRes.GOBP.clus3 %>%
  as_tibble() %>%
  dplyr::arrange(desc(NES))

fgseaResTidy.GOBP.clus3 %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  dplyr::arrange(padj) %>% 
  head(10)
```

```{r BarplotFGSEAmsigdbGOBPclust3ons76cbo, fig.height=5, fig.width=6}
# only plot the top 20 pathways, but note the 'padj' filter which will limit the number of plotted
# genesets to the top few

ggplot(fgseaResTidy.GOBP.clus3 %>% filter(padj < 0.001) %>% head(n=20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill = NES > 0)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="GO:Biological Pathways NES from GSEA") + 
  theme_minimal()
```

****Cluster 3 - HALLMARK genesets****
The Hallmark genesets are enriched for 'mitotic spindle', MYC targets, E2F targets - i.e. all cell cycle related - and this is quite similar to the healthy granule cluster, cluster 4 which could explain why the correlation heatmap and the UMAP plots show cluster 4 next to cluster 3.
```{r}
H.geneset <- msigdbr(species = "Homo sapiens", category = "H")
head(H.geneset)
```

```{r}
fgsea_Hsets <- H.geneset %>% split(x = .$gene_symbol, f = .$gs_name)
```

```{r}
# set eps to zero to get lower p-values
fgseaRes.H.clus3 <- 
  fgseaMultilevel(fgsea_Hsets, eps = 0, stats = ranks.clus3, minSize = 10, scoreType = "pos", maxSize = 300)
```

```{r}
fgseaResTidy.H.clus3 <- fgseaRes.H.clus3 %>%
  as_tibble() %>%
  dplyr::arrange(desc(NES))

fgseaResTidy.H.clus3 %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  dplyr::arrange(padj) %>% 
  head(50)
```


```{r BarplotFGSEAmsigdbHallmarkClust3ons76cbo, fig.height=5, fig.width=6}
# only plot the top 20 pathways, but note the 'padj' filter which will limit the number of plotted
# genesets to the top few

ggplot(fgseaResTidy.H.clus3 %>% filter(padj < 0.001) %>% head(n=20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill = NES > 0)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="HALLMARK Pathways NES from GSEA") + 
  theme_minimal()
```

****Cluster 3 - Reactome gene sets****
```{r}
React.geneset <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:REACTOME")
head(React.geneset)
```

```{r}
fgsea_ReactomeSets <- React.geneset %>% split(x = .$gene_symbol, f = .$gs_name)
```

```{r}
# set eps to zero to get lower p-values
fgseaRes.Reactome.clus3 <- 
  fgseaMultilevel(fgsea_ReactomeSets, eps = 0, stats = ranks.clus3, minSize = 15, scoreType = "pos", maxSize = 200)
```

```{r}
# cell cycle related genesets are enriched which is consistent with GO:BP
fgseaResTidy.reactome.clus3 <- fgseaRes.Reactome.clus3 %>%
  as_tibble() %>%
  arrange(desc(NES))

fgseaResTidy.reactome.clus3 %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  arrange(padj) %>% 
  head(10)
```

```{r BarplotFGSEAmsigdbREACTOMEclust3ons76cbo, fig.width=6}
# only plot the top 20 pathways, but note the 'padj' filter which will limit the number of plotted
# genesets to the top few. The padj is < 0.05 which is higher than padj for HALLMARK set which
# is set to padj < 0.001

ggplot(fgseaResTidy.reactome.clus3 %>% filter(padj < 0.05) %>% head(n=20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill = NES > 0)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="REACTOME Pathways NES from GSEA") + 
  theme_minimal()
```

***Cluster 8 ONS76-CBO - geneset enrichment***
Minimal genesets enriched!
```{r}
# filter gene expression for cluster 8, the ONS76 cluster
# the presto gene list puts EYA2 and NEUROD1 near the top which is reassuring
ons76cbo.genes.presto %>%
  dplyr::filter(group == "8") %>%
  arrange(desc(logFC), desc(auc)) %>%
  head(10)
```

```{r}
ons76cbo.cl8.UPgenes.presto <- ons76cbo.genes.presto %>% 
  dplyr::filter(group == "8" & logFC > 0.25) %>%
  dplyr::arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)

nrow(ons76cbo.cl8.UPgenes.presto) # I used only 127 genes in g:Profiler
```

```{r}
# this will be used as input for UPregulated genes in g:Profiler. Transfer file via Filezilla to
# Excel on Mac
write.csv(ons76cbo.cl8.UPgenes.presto, 
          "./outputs50K/ONS76-CBO/ONS76cbo.cluster8.127UPgenes.PRESTO.gProfilerInput.csv", 
          quote = F)
```

```{r}
# continue with preparing data for input to fgsea and getting bar charts of gene set 
# enrichment analysis
clust8.genes.presto <- ons76cbo.genes.presto %>%
  dplyr::filter(group == "8") %>%
  arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)
```

```{r}
ranks.clus8 <- deframe(clust8.genes.presto)
head(ranks.clus8)
```

****Cluster 8 - GO:Biological Process****
```{r}
GOBP.geneset <- msigdbr(species = "Homo sapiens", category = "C5", subcategory = "GO:BP")
head(GOBP.geneset)
```

```{r}
fgsea_GOBPsets <- GOBP.geneset %>% split(x = .$gene_symbol, f = .$gs_name)
```

```{r}
# set eps to zero to get lower p-values
fgseaRes.GOBP.clus8 <- 
  fgseaMultilevel(fgsea_GOBPsets, eps = 0, stats = ranks.clus8, minSize = 15, scoreType = "pos", maxSize = 200)
```

```{r}
# no significantly enriched GO terms!
fgseaResTidy.GOBP.clus8 <- fgseaRes.GOBP.clus8 %>%
  as_tibble() %>%
  arrange(desc(NES))

fgseaResTidy.GOBP.clus8 %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  arrange(padj) %>% 
  head(10)
```

There are no statistically significant pathways (i.e. p_adj < 0.05)! This is even when I make logFC > 0 and use more than 6000 genes.
```{r BarplotFGSEAmsigdbGOBPclust3ons76cbo, fig.width=6}
# only plot the top 20 pathways, but note the 'padj' filter which will limit the number of plotted
# genesets to the top few, also padj is < 0.05 compared to padj < 0.001 for cluster 3

ggplot(fgseaResTidy.GOBP.clus8 %>% filter(padj < 0.05) %>% head(n=20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill = NES > 0)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="GO:Biological Pathways NES from GSEA") + 
  theme_minimal()
```


****Cluster 8 - Reactome geneset****

```{r}
React.geneset <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:REACTOME")
head(React.geneset)
```

```{r}
fgsea_ReactomeSets <- React.geneset %>% split(x = .$gene_symbol, f = .$gs_name)
```

```{r}
# set eps to zero to get lower p-values
fgseaRes.Reactome.clus8 <- 
  fgseaMultilevel(fgsea_ReactomeSets, eps = 0, stats = ranks.clus8, minSize = 15, scoreType = "pos", maxSize = 200)
```

```{r}
# cell cycle related genesets are enriched which is consistent with GO:BP
fgseaResTidy.reactome.clus8 <- fgseaRes.Reactome.clus8 %>%
  as_tibble() %>%
  arrange(desc(NES))

fgseaResTidy.reactome.clus8 %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  arrange(padj) %>% 
  head(10)
```

```{r BarplotFGSEAmsigdbREACTOMEclust8ons76cbo}
# only plot the top 20 pathways, but note the 'padj' filter which will limit the number of plotted
# genesets to the top few. padj is < 0.05 which is higher padj value than used for cluster 3 where
# padj < 0.001

ggplot(fgseaResTidy.reactome.clus8 %>% filter(padj < 0.05) %>% head(n=20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill = NES > 0)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="REACTOME Pathways NES from GSEA") + 
  theme_minimal()
```

****Cluster 8 - HALLMARK geneset****
```{r}
H.geneset <- msigdbr(species = "Homo sapiens", category = "H")
head(H.geneset)
```

```{r}
fgsea_Hsets <- H.geneset %>% split(x = .$gene_symbol, f = .$gs_name)
```

```{r}
# set eps to zero to get lower p-values
fgseaRes.H.clus8 <- 
  fgseaMultilevel(fgsea_Hsets, eps = 0, stats = ranks.clus8, minSize = 15, scoreType = "pos", maxSize = 200)
```

No significant pathways (ie with p_adj < 0.05)
```{r}

fgseaResTidy.H.clus8 <- fgseaRes.H.clus8 %>%
  as_tibble() %>%
  arrange(desc(NES))

fgseaResTidy.H.clus8 %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  arrange(padj) %>% 
  head(10)
```

***Cluster 9 ONS76-CBO - geneset enrichment including input gene list for g:Profiler***
```{r}
ons76cbo.cl9.UPgenes.presto <- ons76cbo.genes.presto %>% 
  dplyr::filter(group == "9" & logFC > 0.25) %>%
  dplyr::arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)

nrow(ons76cbo.cl9.UPgenes.presto) # nrow=471, so I used 471 genes in g:Profiler
```

```{r}
# this will be used as input for UPregulated genes in g:Profiler. Transfer file via Filezilla to
# Excel on Mac
write.csv(ons76cbo.cl9.UPgenes.presto, 
          "./outputs50K/ONS76-CBO/ONS76cbo.cluster9.UPgenes.PRESTO.gProfilerInput.csv", 
          quote = F)
```


```{r}
# continue with preparing data for input to fgsea and getting bar charts of gene set 
# enrichment analysis
clust9.genes.presto <- ons76cbo.genes.presto %>%
  dplyr::filter(group == "9") %>%
  arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)
```

```{r}
ranks.clus9 <- deframe(clust9.genes.presto)
head(ranks.clus9)
```

****Cluster 9 - GO:Biological Process****
```{r}
GOBP.geneset <- msigdbr(species = "Homo sapiens", category = "C5", subcategory = "GO:BP")
head(GOBP.geneset)
```

```{r}
fgsea_GOBPsets <- GOBP.geneset %>% split(x = .$gene_symbol, f = .$gs_name)
```

```{r}
# set eps to zero to get lower p-values
fgseaRes.GOBP.clus9 <- 
  fgseaMultilevel(fgsea_GOBPsets, eps = 0, stats = ranks.clus9, minSize = 10, scoreType = "pos", maxSize = 300)
```

```{r}
fgseaResTidy.GOBP.clus9 <- fgseaRes.GOBP.clus9 %>%
  as_tibble() %>%
  dplyr::arrange(desc(NES))

fgseaResTidy.GOBP.clus9 %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  dplyr::arrange(padj) %>% 
  head(10)
```

```{r BarplotFGSEAmsigdbGOBPclust9ons76cbo, fig.width=6}
# only plot the top 20 pathways, but note the 'padj' filter which will limit the number of plotted
# genesets to the top few

ggplot(fgseaResTidy.GOBP.clus9 %>% filter(padj < 0.001) %>% head(n=20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill = NES > 0)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="GO:Biological Pathways NES from GSEA") + 
  theme_minimal()
```

****Cluster 9 - Reactome geneset****

```{r}
# load the Reactome genesets
React.geneset <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:REACTOME")
head(React.geneset)
```

```{r}
fgsea_ReactomeSets <- React.geneset %>% split(x = .$gene_symbol, f = .$gs_name)
```

```{r}
# set eps to zero to get lower p-values
fgseaRes.Reactome.clus9 <- 
  fgseaMultilevel(fgsea_ReactomeSets, eps = 0, stats = ranks.clus9, minSize = 10, scoreType = "pos", maxSize = 300)
```

```{r}
# cell cycle related genesets are enriched which is consistent with GO:BP
fgseaResTidy.reactome.clus9 <- fgseaRes.Reactome.clus9 %>%
  as_tibble() %>%
  arrange(desc(NES))

fgseaResTidy.reactome.clus9 %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  arrange(padj) %>% 
  head(10)
```

```{r BarplotFGSEAmsigdbREACTOMEclust9ons76cbo, fig.height=5, fig.width=6}
# only plot the top 20 pathways, but note the 'padj' filter which will limit the number of plotted
# genesets to the top few

ggplot(fgseaResTidy.reactome.clus9 %>% filter(padj < 0.001) %>% head(n=20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill = NES > 0)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="REACTOME Pathways NES from GSEA") + 
  theme_minimal()
```


***Comparison of ONS76 clusters in co-culture to the healthy granule cell cluster, cluster 4 in ONS76-CBO sample***
Struck by the annotations in gProfiler and FGSEA for cluster 3 which is enriched for cell cycle, yet the correlation heatmap shows it matches quite well to cluster 4, the healthy granule cluster. On checking the geneset enrichment for cluster 5 of CBO this now makes sense - healthy granule cell cluster in CBO and ONS76-CBO are precursors not differentiated neurons

```{r}
# preparing data for input to fgsea and getting bar charts of gene set 
# enrichment analysis
clust4.genes.presto <- ons76cbo.genes.presto %>%
  dplyr::filter(group == "4") %>%
  arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)
```

```{r}
ranks.clus4 <- deframe(clust4.genes.presto)
head(ranks.clus4)
```

****Cluster 4 - GO:Biological Process****
```{r}
GOBP.geneset <- msigdbr(species = "Homo sapiens", category = "C5", subcategory = "GO:BP")
head(GOBP.geneset)
```

```{r}
fgsea_GOBPsets <- GOBP.geneset %>% split(x = .$gene_symbol, f = .$gs_name)
```

```{r}
# set eps to zero to get lower p-values
fgseaRes.GOBP.clus4 <- 
  fgseaMultilevel(fgsea_GOBPsets, eps = 0, stats = ranks.clus4, minSize = 10, scoreType = "pos", maxSize = 300)
```

```{r}
# top GO terms are related to proliferation
fgseaResTidy.GOBP.clus4 <- fgseaRes.GOBP.clus4 %>%
  as_tibble() %>%
  dplyr::arrange(desc(NES))

fgseaResTidy.GOBP.clus4 %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  dplyr::arrange(padj) %>% 
  head(50)
```

```{r BarplotFGSEAmsigdbGOBPclust3ons76cbo, fig.height=5, fig.width=6}
# only plot the top 20 pathways, but note the 'padj' filter which will limit the number of plotted
# genesets to the top few

ggplot(fgseaResTidy.GOBP.clus4 %>% filter(padj < 0.001) %>% head(n=20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill = NES > 0)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="GO:Biological Pathways NES from GSEA") + 
  theme_minimal()
```

****CLUSTER 4 - HALLMARK Geneset****
Using HALLMARK genesets there is an enrichment for 'mitotic spindle' at the top of the list, which is different from GO:BP! The 'mitotic spindle' geneset has 199 genes and is the largest geneset found.
```{r}
H.geneset <- msigdbr(species = "Homo sapiens", category = "H")
head(H.geneset)
```

```{r}
fgsea_Hsets <- H.geneset %>% split(x = .$gene_symbol, f = .$gs_name)
```

```{r}
# set eps to zero to get lower p-values
fgseaRes.H.clus4 <- 
  fgseaMultilevel(fgsea_Hsets, eps = 0, stats = ranks.clus4, minSize = 10, scoreType = "pos", maxSize = 300)
```


```{r}

fgseaResTidy.H.clus4 <- fgseaRes.H.clus4 %>%
  as_tibble() %>%
  dplyr::arrange(desc(NES))

fgseaResTidy.H.clus4 %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  dplyr::arrange(padj) %>% 
  head(10)
```

```{r BarplotFGSEAmsigdbHallmarkClust4ons76cbo, fig.height=1.5, fig.width=4}
# only plot the top 20 pathways, but note the 'padj' filter (set to < 0.05 in this case) 
# which will limit the number of plotted genesets to the top few

ggplot(fgseaResTidy.H.clus4 %>% filter(padj < 0.05) %>% head(n=20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill = NES > 0)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="HALLMARK Pathways NES from GSEA") + 
  theme_minimal()
```


--------------------------------------------------------------------------------------------------
***Examine cell cycle and proliferating cells using AddModuleScore()***
Find the cells that are actively proliferating in this sample
```
{r}
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
prolif.genes <- list(c(s.genes, g2m.genes))
```

```
{r AddModuleScoreProlifernONS76cbo, fig.height=4}
phase.ons76cbo.sct.clust <- AddModuleScore(phase.ons76cbo.sct.clust,
                        features = prolif.genes,
                        name = "prolif.genes")
FeaturePlot(phase.ons76cbo.sct.clust, features = 'prolif.genes1', label = TRUE, repel = TRUE) +
   scale_colour_gradientn(colours = rev(brewer.pal(n = 11, name = "RdBu")))
```

```
save.image("./RDataFiles/ONS76-CBO_Seurat_EXPLORE1_50K.RData")
```


UNUSED CODE RETAINED IN CASE IT TURNS OUT TO BE USEFUL
-----------------------------------------------------------------------------------
***Expression of specific marker genes of each cluster in phase SCTransformed ONS76-CBO seurat object***
This type of plot was not that useful
```
{r DotplotAllSeppMarkersAndCiliaONS76cbo, fig.height=10}
# from the Sepp paper repeat Dotplot with TF and some key non-TF markers
# added markers of ciliated cells from cluster10 above.
# used the argument col.min = 0 so that average expression scale is from 0 to 2.5 in key

sepp.all.markers <- c("TFAP2A", "PAX3", "RFX4", "TCFL71", "TCF7L2", "GLIS3", "SOX6", "HOPX", "TSC22D4",  
                      "ID4", "NCKAP5", "SLIT2", "CLYBL", "SLC1A3", "SLC6A11", "AQP4",  #astroglial/progenitor
                      
                      "ESRRB", "FOXP1", "FOXP2", "FOXP4", "MEF2C", "RORB", "CDH9", "ETV1", "EBF1", "EBF2", 
                      "SKOR2", "DAB1", "IPTR",   # Purkinje
                      
                      "NOTCH1",    # VZ neuroblast
             
                      "PAX2", "NFIA", "NFIB", "TFAP2B", "BHLHE22", "PRDM8",  # interneuron
                      
                      "ZFHX3", "ZFHX4",   # GABAergic deep nuclei neurons
             
                      "TOP2A", "ATOH1", "LMX1A", "OTX2","ETV6", "PAX6", "BARHL1", "E2F3", "ST18", "SATB2", 
                      "ZBTB18", "NEUROD1", "ZIC1", "ZIC4", "GABAR6", "CNTN1", "CNTN2", "KLF9", "NR3C1", "LMO3", 
                      "SYNPR", "KCNIP4", "HCRTR2", 
                      "EOMES", "TRPC3", "GRM1")
                      # UBC/GCP/granule/glut deep cerebellar nuclei                                                                   
seppAll.cilia.stem.markers <- c(sepp.all.markers, "CCNO", "CFAP299", "ROPN1L", "C11orf88", "ANKRD66", "C20orf85", 
                                "STOML3", "TEKT4", "SNTN", "MEIS1", "POU3F2", "HOXB2", "MEIS2", "GRIA4")

DotPlot(object = phase.ons76cbo.sct.clust, features = seppAll.cilia.stem.markers, col.min = 0) +
  coord_flip()
```


# code below is in case I need to check expression of specific gene(s) in certain clusters.
```
# CHECK ABOVE CANDIDATE GENES WHICH ARE UPREGULATED IN ONS76 CELLS VS WT CELLS IN CO-CULTURE
#----------------------------------------------------------------------------------------------------

cntn2.clusters.2.10 <- ons76cbo.cluster.markers.noRiboMito[((ons76cbo.cluster.markers.noRiboMito$cluster == 2 | ons76cbo.cluster.markers.noRiboMito$cluster == 10) & ons76cbo.cluster.markers.noRiboMito$gene == "CNTN2"), ]

cntn2.clusters.2.10
```
# code below is not needed as Venn diagrams are not the correct way to compare gene expression in cells from different clusters. However, it is retained in case Venn Diagrams of expressed genes are useful in future.
***Venn diagrams for overlap of upregulated genes***

```
cluster2.markers <- ons76cbo.cluster.markers.noRiboMito[ons76cbo.cluster.markers.noRiboMito$cluster == 2, ]
cluster6.markers <- ons76cbo.cluster.markers.noRiboMito[ons76cbo.cluster.markers.noRiboMito$cluster == 6, ]
cluster9.markers <- ons76cbo.cluster.markers.noRiboMito[ons76cbo.cluster.markers.noRiboMito$cluster == 9, ]
cluster10.markers <- ons76cbo.cluster.markers.noRiboMito[ons76cbo.cluster.markers.noRiboMito$cluster == 10, ]
```

```
write.csv(cluster2.markers, file = "./outputs50K/DEG_VennDiag/ONS76cbo_cluster2.csv")
write.csv(cluster6.markers, file = "./outputs50K/DEG_VennDiag/ONS76cbo_cluster6.csv")
write.csv(cluster9.markers, file = "./outputs50K/DEG_VennDiag/ONS76cbo_cluster9.csv")
write.csv(cluster10.markers, file = "./outputs50K/DEG_VennDiag/ONS76cbo_cluster10.csv")
```

```
ONS76cbo_clus2 <- read.csv("outputs50K/DEG_VennDiag/ONS76cbo_cluster2.csv",
                           header = TRUE,
                           stringsAsFactors = FALSE)
ONS76cbo_clus6 <- read.csv("outputs50K/DEG_VennDiag/ONS76cbo_cluster6.csv",
                           header = TRUE,
                           stringsAsFactors = FALSE)
ONS76cbo_clus9 <- read.csv("outputs50K/DEG_VennDiag/ONS76cbo_cluster9.csv",
                           header = TRUE,
                           stringsAsFactors = FALSE)
ONS76cbo_clus10 <- read.csv("outputs50K/DEG_VennDiag/ONS76cbo_cluster10.csv",
                           header = TRUE,
                           stringsAsFactors = FALSE)
cbo_clus6_granule <- read.csv("outputs50K/DEG_VennDiag/cbo_cluster6_granule.csv",
                           header = TRUE,
                           stringsAsFactors = FALSE)
```

``` VennDiagClus6CBOvsClus6ONS76CBO
gene_list1 <- list(cbo_clus6_granule$gene, ONS76cbo_clus6$gene)
cbo.clus6.granule_ons76cbo.clus6 <- ggVennDiagram(gene_list1, category.names = c("cbo_clus6_granule", "ONS76cbo_clus6"))
cbo.clus6.granule_ons76cbo.clus6
```

```VennDiagClus6CBOvsClus2ONS76CBO
# cluster 2 in ONS76-CBO and cluster 6 in CBO seem to share the greatest percentage of genes,
# so it is likely that CBO cluster 6 and ONS76-CBO cluster 2 are the most similar

gene_list2 <- list(cbo_clus6_granule$gene, ONS76cbo_clus2$gene)
cbo.clus6.granule_ons76cbo.clus2 <- ggVennDiagram(gene_list2, category.names = c("cbo_clus6_granule", "ONS76cbo_clus2"))
cbo.clus6.granule_ons76cbo.clus2
```

```VennDiagClus6CBOvsClus9ONS76CBO
gene_list3 <- list(cbo_clus6_granule$gene, ONS76cbo_clus9$gene)
cbo.clus6.granule_ons76cbo.clus9 <- ggVennDiagram(gene_list3, category.names = c("cbo_clus6_granule", "ONS76cbo_clus9"))
cbo.clus6.granule_ons76cbo.clus9
```

```VennDiagClus6CBOvsClus10ONS76CBO
gene_list4 <- list(cbo_clus6_granule$gene, ONS76cbo_clus10$gene)
cbo.clus6.granule_ons76cbo.clus10 <- ggVennDiagram(gene_list4, category.names = c("cbo_clus6_granule", "ONS76cbo_clus10"))
cbo.clus6.granule_ons76cbo.clus10
```

----------------------------------------------------------------------------------------------------
***Generate correlation matrix of gene expression in each of the NEUROD1+ clusters of ONS76-CBO and CBO control***
Find AverageExpression() of each of the NEUROD1+ clusters, and eventually construct a dataframe of the Average Expression of all 4 NEUROD1+ clusters. The aim is to generate a correlation matrix, first using all the genes, and then using the top 500 DEGs of each cluster.
```
{r}
av.exp.cl5.cbo <- AverageExpression(nd1.clus5.cbo)$RNA
head(av.exp.cl5.cbo)
```

Is the above result the same as if I do AverageExpression on all the seurat clusters? YES!!
```
{r}
# converted to a dataframe (original data structure is matrix)
av.exp.all <- as.data.frame(AverageExpression(phase.cbo.sct.clust)$RNA)
head(av.exp.all)
```

```
{r}
class(av.exp.all)
```


```
{r}
# subset each NEUROD1 cluster in the ONS76-CBO sample (cluster 9 is NOT a NeuroD1 cluster!)
nd1.cl3.ons76cbo <- subset(phase.ons76cbo.sct.clust, idents = c("3"))

nd1.cl4.ons76cbo <- subset(phase.ons76cbo.sct.clust, idents = c("4"))

nd1.cl8.ons76cbo <- subset(phase.ons76cbo.sct.clust, idents = c("8"))

cl9.ons76cbo <- subset(phase.ons76cbo.sct.clust, idents = c("9")) 

```

```
{r}
# compute AverageExpression of each of the ONS76-CBO NEUROD1+ clusters
av.exp.cl3.ons76cbo <- AverageExpression(nd1.cl3.ons76cbo)$RNA
av.exp.cl4.ons76cbo <- AverageExpression(nd1.cl4.ons76cbo)$RNA
av.exp.cl8.ons76cbo <- AverageExpression(nd1.cl8.ons76cbo)$RNA
av.exp.cl9.ons76cbo <- AverageExpression(cl9.ons76cbo)$RNA
```

Check lengths of the objects. They are different. Cluster 5 from the CBO control has the largest number of genes 
```
{r}
length(av.exp.cl5.cbo)
length(av.exp.cl3.ons76cbo)
length(av.exp.cl4.ons76cbo)
length(av.exp.cl8.ons76cbo)
length(av.exp.cl9.ons76cbo)
```

```
{r}
# change colnames of objects (clusters)
colnames(av.exp.cl5.cbo) <- "cl5_cbo"
colnames(av.exp.cl3.ons76cbo) <- "cl3_ons76cbo"
colnames(av.exp.cl4.ons76cbo) <- "cl4_ons76cbo"
colnames(av.exp.cl8.ons76cbo) <- "cl8_ons76cbo"
colnames(av.exp.cl9.ons76cbo) <- "cl9_ons76cbo"
```

Make the rownames (which are gene symbols) of each of the "av.exp..." objects into a new column  with rownames_to_column(). First convert the object to a dataframe.
```
{r}
av.exp.cl5.cbo.2cols <- as.data.frame(av.exp.cl5.cbo) %>% 
  rownames_to_column("genes")
av.exp.cl3.ons76cbo.2cols <- as.data.frame(av.exp.cl3.ons76cbo) %>% 
  rownames_to_column("genes")
av.exp.cl4.ons76cbo.2cols <- as.data.frame(av.exp.cl4.ons76cbo) %>% 
  rownames_to_column("genes")
av.exp.cl8.ons76cbo.2cols <- as.data.frame(av.exp.cl8.ons76cbo) %>% 
  rownames_to_column("genes")
av.exp.cl9.ons76cbo.2cols <- as.data.frame(av.exp.cl9.ons76cbo) %>% 
  rownames_to_column("genes")
```

This worked, as shown by below example
```
{r}
head(av.exp.cl5.cbo.2cols)
```

Are there any duplicate genes in the ONS76-CBO sample? No
```
{r}
sum(duplicated(av.exp.cl3.ons76cbo.2cols))
sum(duplicated(av.exp.cl4.ons76cbo.2cols))
sum(duplicated(av.exp.cl8.ons76cbo.2cols))
```

Are there any duplicated genes in the CBO sample? No
```
{r}
sum(duplicated(av.exp.cl5.cbo.2cols))
```

Are the genes in 'av.exp.cl5.cbo.2cols' the same as the genes in 'av.exp.cl4.ons76cbo.2cols$genes'. NO!!
```
{r}
# compare the genes in Cluster 5 of CBO and Cluster 4 of ONS76-CBO sample - they are NOT identical
identical(av.exp.cl5.cbo.2cols$genes, av.exp.cl4.ons76cbo.2cols$genes)
```

Are the genes in NeuroD1 cluster 3, 4, 8 of the ONS76-CBO sample identical - they are.
```
{r}
# compare the genes in Cluster 3 and Cluster 4 of ONS76-CBO sample - they are identical
identical(av.exp.cl3.ons76cbo.2cols$genes, av.exp.cl4.ons76cbo.2cols$genes)
```

```
{r}
# compare the genes in Cluster 4 of ONS76-CBO and Cluster 8 of ONS76-CBO samples - they are identical
identical(av.exp.cl4.ons76cbo.2cols$genes, av.exp.cl8.ons76cbo.2cols$genes)
```

Join the 'av.exp..' dataframes by "genes" so that all the 'genes' in the smaller dfs are kept and the genes that are in the longer df that are not also present in the shorter dfs are removed

```
{r}
merge3 <- join_all(list(av.exp.cl3.ons76cbo.2cols, 
                        av.exp.cl4.ons76cbo.2cols, 
                        av.exp.cl8.ons76cbo.2cols,
                        av.exp.cl9.ons76cbo.2cols,
                        av.exp.cl5.cbo.2cols), 
                       by = "genes", type = "left")
head(merge3)
```

```
{r}
colSums(is.na(merge3))
```

```
{r}
# Replace all the NA values with zeros
merge3[is.na(merge3)] <- 0
colSums(is.na(merge3))
```

```
{r}
# check expression of NEUROD1, RBFOX3 and RTN1
merge3 %>% 
  filter(genes %in% c("NEUROD1", "RBFOX3", "RTN1", "MEIS1", "CNTN2"))
```


```
{r}
# save the 'merge3' object which is the Average Expression of all the clusters (NA values removed)
saveRDS(merge3, 
        file = "./outputs50K/ONS76-CBO/neurod1ClustsANDons76CboClus9.ons76cbo.cbo.allGenes.AvgExpressn.dataframe.rds")
```

```
{r}
# load the AverageExpression dataframe containing ALL genes
merge3 <- 
  readRDS("outputs50K/ONS76-CBO/neurod1ClustsANDons76CboClus9.ons76cbo.cbo.allGenes.AvgExpressn.dataframe.rds")
```

****Correlation matrix using ALL genes****
The correlation matrix below shows the correlation strength when comparing Average Expression of ALL genes in the matrix. However, plotting the correlation matrix in this form will not look good!
```
{r}
# use dplyr to select only the columns with numeric values (the RNA counts)
RNAcounts <- merge3 %>% 
  select(-genes)

ons76cbo.cor <- cor(RNAcounts[, colnames(RNAcounts) != "cl5_cbo"], RNAcounts$cl5_cbo)
ons76cbo.cor
```

Get the full correlation matrix - correlation between all clusters
```
{r}
corr.ons76sbo.cbo <- cor(RNAcounts)
corr.ons76sbo.cbo
```

Refer to https://dk81.github.io/dkmathstats_site/rvisual-corrplots.html to plot this upper triangular correlation plot
```
{r}
upper.ons76cbo.cbo <- corr.ons76sbo.cbo

# Make upper triangular matrix by setting NA to lower triangular part:
upper.ons76cbo.cbo[lower.tri(upper.ons76cbo.cbo)] <- NA
upper.ons76cbo.cbo
```

```
{r}
# Melt this upper triangular matrix (package reshape2 needed) and remove NA values:
up.m.ons76cbo.cbo <- melt(upper.ons76cbo.cbo, na.rm = TRUE)
head(up.m.ons76cbo.cbo)
nrow(up.m.ons76cbo.cbo)
```

This is a much nicer looking plot - the CORRELATION HEATMAP FOR ALL GENES
```
{r CorrAllGenesNeurod1ClustsONS76cboVScbo}
# ggplot upper triangular correlation matrix:
ggplot(data = up.m.ons76cbo.cbo, aes(x = Var1, y = Var2, fill = value)) +
  geom_tile() +
  theme_classic() +
  labs(x = "", y = "", fill = "Correlation \n Measure \n") +
  scale_x_discrete(labels = c("Clust 3_ONS76-CBO", 
                              "Clust 4_ONS76-CBO",
                              "Clust 8_ONS76-CBO",
                              "Clust 9_ONS76-CBO",
                              "Clust 5_CBO")) +
  scale_y_discrete(labels = c("Clust 3_ONS76-CBO", 
                              "Clust 4_ONS76-CBO",
                              "Clust 8_ONS76-CBO",
                              "Clust 9_ONS76-CBO",
                              "Clust 5_CBO")) +
  theme(axis.text.x = element_text(face = "bold", angle = 90, hjust = 1),
        axis.text.y = element_text(face = "bold", vjust = 1),
        axis.ticks = element_blank(),
        axis.line = element_line(linetype = "blank"),
        panel.background = element_rect(fill = "gray87"),
        panel.grid.major = element_line(colour = "gray47", linetype = "dotted", size = 0.2)
        ) +
  geom_text(aes(label = round(value, 3)), color = "white", fontface = "bold", size = 3) +
  coord_fixed()
```

***Test whether it is better to select the top DEGs, using FindMarkers, in the healthy CBO NEUROD1 cluster 5 rather than all genes as I have done.*** 
It is NOT better to select the top DEGs than useing all genes
Note I am using the SCTransformed, CC difference regressed 'phase' object

```
# NO NEED TO RE-RUN THESE CODE CHUNKS

# choose top 500 DEGs of cluster 5 of CBO sample
cl5.cbo.sct.DEG <- FindMarkers(object = phase.cbo.sct.clust,
                                    ident.1 = 5,
                                    min.pct = 0.5)
cl5.cbo.top500DEG <- head(cl5.cbo.sct.DEG, n = 500)
head(cl5.cbo.top500DEG, n=10)
```

Find the top 500 DEGs in each of the 3 clusters in ONS76-CBO sample, generated after applying SCTransform
```
# choose top 500 DEGs of cluster 3 of ONS76-CBO sample
cl3.ons76cbo.sct.DEG <- FindMarkers(object = phase.ons76cbo.sct.clust,
                                    ident.1 = 3,
                                    min.pct = 0.5)
cl3.ons76cbo.top500DEG <- head(cl3.ons76cbo.sct.DEG, n = 500)
head(cl3.ons76cbo.top500DEG, n=10)
```
```
# choose top 500 DEGs of cluster 4 of ONS76-CBO sample
cl4.ons76cbo.sct.DEG <- FindMarkers(object = phase.ons76cbo.sct.clust,
                                    ident.1 = 4,
                                    min.pct = 0.5)
cl4.ons76cbo.top500DEG <- head(cl4.ons76cbo.sct.DEG, n = 500)
head(cl4.ons76cbo.top500DEG, n=10)
```

```
# choose top 500 DEGs of cluster 8 of ONS76-CBO sample
cl8.ons76cbo.sct.DEG <- FindMarkers(object = phase.ons76cbo.sct.clust,
                                    ident.1 = 8,
                                    min.pct = 0.5)
cl8.ons76cbo.top500DEG <- head(cl8.ons76cbo.sct.DEG, n = 500)
head(cl8.ons76cbo.top500DEG, n=10)
```

So, now we have 500 DEGs for each of our 4 clusters, we need to extract the RNA counts for each DEG for each cluster from merge3 which already has all counts for all genes, including these.
```
cl5.cbo.top500DEG.2 <- cl5.cbo.top500DEG %>% 
  rownames_to_column(var = "gene")

cl3.ons76cbo.top500DEG.2 <- cl3.ons76cbo.top500DEG %>% 
  rownames_to_column(var = "gene")

cl4.ons76cbo.top500DEG.2 <- cl4.ons76cbo.top500DEG %>% 
  rownames_to_column(var = "gene")

cl8.ons76cbo.top500DEG.2 <- cl8.ons76cbo.top500DEG %>% 
  rownames_to_column(var = "gene")

allcl.deg <- rbind(cl5.cbo.top500DEG.2,
                   cl3.ons76cbo.top500DEG.2, 
                   cl4.ons76cbo.top500DEG.2,
                   cl8.ons76cbo.top500DEG.2)

genes.deg <- allcl.deg %>% 
  select(gene)

str(genes.deg)
```

```
# convert gene.vec to a vector
gene.vec <- genes.deg$gene
str(gene.vec)
```

```
# count the number of duplicates
sum(duplicated(gene.vec))
```

```
# get rid of duplicates in the vector 'gene.vect'
gene.vec.uniq <- gene.vec %>% 
  unique()
length(gene.vec.uniq)
```

```
# subset merge3 dataframe so we only keep these 1196 genes
merge3.deg <- merge3 %>% 
  filter(genes %in% gene.vec.uniq)
dim(merge3.deg)
```

```
# save merge3.deg which only contains a subset of genes of merge3, corresponding to DEG genes of all
# the NeuroD1+ clusters from ONS76-CBO and CBO control samples
saveRDS(merge3.deg, file = "outputs50K/ONS76-CBO/neurod1Clusts.ons76cbo.cbo.DEGgenes.AvgExpressn.dataframe.rds")
```

```
{r}
# load the AverageExpression dataframe containing only DEGs
merge3.deg <- readRDS("outputs50K/ONS76-CBO/neurod1Clusts.ons76cbo.cbo.DEGgenes.AvgExpressn.dataframe.rds")
```


The correlation is maximal for cluster 4 in the ON76-CBO sample with cluster 5 of the CBO sample. This means, cluster 4 in ONS76-CBO is the healthy cluster. After repeating the correlation using just (top 500) DEGs, the result is pretty similar to using the entire gene set of ~20K genes
```
{r}
# use dplyr to select only the columns with numeric values (the RNA counts)
RNAcounts.degs <- merge3.deg %>% 
  select(-genes)

ons76cbo.cor.degs <- cor(RNAcounts.degs[, colnames(RNAcounts.degs)], RNAcounts.degs$cl5_cbo)
ons76cbo.cor.degs
```

```
{r}
# convert to dataframe
ons76cbo.cor.degs.df <- as.data.frame(ons76cbo.cor.degs) %>% 
  rownames_to_column(var = "cluster")
```

```
{r}
# overview of object
str(ons76cbo.cor.degs.df)
```

```
{r}
ons76cbo.cor.degs.df$V2 <- "cl5_cbo"
str(ons76cbo.cor.degs.df)
```

This heatmap shows that the correlation plot works and that cluster 4 of ONS76-CBO has the greatest similarity to cluster 5 CBO, which was the expected result. However, using a subset of genes of merge 3, that are DEGs of each cluster is no better than using the full list of genes. The findings are consistent using either the full list of genes or just the DEGs of each cluster. This heatmap appearance is just shown as an example as this plot is not of publication quality!
```
{r CorrlnPlotHealthyNeuroD1VsHealthyAndONS76cocultNeuroD1}
ggplot(data = ons76cbo.cor.degs.df, aes(x = cluster, y = V2, fill = V1)) +
  geom_tile() +
  coord_fixed() +
  theme_classic()
```

A better way to show the correlation using just DEGs is with the upper triangular matrix plot. Start by generating the correlation matrix as before
```
{r}
# generate correlation matrix
corr.ons76sbo.cbo.degs <- cor(RNAcounts.degs)
corr.ons76sbo.cbo.degs
```

```
{r}
upper.ons76cbo.cbo.degs <- corr.ons76sbo.cbo.degs

# Make upper triangular matrix by setting NA to lower triangular part:
upper.ons76cbo.cbo.degs[lower.tri(upper.ons76cbo.cbo.degs)] <- NA
upper.ons76cbo.cbo.degs
```

```
{r}
# 'Melt' this upper triangular matrix (package reshape2 needed) and remove NA values:
up.m.ons76cbo.cbo.degs <- melt(upper.ons76cbo.cbo.degs, na.rm = TRUE)
head(up.m.ons76cbo.cbo.degs)
```

```
{r CorrlnPlotNeuroD1ons76cocultCboSampleAllgenes}
# ggplot upper triangular correlation matrix:
ggplot(data = up.m.ons76cbo.cbo.degs, aes(x = Var1, y = Var2, fill = value)) +
  geom_tile() +
  theme_classic() +
  labs(x = "", y = "", fill = "Correlation \n Measure") +
  scale_x_discrete(labels = c("Clust 3_ONS76-CBO", 
                              "Clust 4_ONS76-CBO",
                              "Clust 8_ONS76-CBO",
                              "Clust 5_CBO")) +
  scale_y_discrete(labels = c("Clust 3_ONS76-CBO", 
                              "Clust 4_ONS76-CBO",
                              "Clust 8_ONS76-CBO",
                              "Clust 5_CBO")) +
  theme(axis.text.x = element_text(face = "bold", angle = 90, hjust = 1),
        axis.text.y = element_text(face = "bold", vjust = 1),
        axis.ticks = element_blank(),
        axis.line = element_line(linetype = "blank"),
        panel.background = element_rect(fill = "gray87"),
        panel.grid.major = element_line(colour = "gray47", linetype = "dotted", size = 0.2)
        ) +
  geom_text(aes(label = round(value, 3)), color = "white", fontface = "bold", size = 4) +
  coord_fixed()
```






















