---
title: "ONS76-2D_Seurat_QC1_50K"
author: "JJ"
date: "05/08/2022"
output: html_document
---

```{r graphical_output}
options(bitmapType='cairo-png') # this line of code needed BEFORE knitr code in chunk below
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = "png",
                      dpi = 300,
                      echo = TRUE,   # prints the code (if FALSE then only code output is shown)
                      cache = TRUE)
```

```{r libraries}
suppressPackageStartupMessages({
library(Seurat)
library(Matrix)
library(patchwork)
library(tidyverse)
})
```

```{r path_to_data}
data_dir <- "./working_data/ONS76-2D/"   # the dot is essential!

list.files(data_dir) # should show matrix.mtx.gz, features.tsv.gz, barcodes.tsv.gz

```

```{r read_expression_matrix}
# note these are .gz files which can be directly read. The number of barcodes detected
# is a few more than with 20K dataset

ons76.2d.data <- ReadMtx(mtx = paste0(data_dir, "matrix.mtx.gz"),
                         features = paste0(data_dir, "features.tsv.gz"), 
                         cells = paste0(data_dir, "barcodes.tsv.gz"))

dim(ons76.2d.data)
```

```{r distrib_detect_genes_stats}
# Check how many genes have at least one transcript in each cell. 

at_least_one <- apply(ons76.2d.data, 2, function(x) sum(x>0)) # counts the sum of rows that have a                                                                         count value (x > 0) for each column
summary(at_least_one) # median number of genes is higher than for 20K dataset

# summary of total expression per cell. It's high (median total expressn=28,030 genes/cell), which is higher than 20K dataset.
summary(colSums(ons76.2d.data))
```

```{r Plots_GenesPerCell, fig.height=10, fig.width=8}

par(mfrow = c(2, 1))

hist(at_least_one, breaks = 100,
     main = "Distribution of detected genes",
     xlab = "Genes with at least one transcript")

hist(colSums(ons76.2d.data),                             # total counts (expression) per cell
     breaks = 100, main = "Expression sum per cell",
     xlab = "Sum expression")

```

```{r genesPerCell}
# manually check the number of genes (that have transcripts) detected in three or more cells
# ~19K genes are expressed in at least 3 cells
# ~17K genes are not expressed in at least 3 cells
tmp <- apply(ons76.2d.data, 1, function(x) sum(x > 0))
table(tmp >= 3)
length(tmp) # total number of genes expressed by all cells in dataset
```

```{r cells_200genes}
# NOT all 648 cells have at least 200 (common Seurat threshold) expressed genes
# (min = 61 genes, see below)

keep <- tmp>=3                  # for each gene this states TRUE or FALSE

tmp2 <- ons76.2d.data[keep, ]   # subset ons76.2d.data to keep only TRUE values, i.e. genes which 
                                # are expressed in 3 or more cells

dim(tmp2)                       # dim is 19,194 by 648, as, for example for some genes the 648th cell may
                                # be one of the cells expressing that gene

genes_in_3orMoreCells <- apply(tmp2, 2, function(x) sum(x > 0))   # for each cell, this counts the number 
                                                                  # of genes that cell expresses for all
                                                                  # genes expressed in 3 or more cells

summary(genes_in_3orMoreCells)
```

```{r Plot_sorted_genesPerCell}
# shape of graph is as expected
genes_per_cell <- Matrix::colSums(ons76.2d.data > 0) # count gene only if it has non-zero reads                                                             mapped.

plot(sort(genes_per_cell), 
     xlab='cell',
     ylab = "sorted genes per cell",
     log = "y",                             # only y axis is logarithmic
     main='Genes per cell (ordered)') 
```

```{r create_seurat_object}

ons76.2d.seurat <- CreateSeuratObject(counts = ons76.2d.data, 
                                    min.cells = 3, 
                                    min.genes = 200, 
                                    project = "ONS76-2D")
dim(ons76.2d.seurat)
```

```{r add_percent_mitochondria_meta.data}
# add mitochondrial percentage to metadata
ons76.2d.seurat[["percent.mt"]] <- PercentageFeatureSet(ons76.2d.seurat, pattern = "^MT-")
head(ons76.2d.seurat@meta.data)
```

```{r metadata_dataframe_642cells}
# create metadata dataframe on its own
metadata <- ons76.2d.seurat@meta.data

# Add cell IDs to metadata
metadata$cells <- rownames(metadata)

# Rename columns to be more intuitive
metadata <- metadata %>%
        dplyr::rename(sample = orig.ident,
                      nUMI = nCount_RNA,
                      nGene = nFeature_RNA)
dim(metadata)
```

```{r violinPlot_genes_transcripts_percent.mt_648cells, fig.height=10, fig.width=5}
# there are some cells with excessive mitochondrial transcripts
# the data looks bimodal, smaller lower 'bulge'

p1 <- metadata %>% 
        ggplot(aes(x = sample, y = nUMI)) +
        geom_violin(colour = "blue", fill = "blue", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.2)

p2 <- metadata %>% 
        ggplot(aes(x = sample, y = nGene)) +
        geom_violin(colour = "brown1", fill = "brown1", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.2)

p3 <- metadata %>% 
        ggplot(aes(x = sample, y = percent.mt)) +
        geom_violin(colour = "cadetblue", fill = "cadetblue", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.2)

p1 / p2 / p3    
                       
```

```{r Plot_UMIcountsPerCell_648cells}
# visualise number of UMIs (transcripts) per cell using all 648 cells. 
# In geom_density use (y = ..scaled..) value that stat_density provides

metadata %>% 
        ggplot(aes(x = nUMI)) +
        geom_density(aes(y = ..scaled..)) +
        scale_x_log10() +
        theme_classic() +
        ylab("cell density") +
        geom_vline(xintercept = 500, color = "blue")
```

```{r mitochondrial_thresholds}
# number of cells with percent.mt > 10%: 139 cells
# number of cells with percent.mt < 10%: 509 cells

nrow(metadata[metadata$percent.mt > 10, ])

nrow(metadata[metadata$percent.mt < 10, ])
```

```{r filtered_mitoch_seuratObject}
# remove cells with percent mitochondrial gene expression > 10, i.e. remove 139 cells
# CELL-LEVEL filtering leaves 509 cells

filtered_ons76.2d.seurat <- subset(ons76.2d.seurat, subset = nFeature_RNA > 200 & percent.mt < 10)
dim(filtered_ons76.2d.seurat)
```

```{r violinPlot_genes_transcripts_percent.mt_2252cells, fig.height=10, fig.width=5}
# create a metadata dataframe
# new plots show remaining data looks unimodal
# mitochondrial gene filtration successful

metadata.2 <- filtered_ons76.2d.seurat@meta.data

# Add cell IDs to metadata
metadata.2$cells <- rownames(metadata.2)

# Rename columns to be more intuitive
metadata.2 <- metadata.2 %>%
        dplyr::rename(sample = orig.ident,
                      nUMI = nCount_RNA,
                      nGene = nFeature_RNA)

p4 <- metadata.2 %>% 
        ggplot(aes(x = sample, y = nUMI)) +
        geom_violin(colour = "blue", fill = "blue", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.1)

p5 <- metadata.2 %>% 
        ggplot(aes(x = sample, y = nGene)) +
        geom_violin(colour = "brown1", fill = "brown1", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.1)

p6 <- metadata.2 %>% 
        ggplot(aes(x = sample, y = percent.mt)) +
        ylim(0, 100) +
        geom_violin(colour = "cadetblue", fill = "cadetblue", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.1)

p4 / p5 / p6
```

```{r extract_countsSlot_509cells}
# subset '@counts' slot (matrix) of filtered seurat object

counts.509 <- GetAssayData(object = filtered_ons76.2d.seurat, slot = "counts")
```

```{r summarise_stats_509cells}
# median genes per cell and expression sum, as above. Min. genes = 556. Median = 6104.
# Median genes is higher than before cells with high mitoch reads filtered out, which makes sense as the latter cells typically have lower number of genes expressed

at_least_one.509 <- apply(counts.509, 2, function(x) sum(x>0))

summary(at_least_one.509)

summary(colSums(counts.509))
```

```{r Plots_GenesPerCell_509cells, fig.height=10, fig.width=8}
# 1 major peak now, but still a group of cells to the left with lower gene expression. 
# There are a few cells in the tail to the right of each graph which look like outliers with 
# very high gene expression.
# In the bottom graph far fewer cells with low expression sum

par(mfrow = c(2, 1))

hist(at_least_one.509, breaks = 100,
     main = "Distribution of detected genes",
     xlab = "Genes with at least one transcript")

hist(colSums(counts.509),                             # total counts (expression) per cell
     breaks = 100, main = "Expression sum per cell",
     xlab = "Sum expression")
```

```{r Plots_UMI_gene_expression_509cells, fig.height=10, fig.width=5}

# looks ok, most of the filtered cells have UMI counts >10000 and <100,000 and express 3,000-10,000 genes

p7 <- metadata.2 %>% 
       ggplot(aes(x = nUMI)) +
        geom_density(aes(y = ..scaled..), colour = "blue", fill = "blue", alpha = 0.2) +
        theme_classic() +
        scale_x_log10() +
        ylab("Cell density")

p8 <- metadata.2 %>% 
        ggplot(aes(x = nGene)) +
        geom_density(aes(y = ..scaled..), colour = "brown1", fill = "brown1", alpha = 0.2) +
        theme_classic() +
        scale_x_log10() +
        ylab("Cell density")

p7 / p8
```

```{r Plot_UMI_v_genes_509cells}
# plot of UMI count vs no. of genes

p9 <- FeatureScatter(filtered_ons76.2d.seurat, 
                      feature1 = "nCount_RNA",        # nCount_RNA is same as UMI count
                      feature2 = "nFeature_RNA")

p9
```

```{r filtered_seuratObject_GeneLevel}
# GENE LEVEL FILTERING (applied on cell-filtered Seurat object)
#-----------------------------------------------------------------------------------
#-----------------------------------------------------------------------------------
#  we will have many genes with zero counts, which can dramatically reduce the average expression for a cell and so we will remove them from our data. Remove genes that have zero expression in all cells.

# # Extract counts
counts <- GetAssayData(object = filtered_ons76.2d.seurat, slot = "counts")

# Output a logical vector for every gene on whether there are more than zero counts per cell
nonzero <- counts > 0

# If a gene is only expressed in a handful of cells, it is not particularly meaningful as it still brings down the averages for all other cells it is not expressed in. For our data we choose to keep only genes which are expressed in 10 or more cells.
# Sums all TRUE values and returns TRUE if more than 10 TRUE values per gene
keep_genes <- Matrix::rowSums(nonzero) >= 10    # sum the TRUE values in a row with 'gene' as rowname

# Only keeping those genes expressed in more than 10 cells
filtered_counts <- counts[keep_genes, ]

# Reassign to filtered Seurat object
filtCellsGenes.ons76.2d.seurat <- CreateSeuratObject(filtered_counts, meta.data = filtered_ons76.2d.seurat@meta.data)

dim(filtCellsGenes.ons76.2d.seurat)       # number of genes further reduced to ~ 15K genes vs 19K genes                                               for only cell-level filtering
```

```
# saved the 'filtCellsGenes.ons76.2d.seurat' object to outputs50K/ONS76_2D folder

saveRDS(filtCellsGenes.ons76.2d.seurat, file = "./outputs50K/ONS76_2D/filtCellsGenes.ons76.2d.seurat.rds")
```

```{r remove_ribo_genes}
# FILTER OUT ALL RIBOSOMAL GENES, THEN ALL MITOCHONDRIAL GENES (ON THE CELLS WHICH HAVE ALREADY BEEN REMOVED
# THAT HAVE PERCENT.MITO < 10% BEFORE DOING THE CLUSTERING), DON'T REMOVE ANY MORE ACTUAL CELLS
#-------------------------------------------------------------------------------------

# remove ribosomal genes first
filt.noRibo.ons76.2d <- fullFilt.ons76.2d.seurat[ ! grepl('^RP[SL]', rownames(fullFilt.ons76.2d.seurat)), ]
```

```{r check_ribo_removed}
# this has worked - 95 ribosomal genes have been removed from the new Seurat object
dim(fullFilt.ons76.2d.seurat)
dim(filt.noRibo.ons76.2d)
```

```{r remove_mitoch_genes}
# remove mitochondrial genes 
filt.noRiboMito.ons76.2d <- filt.noRibo.ons76.2d[ ! grepl("^MT-", rownames(filt.noRibo.ons76.2d)), ]
```

```{r check_filtering}
# check if ribosome and mitochondrial gene filtering works - it does!
dim(filt.noRibo.ons76.2d)
dim(filt.noRiboMito.ons76.2d)
```

```{r}
# SAVE this seurat object from code line 344

saveRDS(filt.noRiboMito.ons76.2d, file = "./outputs50K/ONS76_2D/filt.NoRiboNoMito.ons76.2d.rds")
```

--------------------------------------------------------------------------------------------
***EXPLORE SOURCES OF UNWANTED VARIATION - Cell Cycle***

```{r load_filtered_CellsGenes_SeuratObject}
# the name of this object reflects it has undergone filtering for cells AND genes

filt.ons76mono <- readRDS("outputs50K/ONS76_2D/filt.NoRiboNoMito.ons76.2d.rds")
```

```{r}
# normalise data - WHY NORMALISE BEFORE RUNNING SCTRANSFORM?? THIS IS NOT DONE IN SCTRANSFORM VIGNETTE!!!
filt.ons76mono <- NormalizeData(filt.ons76mono)

# run cell cycle scoring
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
filt.ons76mono <- CellCycleScoring(filt.ons76mono,
                                 g2m.features = g2m.genes,
                                 s.features = s.genes)
```

Run SCTransfrom WITHOUT vars.to.regress argument
```{r DimplotSCTCCdifferenceRetained}
# there is clustering by cell cycle with 3 clusters G2/M cluster, S cluster, G1 cluster

filt.ons76mono.sct <- SCTransform(filt.ons76mono,
                            method = "glmGamPoi", 
                            verbose = FALSE)

filt.ons76mono.sct <- RunPCA(filt.ons76mono.sct, features = c(s.genes, g2m.genes))

DimPlot(filt.ons76mono.sct, reduction = "pca", group.by = "Phase")
```

--------------------------------------------------------------
***Using SCTransform with cell cycle regression***

Run SCTransfrom WITH vars.to.regress argument
```{r DimplotSCTRegressOutCCdifference}
# regress the cell cycle difference - this worked: G2/M and S become a single cluster distinct from G1
filt.ons76mono.sct$CC.difference <- filt.ons76mono.sct$S.Score - filt.ons76mono.sct$G2M.Score

filt.ons76mono.sct <- SCTransform(filt.ons76mono.sct,
                                  vars.to.regress = "CC.difference",
                                  method = "glmGamPoi", 
                                  verbose = FALSE)

filt.ons76mono.sct <- RunPCA(filt.ons76mono.sct, features = c(s.genes, g2m.genes))

DimPlot(filt.ons76mono.sct, reduction = "pca", group.by = "Phase")
```


```{r normalize_fully_filtered_seuratObject}
# standard normalization function is sufficient to explore whether cells cluster by cell-cycle phase

seurat.phase <- NormalizeData(fullFilt.ons76.2d.seurat)
seurat.phase <- FindVariableFeatures(seurat.phase, selection.method = "vst")
seurat.phase <- ScaleData(seurat.phase, features = rownames(seurat.phase))
seurat.phase <- RunPCA(seurat.phase, features = VariableFeatures(seurat.phase))
```

```{r cellCycle_effects}
# Score cells for cell cycle.
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
seurat.phase <- CellCycleScoring(seurat.phase,
                                 g2m.features = g2m.genes,
                                 s.features = s.genes)
```
Let's check the effect of cell cycle phase - indeed, cell cycle phase has a strong effect on clustering!
```{r DimplotCellCycleEffect}
seurat.phase <- RunPCA(seurat.phase, features = c(s.genes, g2m.genes))
DimPlot(seurat.phase, reduction = "pca", group.by = "Phase")
```

Next step is to adjust for the cell cycle  (based on https://satijalab.org/seurat/articles/cell_cycle_vignette.html).
We do not want to entirely regress out the cell cycle, as we want to separate apart cycling from NON-cycling cells. However, differences in cell cycle phase among proliferating cells will be regressed out.
```{r}
seurat.phase$CC.difference <- seurat.phase$S.Score - seurat.phase$G2M.Score
seurat.phase.ons762d <- ScaleData(seurat.phase, vars.to.regress = "CC.difference", features = rownames(seurat.phase))
```

```{r}
# cell cycle effects strongly mitigated in PCA
seurat.phase.ons762d <- RunPCA(seurat.phase.ons762d, features = VariableFeatures(seurat.phase), nfeatures.print = 5)
```

This worked! Compared to before regressing out the 'cell cycle difference', the DimPlot below recapitulates the final DimPlot on that webpage.
```{r DimplotCCdiffRegress}
seurat.phase.ons762d <- RunPCA(seurat.phase.ons762d, features = c(s.genes, g2m.genes))
DimPlot(seurat.phase.ons762d, reduction = "pca", group.by = "Phase")
```


```{r}
# normalise data - WHY NORMALISE BEFORE RUNNING SCTRANSFORM?? THIS IS NOT DONE IN SCTRANSFORM VIGNETTE!!!
seurat.phase.sct <- NormalizeData(fullFilt.ons76.2d.seurat)

# run cell cycle scoring
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
seurat.phase.sct <- CellCycleScoring(seurat.phase.sct,
                                 g2m.features = g2m.genes,
                                 s.features = s.genes)
```

Normalisation, followed by CellCycleScoring, then SCTransform, then PCA worked! Get a similar result to DimplotRegressOutCCdifference code chunk above. Much faster to run. This code separates G1 from other cell cycle phases.


```{r}
# save the seurat.phase.sct object as a .RDS file in outputs50K
saveRDS(seurat.phase.sct, file = "./outputs50K/ONS76_2D/seurat.phase.SCT.regressedCCdiff.ons76.2d.rds")
```


```{r mitochondrial_expression}
# Check quartile values
summary(seurat.phase@meta.data$percent.mt)

# Turn percent.mt into categorical factor vector based on quartile values
seurat.phase@meta.data$mito.prop <- cut(seurat.phase@meta.data$percent.mt, 
                  breaks=c(-Inf, 1.8, 5.0, 9.99, Inf), 
                  labels=c("Low","Medium","Medium high", "High"))
```

```{r plot_mitochondrialExpression}
# repeat PCA and plot the PCA colored by mitochondrial percentage - this variable has no effect
# on the clustering, just like cell-cycle phase
seurat.phase <- RunPCA(seurat.phase)

DimPlot(seurat.phase,
        reduction = "pca",
        group.by= "mito.prop",
        split.by = "mito.prop")
```


------------------------------------------------------------------------------
Run SCTransform again on the seurat object below - fully filtered object for cells, genes, ribosomal and mitochondrial genes
```{r load_noRiboNoMito_seuratObject}
filt.noRiboMito.ons76.2d <- readRDS("outputs50K/ONS76_2D/filt.NoRiboNoMito.ons76.2d.rds")
```

```{r}
# THIS IS THE OBJECT TO LOAD FOR DOWNSTREAM ANALYSIS!!!
# save the GENE, CELL, MITOCHONDRIAL and RIBOSOMAL FILTERED seurat object with regression of the cell cycle
# this object has been saved
saveRDS(phase.ons762d.filtRiboMito.sct, file = "./outputs50K/ONS76_2D/phase.filtRiboMito.SCT.regressedCCdiff.ons762d.rds")
```










