---
title: "ONS76-SPHER_Seurat_QC1_50K"
author: "JJ"
date: "05/08/2022"
output: html_document
---

```{r graphical_output}
options(bitmapType='cairo-png') # this line of code needed BEFORE knitr code in chunk below
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = "png",
                      dpi = 300,
                      echo = TRUE,   # prints the code (if FALSE then only code output is shown)
                      cache = TRUE)
```

```{r libraries}
library(Seurat)
library(Matrix)
library(patchwork)
library(tidyverse)
```

```{r path_to_data}
data_dir <- "./working_data/ONS76-SPHER/"   # the dot is essential!

list.files(data_dir) # should show matrix.mtx.gz, features.tsv.gz, barcodes.tsv.gz
```

```{r read_expression_matrix}
# note these are .gz files which can be directly read. The number of barcodes detected
# is 1322, which is a few less than with the 20K analysis

ons76.spher.data <- ReadMtx(mtx = paste0(data_dir, "matrix.mtx.gz"),
                         features = paste0(data_dir, "features.tsv.gz"), 
                         cells = paste0(data_dir, "barcodes.tsv.gz"))

dim(ons76.spher.data)
```

```{r distrib_detect_genes_stats}
# Check how many genes have at least one transcript in each cell.
# Minimum genes per cell is 108, and median expression is 4364.

at_least_one <- apply(ons76.spher.data, 2, function(x) sum(x>0)) # counts the sum of rows that have a                                                                                  count value (x > 0) for each column 
#                                                                (counts the number of TRUE values per column)
summary(at_least_one)

# summary of total expression per cell.
summary(colSums(ons76.spher.data))
```

```{r Plots_GenesPerCell, fig.height=10, fig.width=8}
# distribution of genes looks unimodal
par(mfrow = c(2, 1))

hist(at_least_one, breaks = 100,
     main = "Distribution of detected genes",
     xlab = "Genes with at least one transcript")

hist(colSums(ons76.spher.data),                             # total counts (expression) per cell
     breaks = 100, main = "Expression sum per cell",
     xlab = "Sum expression")

```

```{r genesPerCell}
# manually check the number of genes detected in three or more cells
# ~21.5K genes are expressed in at least 3 cells
# ~15K genes are not expressed in at least 3 cells
tmp <- apply(ons76.spher.data, 1, function(x) sum(x > 0))
table(tmp >= 3)
```

```{r cells_gene_counts}
# Do all 1322 cells have at least 200 expressed genes? Ans = NO! Min. genes = 108

keep <- tmp>=3                  # for each gene this states TRUE or FALSE

tmp2 <- ons76.spher.data[keep, ]        # subset ons76.spher.data to keep only TRUE values, i.e. genes which 
                                       # are expressed in 3 or more cells

class(tmp2)                     # this is a matrix
dim(tmp2)                       # dim is 21565 by 1322 (cells)

genes_in_3orMoreCells <- apply(tmp2, 2, function(x) sum(x > 0))   # for each cell, this counts the number 
                                                                  # of genes that cell expresses for all
                                                                  # genes expressed in 3 or more cells

summary(genes_in_3orMoreCells)
```

```{r Plot_sorted_genesPerCell}
# rank each cell by its library complexity, i.e. number of genes detected per cell. Shape looks OK.

genes_per_cell <- Matrix::colSums(ons76.spher.data > 0) # count gene only if it has non-zero reads                                                                   mapped.

plot(sort(genes_per_cell), 
     xlab='cell',
     ylab = "sorted genes per cell",
     log = "y",                             # only y axis is logarithmic
     main='Genes per cell (ordered)') 
```

```{r create_seurat_object}

ons76spher.seurat <- CreateSeuratObject(counts = ons76.spher.data, 
                                    min.cells = 3, 
                                    min.genes = 200, 
                                    project = "ons76spher")
dim(ons76spher.seurat)
```

```{r add_percent_mitochondria_meta.data}
# mitochondrial percentage
ons76spher.seurat[["percent.mt"]] <- PercentageFeatureSet(ons76spher.seurat, pattern = "^MT-")
head(ons76spher.seurat@meta.data)
```

```{r metadata_dataframe_1322cells}
# create metadata dataframe
metadata <- ons76spher.seurat@meta.data

# Add cell IDs to metadata
metadata$cells <- rownames(metadata)

# Rename columns to be more intuitive
metadata <- metadata %>%
        dplyr::rename(sample = orig.ident,
                      nUMI = nCount_RNA,
                      nGene = nFeature_RNA)
dim(metadata)
```

```{r violinPlot_genes_transcripts_percent.mt_3729cells, fig.height=10, fig.width=5}
# there are some cells with excessive mitochondrial transcripts
# the data looks relatively unimodal

p1 <- metadata %>% 
        ggplot(aes(x = sample, y = nUMI)) +
        geom_violin(colour = "blue", fill = "blue", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.1)

p2 <- metadata %>% 
        ggplot(aes(x = sample, y = nGene)) +
        geom_violin(colour = "brown1", fill = "brown1", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.1)

p3 <- metadata %>% 
        ggplot(aes(x = sample, y = percent.mt)) +
        geom_violin(colour = "cadetblue", fill = "cadetblue", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.1)

p1 / p2 / p3    
                       
```

```{r Plot_UMIcountsPerCell_1322cells}

# visualise number of UMIs (transcripts) per cell using all 1334 cells. 
# In geom_density use (y = ..scaled..) value that stat_density provides

metadata %>% 
        ggplot(aes(x = nUMI)) +
        geom_density(aes(y = ..scaled..)) +
        scale_x_log10() +
        theme_classic() +
        ylab("cell density") +
        geom_vline(xintercept = 500, color = "blue")
```

```{r mitochondrial_thresholds}
# number of cells with percent.mt > 10%: 151 cells
# number of cells with percent.mt < 10%: 1171 cells

nrow(metadata[metadata$percent.mt > 10, ])

nrow(metadata[metadata$percent.mt < 10, ])
```

```{r filtered_seuratObject_CellLevel}
# CELL LEVEL FILTERING - leaves ~21K genes and 1171 cells
#-------------------------------------------------------------------------------------------

# remove cells with percent mitochondrial gene expression > 10 and where no.genes/cell < 200.

filt.ons76spher.seurat <- subset(ons76spher.seurat, subset = nFeature_RNA > 200 & percent.mt < 10)
dim(filt.ons76spher.seurat)
```

```{r violinPlot_genes_transcripts_percent.mt_1171cells, fig.height=10, fig.width=5}
# create a metadata dataframe
# new plots show remaining data looks unimodal

metadata.2 <- filt.ons76spher.seurat@meta.data

# Add cell IDs to metadata
metadata.2$cells <- rownames(metadata.2)

# Rename columns to be more intuitive
metadata.2 <- metadata.2 %>%
        dplyr::rename(sample = orig.ident,
                      nUMI = nCount_RNA,
                      nGene = nFeature_RNA)

p4 <- metadata.2 %>% 
        ggplot(aes(x = sample, y = nUMI)) +
        geom_violin(colour = "blue", fill = "blue", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.1)

p5 <- metadata.2 %>% 
        ggplot(aes(x = sample, y = nGene)) +
        geom_violin(colour = "brown1", fill = "brown1", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.1)

p6 <- metadata.2 %>% 
        ggplot(aes(x = sample, y = percent.mt)) +
        ylim(0, 100) +
        geom_violin(colour = "cadetblue", fill = "cadetblue", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.1)

p4 / p5 / p6

```

```{r extract_countsSlot_1171cells}
# extract '@counts' slot (matrix) of filtered seurat object

counts.1171 <- GetAssayData(object = filt.ons76spher.seurat, slot = "counts")
```

```{r summarise_stats_1171cells}
# median genes per cell and expression sum, as above. Min. genes = 345, median genes = 4486

at_least_one.1171 <- apply(counts.1171, 2, function(x) sum(x>0))

summary(at_least_one.1171)

summary(colSums(counts.1171))
```

```{r Plots_GenesPerCell_1171cells, fig.height=10, fig.width=8}
# There are cells in the tail to the right of each graph which might be outliers with 
# very high gene expression.

par(mfrow = c(2, 1))

hist(at_least_one.1171, breaks = 100,
     main = "Distribution of detected genes",
     xlab = "Genes with at least one transcript")

hist(colSums(counts.1171),                              # total counts (expression) per cell
     breaks = 100, main = "Expression sum per cell",
     xlab = "Sum expression")
```

```{r Plots_UMI_gene_expression_1171cells, fig.height=10, fig.width=5}

# looks ok, most of the filtered ons76spher cells have UMI counts > 10,000 and express 3000-8000 genes

p7 <- metadata.2 %>% 
       ggplot(aes(x = nUMI)) +
        geom_density(aes(y = ..scaled..), colour = "blue", fill = "blue", alpha = 0.2) +
        theme_classic() +
        scale_x_log10() +
        ylab("Cell density")

p8 <- metadata.2 %>% 
        ggplot(aes(x = nGene)) +
        geom_density(aes(y = ..scaled..), colour = "brown1", fill = "brown1", alpha = 0.2) +
        theme_classic() +
        scale_x_log10() +
        ylab("Cell density")

p7 / p8
        
```

```{r Plot_UMI_v_genes_1171cells}
# plot of UMI count vs no. of genes

p9 <- FeatureScatter(filt.ons76spher.seurat, 
                      feature1 = "nCount_RNA",        # nCount_RNA is same as UMI count
                      feature2 = "nFeature_RNA")

p9

```

```{r filtered_seuratObject_GeneLevel}
# GENE LEVEL FILTERING (applied on cell-filtered Seurat object)
#-----------------------------------------------------------------------------------

#  we will have many genes with zero counts, which can dramatically reduce the average expression for a cell and so we will remove them from our data. Remove genes that have zero expression in all cells.

# # Extract counts
counts <- GetAssayData(object = filt.ons76spher.seurat, slot = "counts")

# Output a logical vector for every gene on whether there are more than zero counts per cell
nonzero <- counts > 0

# If a gene is only expressed in a handful of cells, it is not particularly meaningful as it still brings down the averages for all other cells it is not expressed in. For our data we choose to keep only genes which are expressed in 10 or more cells.
# Sums all TRUE values and returns TRUE if more than 10 TRUE values per gene
keep_genes <- Matrix::rowSums(nonzero) >= 10     # sum the TRUE values in a row with 'gene' as rowname

# Only keeping those genes expressed in more than 10 cells
filtered_counts <- counts[keep_genes, ]

# Reassign to filtered Seurat object
filtCellsGenes.ons76spher.seurat <- CreateSeuratObject(filtered_counts, meta.data = filt.ons76spher.seurat@meta.data)
```

```{r genes_cells_SeuratObject}
# 17.3K genes compared to 21.5K genes (latter is from code line 193)
dim(filtCellsGenes.ons76spher.seurat)
```

```{r}
# saved the 'filtCellsGenes.ons76spher.seurat' object to outputs50K/ONS76_SPHER folder

saveRDS(filtCellsGenes.ons76spher.seurat, file = "./outputs50K/ONS76_SPHER/filtCellsGenes.ons76spher.seurat.rds")
```

-------------------------------------------------------------------------------------
***Remove mitochondrial and ribosomal genes***
```{r remove_ribo_genes}
# FILTER OUT ALL RIBOSOMAL GENES, THEN ALL MITOCHONDRIAL GENES (ON THE CELLS WHICH HAVE ALREADY BEEN REMOVED
# THAT HAVE PERCENT.MITO < 10% BEFORE DOING THE CLUSTERING), DON'T REMOVE ANY MORE ACTUAL CELLS


# remove ribosomal genes first
filt.noRibo.ons76spher <- fullFilt.ons76spher.seurat[ ! grepl('^RP[SL]', rownames(fullFilt.ons76spher.seurat)), ]
```

```{r check_ribo_removed}
# this has worked - 95 ribosomal genes have been removed from the new Seurat object
dim(fullFilt.ons76spher.seurat)
dim(filt.noRibo.ons76spher)
```

```{r remove_mitoch_genes}
# remove mitochondrial genes 
filt.noRiboMito.ons76spher <- filt.noRibo.ons76spher[ ! grepl("^MT-", rownames(filt.noRibo.ons76spher)), ]
```

```{r check_filtering}
# check if ribosome and mitochondrial gene filtering works - it does!
dim(filt.noRibo.ons76spher)
dim(filt.noRiboMito.ons76spher)
```

```{r}
# SAVE the seurat object above (filt.noRiboMito.ons76spher) which has been filtered for CELLS, GENES, RIBOSOMAL and MITOCHODRIAL genes

saveRDS(filt.noRiboMito.ons76spher, file = "./outputs50K/ONS76_SPHER/filt.NoRiboNoMito.ons76spher.rds")
```

```{r}
# load the 'filt.noRiboMito.ons76spher' .RDS seurat object

filt.RiboMito.ons76spher <- readRDS("outputs50K/ONS76_SPHER/filt.NoRiboNoMito.ons76spher.rds")
```



--------------------------------------------------------------------------------------------
***EXPLORE SOURCES OF UNWANTED VARIATION, e.g. cell cycle***
Run SCTransform on the seurat object below - fully filtered object for cells, genes, ribosomal and mitochondrial genes
```{r}
# normalise data - WHY NORMALISE BEFORE RUNNING SCTRANSFORM?? THIS IS NOT DONE IN SCTRANSFORM VIGNETTE!!!
phase.ons76spher.filtRiboMito.sct <- NormalizeData(filt.RiboMito.ons76spher)

# run cell cycle scoring
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
phase.ons76spher.filtRiboMito.sct <- CellCycleScoring(phase.ons76spher.filtRiboMito.sct,
                                 g2m.features = g2m.genes,
                                 s.features = s.genes)
```

First run SCTransfrom WITHOUT vars.to.regress argument
```{r DimplotSCTRegressOutCCdifference}
# there is clustering by cell cycle phase, but not that many cells in G2M and S phase.
phase.ons76spher.filtRiboMito.sct <- SCTransform(phase.ons76spher.filtRiboMito.sct,
                                   method = "glmGamPoi", 
                                   verbose = FALSE)

phase.ons76spher.filtRiboMito.sct <- RunPCA(phase.ons76spher.filtRiboMito.sct, features = c(s.genes, g2m.genes))

DimPlot(phase.ons76spher.filtRiboMito.sct, reduction = "pca", group.by = "Phase")
```

Then run SCTransform with vars.to.regress argument = cell cycle difference
```{r DimplotSCTRegressOutCCdifferenceFiltRiboMito}
# now G1 and G2/M are just one cluster, separate from G1, i.e. there are 2 clusters instead of 3.

phase.ons76spher.filtRiboMito.sct$CC.difference <- phase.ons76spher.filtRiboMito.sct$S.Score - phase.ons76spher.filtRiboMito.sct$G2M.Score # get the cell cycle difference

phase.ons76spher.filtRiboMito.sct <- SCTransform(phase.ons76spher.filtRiboMito.sct,
                                   vars.to.regress = "CC.difference",
                                   method = "glmGamPoi", 
                                   verbose = FALSE)

phase.ons76spher.filtRiboMito.sct <- RunPCA(phase.ons76spher.filtRiboMito.sct, features = c(s.genes, g2m.genes))

DimPlot(phase.ons76spher.filtRiboMito.sct, reduction = "pca", group.by = "Phase")

```

```{r}
# saved the GENE, CELL, MITOCHONDRIAL and RIBOSOMAL FILTERED seurat object with regression of the cell cycle

saveRDS(phase.ons76spher.filtRiboMito.sct, file = "./outputs50K/ONS76_SPHER/phase.ons76spher.filtRiboMito.SCT.regressedCCdiff.rds")
```

















