---
title: "ONS76-CBO_Seurat_QC1_50K"
author: "JJ"
date: "07/08/2022"
output: html_document
---

```{r graphical_output}
options(bitmapType='cairo-png') # this line of code needed BEFORE knitr code in chunk below
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = "png",
                      dpi = 300,
                      echo = TRUE,   # prints the code (if FALSE then only code output is shown)
                      cache = TRUE)
```

```{r libraries}
library(Seurat)
library(Matrix)
library(patchwork)
library(tidyverse)
```

```{r path_to_data}
data_dir <- "./working_data/ONS76-CBO/"   # the dot is current dir and is essential!

list.files(data_dir) # should show matrix.mtx.gz, features.tsv.gz, barcodes.tsv.gz
```

```{r read_expression_matrix}
# note these are .gz files which can be directly read. There are 3456 cells and 36601 genes.

ons76cbo.data <- ReadMtx(mtx = paste0(data_dir, "matrix.mtx.gz"),
                         features = paste0(data_dir, "features.tsv.gz"), 
                         cells = paste0(data_dir, "barcodes.tsv.gz"))

dim(ons76cbo.data)
```

```{r distrib_detect_genes_stats}
# Check how many genes have at least one transcript in each cell.
# Minimum genes per cell is 112.
# Median expression per cell = 2742 genes

at_least_one <- apply(ons76cbo.data, 2, function(x) sum(x>0)) # counts the sum of rows that have a                                                                        count value (x > 0) for each column (counts                                                                  the number of TRUE values per column)
summary(at_least_one)

# summary of total expression per cell.
summary(colSums(ons76cbo.data))
```

```{r Plots_GenesPerCell, fig.height=10, fig.width=8}
# the distribution of detected genes looks bimodal
par(mfrow = c(2, 1))

hist(at_least_one, breaks = 100,
     main = "Distribution of detected genes",
     xlab = "Genes with at least one transcript")

hist(colSums(ons76cbo.data),                             # total counts (expression) per cell
     breaks = 100, main = "Expression sum per cell",
     xlab = "Sum expression")
```

```{r genesPerCell}
# manually check the number of genes detected in three or more cells
# ~25K genes are expressed in at least 3 cells
# ~11K genes are not expressed in at least 3 cells

tmp <- apply(ons76cbo.data, 1, function(x) sum(x > 0))
table(tmp >= 3)
```

```{r cells_gene_counts}
# Do all 3456 cells have at least 200 expressed genes? Ans = NO! Min. genes = 112

keep <- tmp>=3                       # for each gene this states TRUE or FALSE

tmp2 <- ons76cbo.data[keep, ]        # subset ons76cbo.data to keep only TRUE values, i.e. genes which 
                                     # are expressed in 3 or more cells

genes_in_3orMoreCells <- apply(tmp2, 2, function(x) sum(x > 0))   # for each cell, this counts the number 
                                                                  # of genes that cell expresses for all
                                                                  # genes expressed in 3 or more cells

summary(genes_in_3orMoreCells)
```

```{r Plot_sorted_genesPerCell}
# rank each cell by its library complexity, i.e. number of genes detected per cell.
# The graph has the expected shape

genes_per_cell <- Matrix::colSums(ons76cbo.data > 0) # count gene only if it has non-zero reads                                                                         mapped.

plot(sort(genes_per_cell), 
     xlab='cell',
     ylab = "sorted genes per cell",
     log = "y",                             # only y axis is logarithmic
     main='Genes per cell (ordered)') 
```

```{r create_seurat_object}
# Seurat object has ~25K genes and 3456 cells

ons76cbo.seurat <- CreateSeuratObject(counts = ons76cbo.data, 
                                    min.cells = 3, 
                                    min.genes = 200, 
                                    project = "ONS76-CBO")
dim(ons76cbo.seurat)
```

```{r add_percent_mitochondria_meta.data}
# mitochondrial percentage
ons76cbo.seurat[["percent.mt"]] <- PercentageFeatureSet(ons76cbo.seurat, pattern = "^MT-")
head(ons76cbo.seurat@meta.data)
```

```{r metadata_dataframe_3456cells}
# create metadata dataframe
metadata <- ons76cbo.seurat@meta.data

# Add cell IDs to metadata
metadata$cells <- rownames(metadata)

# Rename columns to be more intuitive
metadata <- metadata %>%
        dplyr::rename(sample = orig.ident,
                      nUMI = nCount_RNA,
                      nGene = nFeature_RNA)
dim(metadata)
```

```{r violinPlot_genes_transcripts_percent.mt_3456cells, fig.height=10, fig.width=5}
# there are a lot of cells with excessive mitochondrial transcripts
# the data looks bimodal for nGene
# lot of cells have excessive mitochondrial gene expression

p1 <- metadata %>% 
        ggplot(aes(x = sample, y = nUMI)) +
        geom_violin(colour = "blue", fill = "blue", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.1)

p2 <- metadata %>% 
        ggplot(aes(x = sample, y = nGene)) +
        geom_violin(colour = "brown1", fill = "brown1", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.1)

p3 <- metadata %>% 
        ggplot(aes(x = sample, y = percent.mt)) +
        geom_violin(colour = "cadetblue", fill = "cadetblue", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.1)

p1 / p2 / p3    
                       
```

```{r Plot_UMIcountsPerCell_3456cells}

# visualise number of UMIs (transcripts) per cell using all 3456 cells. 
# In geom_density use (y = ..scaled..) value that stat_density provides.
# Shape of curve has small hump on the left - possibly reflects cancerous and non-cancerous cell populations?

metadata %>% 
        ggplot(aes(x = nUMI)) +
        geom_density(aes(y = ..scaled..)) +
        scale_x_log10() +
        theme_classic() +
        ylab("cell density") +
        geom_vline(xintercept = 500, color = "blue")
```

```{r mitochondrial_thresholds}
# number of cells with percent.mt > 10%: 911 cells
# number of cells with percent.mt < 10%: 2545 cells

nrow(metadata[metadata$percent.mt > 10, ])

nrow(metadata[metadata$percent.mt < 10, ])
```

```{r filtered_seuratObject_CellLevel}
# CELL LEVEL FILTERING - leaves 25K genes and 2545 cells
#-------------------------------------------------------------------------------------------

# remove cells with percent mitochondrial gene expression > 10 and where no.genes/cell < 200.

filt.ons76cbo.seurat <- subset(ons76cbo.seurat, subset = nFeature_RNA > 200 & percent.mt < 10)
dim(filt.ons76cbo.seurat)
```

```{r violinPlot_genes_transcripts_percent.mt_2545cells, fig.height=10, fig.width=5}
# create a metadata dataframe
# new plots show remaining data looks unimodal

metadata.2 <- filt.ons76cbo.seurat@meta.data

# Add cell IDs to metadata
metadata.2$cells <- rownames(metadata.2)

# Rename columns to be more intuitive
metadata.2 <- metadata.2 %>%
        dplyr::rename(sample = orig.ident,
                      nUMI = nCount_RNA,
                      nGene = nFeature_RNA)

p4 <- metadata.2 %>% 
        ggplot(aes(x = sample, y = nUMI)) +
        geom_violin(colour = "blue", fill = "blue", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.1)

p5 <- metadata.2 %>% 
        ggplot(aes(x = sample, y = nGene)) +
        geom_violin(colour = "brown1", fill = "brown1", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.1)

p6 <- metadata.2 %>% 
        ggplot(aes(x = sample, y = percent.mt)) +
        ylim(0, 100) +
        geom_violin(colour = "cadetblue", fill = "cadetblue", alpha = 0.1) +
        theme_classic() +
        geom_jitter(alpha = 0.1)

p4 / p5 / p6
```

```{r extract_countsSlot_2545cells}
# subset '@counts' slot (matrix) of filtered seurat object

counts.2545 <- GetAssayData(object = filt.ons76cbo.seurat, slot = "counts")
```

```{r summarise_stats_2545cells}
# median genes per cell and expression sum, as above. Min. genes = 392. Median = 3198.
# Median genes is higher than before cells with high mitoch reads filtered out, which makes sense as the latter cells typically have lower number of genes expressed

at_least_one.2545 <- apply(counts.2545, 2, function(x) sum(x>0))

summary(at_least_one.2545)

summary(colSums(counts.2545))
```

```{r Plots_GenesPerCell_2545cells, fig.height=10, fig.width=8}
# 1 major peak now, but still a group of cells to the left with lower gene expression. 
# There are a few cells in the tail to the right of each graph which look like outliers with 
# very high gene expression.

par(mfrow = c(2, 1))

hist(at_least_one.2545, breaks = 100,
     main = "Distribution of detected genes",
     xlab = "Genes with at least one transcript")

hist(colSums(counts.2545),                             # total counts (expression) per cell
     breaks = 100, main = "Expression sum per cell",
     xlab = "Sum expression")
```

```{r Plots_UMI_gene_expression_2545cells, fig.height=10, fig.width=5}

# looks ok, most of the filtered cells have UMI counts ~ 10,000 and express > 3000 genes

p7 <- metadata.2 %>% 
       ggplot(aes(x = nUMI)) +
        geom_density(aes(y = ..scaled..), colour = "blue", fill = "blue", alpha = 0.2) +
        theme_classic() +
        scale_x_log10() +
        ylab("Cell density")

p8 <- metadata.2 %>% 
        ggplot(aes(x = nGene)) +
        geom_density(aes(y = ..scaled..), colour = "brown1", fill = "brown1", alpha = 0.2) +
        theme_classic() +
        scale_x_log10() +
        ylab("Cell density")

p7 / p8
```

```{r Plot_UMI_v_genes_2252cells}
# plot of UMI count vs no. of genes shows a positive correlation

p9 <- FeatureScatter(filt.ons76cbo.seurat, 
                      feature1 = "nCount_RNA",        # nCount_RNA is same as UMI count
                      feature2 = "nFeature_RNA")

p9
```

```{r filtered_seuratObject_GeneLevel}
# GENE LEVEL FILTERING (applied on cell-filtered Seurat object)
#-----------------------------------------------------------------------------------

#  we will have many genes with zero counts, which can dramatically reduce the average expression for a cell and so we will remove them from our data. Remove genes that have zero expression in all cells.

# # Extract counts
counts <- GetAssayData(object = filt.ons76cbo.seurat, slot = "counts")

# Output a logical vector for every gene on whether there are more than zero counts per cell
nonzero <- counts > 0

# If a gene is only expressed in a handful of cells, it is not particularly meaningful as it still brings down the averages for all other cells it is not expressed in. For our data we choose to keep only genes which are expressed in 10 or more cells.
# Sums all TRUE values and returns TRUE if more than 10 TRUE values per gene
keep_genes <- Matrix::rowSums(nonzero) >= 10    # sum the TRUE values in a row with 'gene' as rowname

# Only keeping those genes expressed in more than 10 cells
filtered_counts <- counts[keep_genes, ]

# Reassign to filtered Seurat object
filtCellsGenes.ons76cbo.seurat <- CreateSeuratObject(filtered_counts, meta.data = filt.ons76cbo.seurat@meta.data)

dim(filtCellsGenes.ons76cbo.seurat)       # number of genes further reduced to ~ 20K genes vs 25K genes                                                      for only cell-level filtering
```

```
# saved the 'filtCellsGenes.ons76cbo.seurat' object to outputs50K folder

saveRDS(filtCellsGenes.ons76cbo.seurat, file = "./outputs50K/ONS76-CBO/filtCellsGenes.ons76cbo.seurat.rds")
```

Load the .RDS object
```{r}
filt.ons76sbo <- readRDS("outputs50K/ONS76-CBO/filtCellsGenes.ons76cbo.seurat.rds")
```

-------------------------------------------------------------------------------------
***Remove mitochondrial and ribosomal genes***
```{r remove_ribo_genes}
# FILTER OUT ALL RIBOSOMAL GENES, THEN ALL MITOCHONDRIAL GENES (ON THE CELLS WHICH HAVE ALREADY BEEN REMOVED
# THAT HAVE PERCENT.MITO < 10% BEFORE DOING THE CLUSTERING), DON'T REMOVE ANY MORE ACTUAL CELLS
#-------------------------------------------------------------------------------------

# remove ribosomal genes first
filtRibo.ons76cbo <- filt.ons76sbo[ ! grepl('^RP[SL]', rownames(filt.ons76sbo)), ]
```

```{r check_ribo_removed}
# this has worked - 97 ribosomal genes have been removed from the new Seurat object
dim(filt.ons76sbo)
dim(filtRibo.ons76cbo)
```

Remove mitochondrial genes
```{r remove_mitoch_genes}
# remove mitochondrial genes 
filtRiboMito.ons76cbo <- filtRibo.ons76cbo[ ! grepl("^MT-", rownames(filtRibo.ons76cbo)), ]
```

```{r check_filtering}
# check if ribosome and mitochondrial gene filtering works - it does!
dim(filtRibo.ons76cbo)
dim(filtRiboMito.ons76cbo)
```

```
# save the CELL, GENE, MITOCH, RIBOSOMAL gene filtered object
saveRDS(filtRiboMito.ons76cbo, file = "./outputs50K/ONS76-CBO/filtRiboMitoCellsGenes.ons76cbo.rds")
```

```{r}
#--------------------------------------------------------------------------------------------
#--------------------------------------------------------------------------------------------
# EXPLORE SOURCES OF UNWANTED VARIATION, e.g. cell cycle
```

```{r}
# load the correct filtered object
filtRiboMito.ons76cbo <- readRDS("outputs50K/ONS76-CBO/filtRiboMitoCellsGenes.ons76cbo.rds")
```


Run SCTransform on the seurat object below - fully filtered object for cells, genes, ribosomal and mitochondrial genes
```{r}
# normalise data
phase.filtRiboMito.ons76cbo.sct <- NormalizeData(filtRiboMito.ons76cbo)

# run cell cycle scoring
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
phase.filtRiboMito.ons76cbo.sct <- CellCycleScoring(phase.filtRiboMito.ons76cbo.sct,
                                 g2m.features = g2m.genes,
                                 s.features = s.genes)
```

First run SCTransfrom WITHOUT vars.to.regress argument
```{r DimplotSCTRegressOutCCdifference}
# there is clustering by cell cycle phase, but not that many cells in G2M and S phase.
phase.filtRiboMito.ons76cbo.sct <- SCTransform(phase.filtRiboMito.ons76cbo.sct,
                                   method = "glmGamPoi", 
                                   verbose = FALSE)

phase.filtRiboMito.ons76cbo.sct <- RunPCA(phase.filtRiboMito.ons76cbo.sct, features = c(s.genes, g2m.genes))

DimPlot(phase.filtRiboMito.ons76cbo.sct, reduction = "pca", group.by = "Phase")
```

Then run SCTransform with vars.to.regress argument = cell cycle difference
```{r DimplotSCTRegressOutCCdifferenceFiltRiboMito}
# now G1 and G2/M are just one cluster, separate from G1, i.e. there are 2 clusters instead of 3.

phase.filtRiboMito.ons76cbo.sct$CC.difference <- phase.filtRiboMito.ons76cbo.sct$S.Score - phase.filtRiboMito.ons76cbo.sct$G2M.Score # get the cell cycle difference

phase.filtRiboMito.ons76cbo.sct <- SCTransform(phase.filtRiboMito.ons76cbo.sct,
                                   vars.to.regress = "CC.difference",
                                   method = "glmGamPoi", 
                                   verbose = FALSE)

phase.filtRiboMito.ons76cbo.sct <- RunPCA(phase.filtRiboMito.ons76cbo.sct, features = c(s.genes, g2m.genes))

DimPlot(phase.filtRiboMito.ons76cbo.sct, reduction = "pca", group.by = "Phase")

```

```
# saved 'phase' seurat object on which the PCA algorithm (for cell-cycle genes ONLY) has already been run

saveRDS(phase.filtRiboMito.ons76cbo.sct, "./outputs50K/ONS76-CBO/phase.filtRiboMito.ons76cbo.SCT.regressedCCdiff.rds")
```

```{r}
sessionInfo()
```


